

##Trying iterative downsampling with SERBP1
```{r}

#junction <- "chr1:67424977:67425082:-" 
#gene <- "SERBP1"

junction <- "chr7:80672040:80672754:+" 
gene <- "CD36"

junction.of.interest <- junction

intron.meta.file.path <- "/gpfs/commons/groups/landau_lab/SF3B1_splice_project/12.Bulk.PSI/merging.patient.data/MDSP5.MDSP6.merged.intron.metadata.txt"

intron.meta <- read.table(intron.meta.file.path, sep="\t", header = T)
intron.meta$intron.junction <- paste0(intron.meta$chr,":",intron.meta$start,":",intron.meta$end,":",intron.meta$strand)
intron.meta

clusterID <- paste(intron.meta[which(intron.meta$intron.junction == junction.of.interest),]$clusterID_5p)

intron.meta.sub <- intron.meta[which(intron.meta$clusterID_5p == clusterID),]
rownames(intron.meta.sub) <- intron.meta.sub$intron.junction
intron.meta.sub

```


##Subset MDS.combined matrix to only look at the SERBP1 cluster for that particualr junction

##Use the 4000 cell bucket and also 


```{r}
#count.matrix.sub <- read.table("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/16.Pseudotime.Window/IterativeDownsampling/cluster.count.matrix.subsets/SERBP1.clu_40101.count.matrix.sub.txt")

count.matrix.info <- read.table("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/16.Pseudotime.Window/IterativeDownsampling/cluster.count.matrix.subsets/CD36.clu_374890.count.matrix.sub.txt")

#rm(count.matrix.sub)
count.matrix.info <- count.matrix.info[-1]

#count.matrix.sub <- m[rownames(intron.meta.sub),1:5]

```


```{r}
length(ont.raw.counts.per.cell.per.window.per.gene[[gene]])
```


##Grab all the cells in bucket one that are MUT and WT cells

##Should we downsample cells ???




```{r}

load("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/4.Integrated_seurat_objs/ONT.seruat.objects/mds.ont.samples.integrated.rna.normscale.adjusted.rownames.Rdata")
mds.ont.samples.integrated@meta.data


num.mut.cells.per.bin <- as.data.frame(matrix(ncol=1, nrow=length(ont.raw.counts.per.cell.per.window.per.gene[[gene]])))
rownames(num.mut.cells.per.bin) <- 1:length(ont.raw.counts.per.cell.per.window.per.gene[[gene]])
colnames(num.mut.cells.per.bin) <- "num.mut.cells"

for (i in 1:length(ont.raw.counts.per.cell.per.window.per.gene[[gene]])){
  
  window.info <- ont.raw.counts.per.cell.per.window.per.gene[[gene]][[i]]
  num.mut.cells.per.bin[i,] <- nrow(window.info[which(window.info$genotype == "MUT"),])
}

num.mut.cells.per.bin
  
window.info <- ont.raw.counts.per.cell.per.window.per.gene[[gene]][[1]]

mut.cells <- rownames(window.info[which(window.info$genotype == "MUT"),])
mut.cells <- mds.ont.samples.integrated@meta.data[mut.cells,]$adjusted.rownames
#wt.cells <- rownames(window.info[which(window.info$genotype == "WT"),])
#wt.cells <- mds.ont.samples.integrated@meta.data[wt.cells,]$adjusted.rownames
length(mut.cells)

```


##Instead of downsampling across the full matrix, maybe do the sum across all the cells, and then downsample across the 4 rows. 

```{r}
downsamp_one = function(x,n){
    hist(sample(rep(1:length(x),x),size = n,replace=F),breaks=(0:length(x)+.5),plot=F)$counts
}
rep(1:length(x),x)

```

```{r}

##Subset the count matrix to only include mut cells from a specific bin

num.bins <- length(ont.raw.counts.per.cell.per.window.per.gene[[gene]])

per.bin.info <- as.data.frame(matrix(ncol=3, nrow=num.bins))
rownames(per.bin.info) <- 1:num.bins
colnames(per.bin.info) <- c("mean.psi","C.I.lower","C.I.upper")

num.iterations <- 1000
mutpsi.per.iteration <- as.data.frame(matrix(ncol=num.iterations, nrow=1))
rownames(mutpsi.per.iteration) <- c(junction)
colnames(mutpsi.per.iteration) <- 1:num.iterations



for (bin in 1:num.bins){
  window.info <- ont.raw.counts.per.cell.per.window.per.gene[[gene]][[bin]]
  mut.cells <- rownames(window.info[which(window.info$genotype == "MUT"),])
  mut.cells <- mds.ont.samples.integrated@meta.data[mut.cells,]$adjusted.rownames
  
  count.matrix.bin.sub <- count.matrix.info[,mut.cells]
  
  for (i in 1:num.iterations){
  intron.meta.sub.cp <- intron.meta.sub
  intron.meta.sub.cp$mut.bulk.counts <- rowSums(count.matrix.bin.sub)
  intron.meta.sub.cp$mut.bulk.counts.ds <- downsamp_one(intron.meta.sub.cp$mut.bulk.counts, 199)
  
  intron.meta.sub.cp <- intron.meta.sub.cp %>% group_by(clusterID_5p) %>% mutate(mut.bulk.psi = (mut.bulk.counts.ds/sum(mut.bulk.counts.ds))*100)
  
  mutpsi.per.iteration[,i] <- as.numeric(intron.meta.sub.cp[which(intron.meta.sub.cp$intron.junction == junction),"mut.bulk.psi"])

  }
  
  x <- as.numeric(mutpsi.per.iteration[1,])
  per.bin.info[bin,"mean.psi"] <- mean(x)
  per.bin.info[bin,"C.I.lower"] <- t.test(x)$conf.int[1]
  per.bin.info[bin,"C.I.upper"] <- t.test(x)$conf.int[2]
  
}

per.bin.info$bin <- rownames(per.bin.info)
per.bin.info$bin <- factor(per.bin.info$bin , levels = 1:nrow(per.bin.info))


```



```{r}
ggplot(per.bin.info) +
    geom_bar( aes(x=bin, y=mean.psi), stat="identity", fill="steel blue") +
    geom_errorbar( aes(x=bin, ymin=C.I.lower, ymax=C.I.upper), width=0.4, colour="orange", alpha=0.9, size=1.3)
```


```{r}
ggplot(per.bin.info) +
    geom_bar( aes(x=bin, y=mean.psi), stat="identity", fill="steel blue") +
    geom_errorbar( aes(x=bin, ymin=C.I.lower, ymax=C.I.upper), width=0.4, colour="orange", alpha=0.9, size=1.3)
```



```{r}
t.test(x)
t.test(x)$conf.int[1]
t.test(x)$conf.int[2]
mean(x)



mean(as.numeric(mutpsi.per.iteration[1,]))
sd(as.numeric(mutpsi.per.iteration[1,]))

get.conf.int = function(x) t.test(x)$conf.int

mean(as.numeric(mutpsi.per.iteration[1,]))
t.test(as.numeric(mutpsi.per.iteration[1,]))


#Calculating CI:
error <- qt(0.975,df=length(x)-1)*sd(x)/sqrt(length(x))
left <- mean(x)-error
right <- mean(x)+error



```



```{r}
ont.raw.counts.per.cell.per.window.per.gene[[gene]][[1]]

cluster.cov.mut.per.window[junction.final,]

min.counts <- min(t(cluster.cov.mut.per.window[junction.final,]))
min.counts <- 100

##Subset the count matrix to only include cells from bin 1
count.matrix.bin.sub <- count.matrix.info[,mut.cells]

library(scater)
rowSums(count.matrix.bin.sub)

##Downsample the count matrix to have only the minimum number of reads across all buckets 
set.seed(100)

prop.val <- min.counts/sum(count.matrix.bin.sub)
trial <- downsampleMatrix(as.matrix(count.matrix.bin.sub), prop=prop.val, bycol = TRUE)
rowSums(as.data.frame(trial))

```






```{r}

```


```{r}
intron.meta.sub.cp <- downsample.counts(intron.meta.sub)
```





```{r}
num.iterations <- 10000
dPSI.per.iteration <- as.data.frame(matrix(ncol=num.iterations, nrow=1))
mutpsi.per.iteration <- as.data.frame(matrix(ncol=num.iterations, nrow=1))

rownames(dPSI.per.iteration) <- c(junction)
colnames(dPSI.per.iteration) <- 1:num.iterations


intron.meta.sub.cp <- intron.meta.sub

for (i in 1:num.iterations){
  mut.cells.sub <- sample(mut.cells, num.cells.ds)
  wt.cells.sub <- sample(wt.cells, num.cells.ds)
  intron.meta.sub.cp$mut_bulk_counts <- rowSums(count.matrix.sub[,mut.cells.sub])
  intron.meta.sub.cp$wt_bulk_counts <- rowSums(count.matrix.sub[,wt.cells.sub])
  
  intron.meta.sub.cp <- intron.meta.sub.cp %>% group_by(clusterID_5p) %>% mutate(cluster_cov_mut_bulk = sum(mut_bulk_counts))
  intron.meta.sub.cp <- intron.meta.sub.cp %>% group_by(clusterID_5p) %>% mutate(cluster_cov_wt_bulk = sum(wt_bulk_counts))

  ##Calculating PSI values 
  intron.meta.sub.cp <- intron.meta.sub.cp %>% group_by(clusterID_5p) %>% mutate(mut_bulk_psi = (mut_bulk_counts/sum(mut_bulk_counts))*100)
  intron.meta.sub.cp <- intron.meta.sub.cp %>% group_by(clusterID_5p) %>% mutate(wt_bulk_psi = (wt_bulk_counts/sum(wt_bulk_counts))*100)
  intron.meta.sub.cp$dPSI <- (intron.meta.sub.cp$mut_bulk_psi - intron.meta.sub.cp$wt_bulk_psi)
  
  intron.meta.sub.cp <- as.data.frame(intron.meta.sub.cp)
  rownames(intron.meta.sub.cp) <- intron.meta.sub.cp$intron.junction
  dPSI.per.iteration[junction,i] <- intron.meta.sub.cp[junction,"dPSI"]
}


trial <- as.data.frame(t(dPSI.per.iteration))

mean(trial$`chr1:67424977:67425082:-`)
sd(trial$`chr1:67424977:67425082:-`)

quantile(trial$`chr1:67424977:67425082:-`,.025)
quantile(trial$`chr1:67424977:67425082:-`,.975)
```



```{r}
intron.meta.sub[,9:16]
```


##Combined function

```{r}

library(plyr)


per.bin.info <- list()

for (i in 1:length(ont.raw.counts.per.cell.per.window.per.gene[[gene]])){
  
  print(paste0("bin ",i))
  window.info <- ont.raw.counts.per.cell.per.window.per.gene[[gene]][[i]]

  mut.cells <- rownames(window.info[which(window.info$genotype == "MUT"),])
  mut.cells <- mds.ont.samples.integrated@meta.data[mut.cells,]$adjusted.rownames
  wt.cells <- rownames(window.info[which(window.info$genotype == "WT"),])
  wt.cells <- mds.ont.samples.integrated@meta.data[wt.cells,]$adjusted.rownames

  length(mut.cells)
  length(wt.cells)

  min.num.cells <- min(length(wt.cells),length(mut.cells))
  num.cells.ds <- round_any(((min.num.cells)*2)/3 ,10)
  
  num.iterations <- 100
  dPSI.per.iteration <- as.data.frame(matrix(ncol=num.iterations, nrow=1))
  mutpsi.per.iteration <- as.data.frame(matrix(ncol=num.iterations, nrow=1))
  wtpsi.per.iteration <- as.data.frame(matrix(ncol=num.iterations, nrow=1))

  rownames(dPSI.per.iteration) <- c(junction)
  colnames(dPSI.per.iteration) <- 1:num.iterations
  
  rownames(mutpsi.per.iteration) <- c(junction)
  colnames(mutpsi.per.iteration) <- 1:num.iterations
  
  rownames(wtpsi.per.iteration) <- c(junction)
  colnames(wtpsi.per.iteration) <- 1:num.iterations
  

  intron.meta.sub.cp <- intron.meta.sub
  
  for (i in 1:num.iterations){
    mut.cells.sub <- sample(mut.cells, num.cells.ds)
    wt.cells.sub <- sample(wt.cells, num.cells.ds)
    intron.meta.sub.cp$mut_bulk_counts <- rowSums(count.matrix.sub[,mut.cells.sub])
    intron.meta.sub.cp$wt_bulk_counts <- rowSums(count.matrix.sub[,wt.cells.sub])
  
    intron.meta.sub.cp <- intron.meta.sub.cp %>% group_by(clusterID_5p) %>% mutate(cluster_cov_mut_bulk = sum(mut_bulk_counts))
    intron.meta.sub.cp <- intron.meta.sub.cp %>% group_by(clusterID_5p) %>% mutate(cluster_cov_wt_bulk = sum(wt_bulk_counts))

    ##Calculating PSI values 
    intron.meta.sub.cp <- intron.meta.sub.cp %>% group_by(clusterID_5p) %>% mutate(mut_bulk_psi = (mut_bulk_counts/sum(mut_bulk_counts))*100)
    intron.meta.sub.cp <- intron.meta.sub.cp %>% group_by(clusterID_5p) %>% mutate(wt_bulk_psi = (wt_bulk_counts/sum(wt_bulk_counts))*100)
    intron.meta.sub.cp$dPSI <- (intron.meta.sub.cp$mut_bulk_psi - intron.meta.sub.cp$wt_bulk_psi)
  
    intron.meta.sub.cp <- as.data.frame(intron.meta.sub.cp)
    rownames(intron.meta.sub.cp) <- intron.meta.sub.cp$intron.junction
    
    dPSI.per.iteration[junction,i] <- intron.meta.sub.cp[junction,"dPSI"]
    mutpsi.per.iteration[junction,i] <- intron.meta.sub.cp[junction,"mut_bulk_psi"]
    wtpsi.per.iteration[junction,i] <- intron.meta.sub.cp[junction,"wt_bulk_psi"]
  }
  
  
  per.bin.info[[i]][["dPSI"]] <- dPSI.per.iteration
  per.bin.info[[i]][["mutpsi"]] <- mutpsi.per.iteration
  per.bin.info[[i]][["wtpsi"]] <- wtpsi.per.iteration

  output <- as.data.frame(t(dPSI.per.iteration))

  print(paste0("mean: ", mean(output$`chr1:67424977:67425082:-`)))
  print(paste0("sd: " ,sd(output$`chr1:67424977:67425082:-`)))

  print(quantile(output$`chr1:67424977:67425082:-`,.025))
  print(quantile(output$`chr1:67424977:67425082:-`,.975))
  print("")
  
}
```

```{r}
mutpsi.per.iteration
wtpsi.per.iteration
```



```{r}
per.bin.info$dPSI
```




#TRIAL
```{r}
##Downsample to 1/2 the number of cells 
count.matrix.sub[,mut.cells.sub]
count.matrix.sub[,wt.cells.sub]


trial1 <- sample(mut.cells, num.cells.ds)
trial2 <- sample(mut.cells, num.cells.ds)

setdiff(trial1,trial2)

trial1 <- sample(mut.cells, num.cells.ds)
trial2 <- sample(mut.cells, num.cells.ds)

setdiff(trial1,trial2)

downSample(x, y, list = FALSE, yname = "Class")
```



##Try doing one single downsample (to the set win min number)

```{r}
library(plyr)

num.bins <- length(ont.raw.counts.per.cell.per.window.per.gene[[gene]])

per.bin.info <- as.data.frame(matrix(ncol= 3, nrow=num.bins))

rownames(per.bin.info) <- 1:num.bins
colnames(per.bin.info) <- c("dPSI", "mut.psi","wt.psi")

for (i in 1:num.bins){
  i <- 1
  #print(paste0("bin ",i))
  window.info <- ont.raw.counts.per.cell.per.window.per.gene[[gene]][[i]]

  mut.cells <- rownames(window.info[which(window.info$genotype == "MUT"),])
  mut.cells <- mds.ont.samples.integrated@meta.data[mut.cells,]$adjusted.rownames
  wt.cells <- rownames(window.info[which(window.info$genotype == "WT"),])
  wt.cells <- mds.ont.samples.integrated@meta.data[wt.cells,]$adjusted.rownames

  length(mut.cells)
  length(wt.cells)

  min.num.cells <- min(length(wt.cells),length(mut.cells))
  #num.cells.ds <- min.num.cells
  num.cells.ds <- round_any(((min.num.cells)*2)/3 ,10)

  num.iterations <- 10000
  dPSI.per.iteration <- as.data.frame(matrix(ncol=num.iterations, nrow=1))
  mutpsi.per.iteration <- as.data.frame(matrix(ncol=num.iterations, nrow=1))
  wtpsi.per.iteration <- as.data.frame(matrix(ncol=num.iterations, nrow=1))

  rownames(dPSI.per.iteration) <- c(junction)
  colnames(dPSI.per.iteration) <- 1:num.iterations
  
  rownames(mutpsi.per.iteration) <- c(junction)
  colnames(mutpsi.per.iteration) <- 1:num.iterations
  
  rownames(wtpsi.per.iteration) <- c(junction)
  colnames(wtpsi.per.iteration) <- 1:num.iterations
  

  intron.meta.sub.cp <- intron.meta.sub
  
  for (iter in 1:num.iterations){
    inter <- 1
    mut.cells.sub <- sample(mut.cells,num.cells.ds, replace =T)
    wt.cells.sub <- sample(wt.cells, num.cells.ds, replace =T)
    intron.meta.sub.cp$mut_bulk_counts <- rowSums(count.matrix.sub[,mut.cells.sub])
    intron.meta.sub.cp$wt_bulk_counts <- rowSums(count.matrix.sub[,wt.cells.sub])
  
    intron.meta.sub.cp <- intron.meta.sub.cp %>% group_by(clusterID_5p) %>% mutate(cluster_cov_mut_bulk = sum(mut_bulk_counts))
    intron.meta.sub.cp <- intron.meta.sub.cp %>% group_by(clusterID_5p) %>% mutate(cluster_cov_wt_bulk = sum(wt_bulk_counts))

    ##Calculating PSI values 
    intron.meta.sub.cp <- intron.meta.sub.cp %>% group_by(clusterID_5p) %>% mutate(mut_bulk_psi = (mut_bulk_counts/sum(mut_bulk_counts))*100)
    intron.meta.sub.cp <- intron.meta.sub.cp %>% group_by(clusterID_5p) %>% mutate(wt_bulk_psi = (wt_bulk_counts/sum(wt_bulk_counts))*100)
    intron.meta.sub.cp$dPSI <- (intron.meta.sub.cp$mut_bulk_psi - intron.meta.sub.cp$wt_bulk_psi)
  
    intron.meta.sub.cp <- as.data.frame(intron.meta.sub.cp)
    rownames(intron.meta.sub.cp) <- intron.meta.sub.cp$intron.junction
    intron.meta.sub
    intron.meta.sub.cp
    dPSI.per.iteration[junction,iter] <- intron.meta.sub.cp[junction,"dPSI"]
    mutpsi.per.iteration[junction,iter] <- intron.meta.sub.cp[junction,"mut_bulk_psi"]
    wtpsi.per.iteration[junction,iter] <- intron.meta.sub.cp[junction,"wt_bulk_psi"]
    #crytpic.cov.per.iteration[junction,iter] <- intron.meta.sub.cp[junction,"wt_bulk_psi"]
    
  }
  
  per.bin.info[i,"dPSI"] <- mean(as.numeric(dPSI.per.iteration[junction,]))
  per.bin.info[i,"mut.psi"] <- mean(as.numeric(mutpsi.per.iteration[junction,]))
  per.bin.info[i,"wt.psi"] <- mean(as.numeric(wtpsi.per.iteration[junction,]))
}

per.bin.info$window <- rownames(per.bin.info)
per.bin.info$window <-  factor(per.bin.info$window,levels = 1:nrow(per.bin.info))
```



```{r}
#1000
per.bin.info
ggplot(per.bin.info, aes(x=window, group = 1)) + 
  geom_line( aes(y=dPSI), color=dPSIColor) + 
  geom_hline(yintercept = 0, linetype = "dotted") + 
  theme_classic()+
  theme(plot.title = element_text(hjust = 0.5, size = 14), 
        axis.title.y = element_text(color =dPSIColor , size=13)) +
  ggtitle(paste0("dPSI and Expression for ", gene ," cryptic junction - ", junction)) 
```


```{r}
#100
per.bin.info
ggplot(per.bin.info, aes(x=window, group = 1)) + 
  geom_line( aes(y=dPSI), color=dPSIColor) + 
  geom_hline(yintercept = 0, linetype = "dotted") + 
  theme_classic()+
  theme(plot.title = element_text(hjust = 0.5, size = 14), 
        axis.title.y = element_text(color =dPSIColor , size=13)) +
  ggtitle(paste0("dPSI and Expression for ", gene ," cryptic junction - ", junction)) 
```

