
##Motif Analysis
##Generating counts per Pseudotime window for MDS 


##Using the Object grab the HSPC, IMP, MEP, EP cells, rank by pseudotime and sperate into quantiles.
```{r}

##----------------- Setting important variables   -----------------##

seurat.object.path <- "/gpfs/commons/groups/landau_lab/SF3B1_splice_project/4.Integrated_seurat_objs/mds.sample.integrated.p4.5.6.in.progress_2_withPseudotime.RData"
#ch.samples.integrated

junction.meta.file.path <- "/gpfs/commons/groups/landau_lab/SF3B1_splice_project/13.Motif.Analysis/junction.matricies/CH.combined.alt3p.junction.info.txt"

junction.count.file.path <- "/gpfs/commons/groups/landau_lab/SF3B1_splice_project/9.Splice_pipeline_example_run/CH259/output_files/leafcutter_outputs/CH259_output/CH259_perind_numbers.counts.txt"

genotyping.table.file.path <- "/gpfs/commons/groups/landau_lab/SF3B1_splice_project/3.Genotyping_info/ch_259.genotype.info.with.cell.type.txt"


pseudotime.column <- "pseudotime_monocle3"
genotype.column <- "Genotype_1UMI"
cluster.column = "CellType" 
cluster.types = c("HSPC","IMP","MEP","EP")

##----------------- Load in Seurat Object   -----------------##

print("Loading in seurat object")
load(seurat.object.path) #ch.samples.integrated

##----------------- Load in counts matrix -----------------##

print("Loading count matrix")
m <- fread(junction.count.file.path, data.table=FALSE)
rownames(m) <- m$intron_junction
m <- m[,-1]

##----------------- Grab the genotyped cells that we want to use (only MUT) -----------------##

genotyping.table <- read.table(genotyping.table.file.path)
cells.to.use <- rownames(genotyping.table[which(genotyping.table[,genotype.column] =="MUT" & genotyping.table$Cell.Assignment_2 %in% cluster.types),])
length(cells.to.use)

##----------------- Load metadata, remove NA's from pseudotime column and subset the count matrix to only include relevant cells and junctions -----------------##

#meta = as.data.frame(mds.samples.integrated@meta.data[which(mds.samples.integrated@meta.data[,cluster.column] %in% cluster.types & mds.samples.integrated@meta.data$Genotype_2UMI == "MUT"),])
cell.meta = as.data.frame(ch.samples.integrated@meta.data[cells.to.use,])
cell.meta <- cell.meta[!is.na(cell.meta[,pseudotime.column]),]
shared.cells <- intersect(colnames(m), rownames(cell.meta))

print("number of cell in long read data")
ncol(m)
print("number of cell in short read data")
nrow(cell.meta)
print("number of shared cells")
length(shared.cells)

junction.meta <- read.table(junction.meta.file.path)
junction.meta <- junction.meta[!duplicated(junction.meta$intron.junction),]
rownames(junction.meta) <- junction.meta$intron.junction
shared.junctions <- intersect(rownames(m), rownames(junction.meta))


#Subset the cell.metadata, junction count matrix and junction.metadata to have the same cells and junctions
cell.meta.sub = meta[shared.cells,]
m.sub = m[shared.junctions,shared.cells]
junction.meta.sub <- junction.meta[shared.junctions,]

##------------------------------- Order cells in pseudotime and group by quantile ------------------------------- ##

ord = rownames(cell.meta.sub)[order(cell.meta.sub[,pseudotime.column],decreasing = F)]
cell.meta.sub = cell.meta.sub[ord,]
m.sub = m.sub[,ord]

num.quantiles <- 4
cell.meta.sub$quantile <- ntile(cell.meta.sub[,pseudotime.column], num.quantiles)
table(cell.meta.sub$quantile)

##------------------------------- Get mut.psi info for each quantile and add to table ------------------------------- ##


for (i in 1:num.quantiles){
  print(i)
  #i <- 1
  quantile.cells <- rownames(cell.meta.sub[which(cell.meta.sub$quantile == i),])
  col.name <- paste0("Q",i,".junc.cov")
  junction.meta.sub[,col.name] <- rowSums(m.sub[,quantile.cells])
}


```



##Add in the results to the existing output table (with the junction counts for each quantile as separate columns)

```{r}

write.table(junction.meta.sub, file="/gpfs/commons/groups/landau_lab/SF3B1_splice_project/13.Motif.Analysis/junction.matricies/CH.combined.alt3p.junction.info.w.quantile.counts.txt" )
```



##Looking at MDS table

```{r}

trial <- read.table("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/13.Motif.Analysis/junction.matricies/MDS.untreated.alt3p.junction.info.w.quantile.counts.txt")

trial
```

