
##Generate a list of cryptic junctions to look into!

```{r}
#First trial run 
#junctions.of.interest <- c("chr20:35556271:35556954:+","chr2:169812524:169815476:-","chr1:67424977:67425082:-","chrX:119625396:119629317:-","chr17:44322552:44322807:-")


#Full list of genes from Gene cards that we shared with Omar and Justin
junctions.of.interest <- c("chr20:35556271:35556954:+",
          "chr2:169812524:169815476:-",
          "chr1:67424977:67425082:-", 
          "chr1:161167234:161167338:+",  
          "chr2:241335212:241335958:+", 
          "chr2:106164799:106166055:-",   
          "chr17:44322552:44322807:-", 
          "chr1:100992754:100995109:-", 
          "chr16:90027722:90031175:+", 
          "chr12:120496216:120496401:+", 
          "chr1:67424322:67424887:-",
          "chr15:58917020:58932355:-",
          "chr17:47223057:47223495:+",
          "chr6:90560234:90561621:-",
          "chrX:75071704:75073688:-", 
          "chr17:42565311:42565605:+",
          "chr17:42562219:42562355:+",
          "chr5:140529796:140537337:+",
          "chr16:30407664:30408034:+",
          "chrX:119625396:119629317:-")

##Reduced list from  gene cards
junctions.of.interest <- c("chr20:35556271:35556954:+",
          "chr2:169812524:169815476:-",
          "chr1:67424977:67425082:-", 
          "chr1:161167234:161167338:+",  
          "chr2:241335212:241335958:+", 
          "chr2:106164799:106166055:-",   
          "chr17:44322552:44322807:-", 
          "chr1:100992754:100995109:-", 
          "chr12:120496216:120496401:+", 
          "chr1:67424322:67424887:-",
          "chr15:58917020:58932355:-",
          "chr17:47223057:47223495:+",
          "chrX:75071704:75073688:-", 
          "chr16:30407664:30408034:+",
          "chrX:119625396:119629317:-")


##Chekcing out riosomal Poisioning - Grab it from a txt file
junctions <- read.table("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/7.Permutation_junction_output/4.Comb_patients_merge_counts/CH_combined/logOR_within_cell_type_ONLY_CRYPTIC_3P_Junctions_with_threshold_info.txt")

junctions[which(junctions$gene == "GATA1"),]


junctions.sub <- junctions[grepl("GATA1|SLC25A39|ERGIC3|DPH5|TMEM141|DUSP12", junctions$gene),]
rownames(junctions.sub) <- junctions.sub$intron_junction

junctions.of.interest <- paste(junctions.sub$intron_junction)
junctions.of.interest.final <- paste0(junctions.sub$gene,":",junctions.sub$intron_junction)


junctions.sub <- junctions[grepl("RPL|RPS|GATA1", junctions$gene),]
junctions.of.interest <- paste(junctions.sub[which(junctions.sub$pvalue <= 0.05),]$intron_junction)

##HSPC specific cryptic junctions:
junctions.hspc <- read.table("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/7.Permutation_junction_output/5.Comb_patients_ind_celltype_merge_counts/CH_combined/HSPC/logOR_within_cell_type_ONLY_CRYPTIC_3P_Junctions_with_threshold_info.txt")

##EP specific cryptic junctions:
junctions.ep <- read.table("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/7.Permutation_junction_output/5.Comb_patients_ind_celltype_merge_counts/CH_combined/EP/logOR_within_cell_type_ONLY_CRYPTIC_3P_Junctions_with_threshold_info.txt")

junctions.hspc.list <- junctions.hspc[which(junctions.hspc$pvalue <= 0.05 & junctions.hspc$dPSI >= 5),]$intron_junction
junctions.ep.list <- junctions.ep[which(junctions.ep$pvalue <= 0.05 & junctions.ep$dPSI >= 5),]$intron_junction

shared.junctions <- intersect(junctions.hspc.list, junctions.ep.list)

junctions.of.interest.hspc <- setdiff(junctions.hspc.list,shared.junctions)
junctions.of.interest.ep <- setdiff(junctions.ep.list,shared.junctions)
length(junctions.of.interest.hspc)
length(junctions.of.interest.ep)
junctions.of.interest <- c(junctions.of.interest.hspc,junctions.of.interest.ep)
length(junctions.of.interest)
junctions.of.interest
```


#For each set, load the intron meta and count matrix and subset each by the set of junctions, calculate the raw counts for the junctions and subset again by just the cryptic junctions before merging
```{r}
##------------------------------- Arguments ------------------------------- ##
#seurat.object <- ch.samples.integrated@meta.data

intron.meta.file.path <- "/gpfs/commons/groups/landau_lab/SF3B1_splice_project/9.Splice_pipeline_example_run/CH259/output_files/leafcutter_outputs/CH259_output/CH259_all.introns.info.w.primary.annotations.txt"

junction.count.file.path <- "/gpfs/commons/groups/landau_lab/SF3B1_splice_project/9.Splice_pipeline_example_run/CH259/output_files/leafcutter_outputs/CH259_output/CH259_perind_numbers.counts.txt"

genotyping.table.file.path <- "/gpfs/commons/groups/landau_lab/SF3B1_splice_project/3.Genotyping_info/ch_259.genotype.info.with.cell.type.txt"

pseudotime.column <- "pseudotime_monocle3"
genotype.column <- "Genotype_1UMI"
step  <- 30
cluster.column = "CellType" 
cluster.types = c("HSPC","IMP","MEP","EP")


##------------------------------- Get intron metadata and junction count matrix  ------------------------------- ##

intron.meta <- read.table(intron.meta.file.path, sep=",", header = T)
intron.meta$intron.junction <- paste0(intron.meta$chr,":",intron.meta$start,":",intron.meta$end,":",intron.meta$strand_1)

clusterIDs <- paste(intron.meta[which(intron.meta$intron.junction %in% junctions.of.interest),]$clusterID_5p)
intron.meta.sub <- intron.meta[which(intron.meta$clusterID_5p %in% clusterIDs),]
rownames(intron.meta.sub) <- intron.meta.sub$intron.junction

library(data.table)
m.trial <- fread("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/9.Splice_pipeline_example_run/CH_combined/1.Counts_matrix/CH259_perind_numbers.counts.txt", sep="\t", data.table=FALSE)
m.trial
m <- read.table(junction.count.file.path, header = T)
rownames(m) <- m$intron_junction
m <- m[,-1]
colnames(m) <- paste0(colnames(m),"_1")
m[1:10,1:10]

genotyping.table <- read.table(genotyping.table.file.path)
```



##meta is already subsetted to only include genotyped cells and so is the count data 
```{r}

##------------------------------- Load the metadata & remove NA's from pseudotime column ------------------------------- ##

meta = as.data.frame(ch.samples.integrated@meta.data[which(ch.samples.integrated@meta.data$orig.ident == "CH_259" & ch.samples.integrated@meta.data$CellType %in% cluster.types & ch.samples.integrated@meta.data$Genotype_1UMI_10x != "NA"),])

#meta = as.data.frame(ch.samples.integrated@meta.data[which(ch.samples.integrated@meta.data$orig.ident == "CH_259" & ch.samples.integrated@meta.data$CellType %in% cluster.types),])
meta <- meta[!is.na(meta[,pseudotime.column]),]

#Subset the metadata and junction matrix to have the same cells (and remove any NAs)
shared.cells <- intersect(colnames(m), rownames(meta))
print("number of cell in long read data")
ncol(m)
print("number of cell in short read data")
nrow(meta)
print("number of shared cells")
length(shared.cells)

m.sub = m[,shared.cells]
m.meta = meta[shared.cells,]

ncol(m.sub)
nrow(m.meta)
#m.meta

##------------------------------- Order cells in pseudotime ------------------------------- ##

ord = rownames(m.meta)[order(m.meta[,pseudotime.column],decreasing = F)]
m.meta = m.meta[ord,]
m.sub = m[,ord]
#m.meta[1:10,1:10]
#m.sub[1:10,1:10]

# Get sliding windows
step.def = seq(from = 1, to = ncol(m.sub), by = step)
define.windows = list()

for(i in 1:(round(ncol(m.sub)/(step))-1)){
  define.windows[[i]] = c(step.def[i],step.def[i+10])
}

ind = lapply(define.windows, function(x) sum(is.na(x)) > 0)
define.windows = define.windows[!unlist(ind)]
#define.windows


##------------------------------- Get mean pseudotime value and cell type fractions per window ------------------------------- ##

pseudotime.mean = unlist(lapply(define.windows, function(x){
  meta.temp = m.meta[c(x[1]:x[2]),]
  mean(meta.temp[,pseudotime.column])
}))



ct.fractions <- list()
if(!is.null(cluster.column) & !is.null(cluster.types)){
  for (ct in cluster.types) {
    ct.fractions[[ct]] = unlist(lapply(define.windows, function(x){
      meta.temp = m.meta[c(x[1]:x[2]),]
      sum(meta.temp[,cluster.column] == ct) / nrow(meta.temp)
    }))
  }
}



ct.fractions.2 <- as.data.frame(matrix(ncol = length(define.windows), nrow = length(cluster.types)))
colnames(ct.fractions.2) <- as.character(1:ncol(ct.fractions.2))
rownames(ct.fractions.2) <- cluster.types

if(!is.null(cluster.column) & !is.null(cluster.types)){
  for (ct in cluster.types) {
    ct.fractions.2[ct,] = unlist(lapply(define.windows, function(x){
      meta.temp = m.meta[c(x[1]:x[2]),]
      sum(meta.temp[,cluster.column] == ct) / nrow(meta.temp)
    }))
  }
}


##------------------------------- Get mut.fraction per window ------------------------------- ##

mut.fraction = unlist(lapply(define.windows, function(x){
  
  meta.temp = m.meta[c(x[1]:x[2]),]
  shared.cells <- intersect(rownames(genotyping.table), rownames(meta.temp))
  genotyping.table.sub <- genotyping.table[shared.cells, ]
  wt <- rownames(genotyping.table.sub[which(genotyping.table.sub[,genotype.column] =="WT"),])
  mut <- rownames(genotyping.table.sub[which(genotyping.table.sub[,genotype.column] =="MUT"),])
  out = length(mut)/(length(wt)+ length(mut))
  
}))

##------------------------------- Get differential splicing score per window ------------------------------- ##

win.dPSI =  lapply(define.windows, function(x){
  
  #x <- unlist(define.windows[1])
  temp = m.sub[,c(x[1]:x[2])]
  meta.temp = m.meta[c(x[1]:x[2]),]
  intron.meta.sub.temp = intron.meta.sub
  intron.meta.sub.temp
  

  shared.cells <- intersect(rownames(genotyping.table), rownames(meta.temp))
  genotyping.table.sub <- genotyping.table[shared.cells, ]
  
  wt <- rownames(genotyping.table.sub[which(genotyping.table.sub[,genotype.column] =="WT"),])
  mut <- rownames(genotyping.table.sub[which(genotyping.table.sub[,genotype.column] =="MUT"),])
  message(paste(length(wt),"WT cells"))
  message(paste(length(mut),"MUT cells"))
  message(x[1],":",x[2])
  
  intron.meta.sub.temp$mut.bulk.counts <- rowSums(temp[rownames(intron.meta.sub.temp),mut])
  intron.meta.sub.temp$wt.bulk.counts <- rowSums(temp[rownames(intron.meta.sub.temp),wt])
  intron.meta.sub.temp <- intron.meta.sub.temp %>% group_by(clusterID_5p) %>% mutate(cluster.cov.mut=sum(mut.bulk.counts))
  intron.meta.sub.temp <- intron.meta.sub.temp %>% group_by(clusterID_5p) %>% mutate(cluster.cov.wt=sum(wt.bulk.counts))
  intron.meta.sub.temp <- intron.meta.sub.temp %>% group_by(clusterID_5p) %>% mutate(mut.psi = (mut.bulk.counts/cluster.cov.mut)*100)
  intron.meta.sub.temp <- intron.meta.sub.temp %>% group_by(clusterID_5p) %>% mutate(wt.psi = (wt.bulk.counts/cluster.cov.wt)*100)
  
  intron.meta.sub.temp$dPSI <- (intron.meta.sub.temp$mut.psi - intron.meta.sub.temp$wt.psi)
  #out = as.data.frame(intron.meta.sub.temp[,"dPSI"])
  
  out = as.data.frame(intron.meta.sub.temp[,c("mut.bulk.counts","wt.bulk.counts","cluster.cov.mut","cluster.cov.wt","mut.psi","wt.psi","dPSI")])
  rownames(out) <- paste0(intron.meta.sub.temp$intron.junction)
  out = out[junctions.of.interest,]
  return(out)

})

```



```{r}
#junctions.of.interest.final <- paste0(intron.meta.sub[junctions.of.interest,]$gene,":",intron.meta.sub[junctions.of.interest,]$intron.junction)


#rownames <- paste0(paste(junctions.sub[rownames(win.dPSI[[1]]),]$gene),":",rownames(win.dPSI[[1]]))
rownames <- junctions.of.interest.final

extracted_cols <- lapply(win.dPSI, function(x) x[,"dPSI"]) #note the added comma!
dPSI.per.window <- as.data.frame(do.call(cbind, extracted_cols))
colnames(dPSI.per.window) = as.character(1:ncol(dPSI.per.window))
rownames(dPSI.per.window) <- rownames

extracted_cols <- lapply(win.dPSI, function(x) x[,"mut.psi"]) #note the added comma!
mut.psi.per.window <- as.data.frame(do.call(cbind, extracted_cols))
colnames(mut.psi.per.window) = as.character(1:ncol(mut.psi.per.window))
rownames(mut.psi.per.window) <- rownames

extracted_cols <- lapply(win.dPSI, function(x) x[,"wt.psi"]) #note the added comma!
wt.psi.per.window <- as.data.frame(do.call(cbind, extracted_cols))
colnames(wt.psi.per.window) = as.character(1:ncol(wt.psi.per.window))
rownames(wt.psi.per.window) <- rownames

dPSI.per.window
mut.psi.per.window
wt.psi.per.window

#OLD CODE
#dPSI.per.window.orig <- dPSI.per.window
#dPSI.per.window = (do.call(cbind, win.dPSI))
#colnames(dPSI.per.window) = as.character(1:ncol(dPSI.per.window))
#out.zscores <- as.data.frame(t(apply(dPSI.per.window, 1, function(x) (x-mean(x))/sd(x))))
#mask <- apply(out, 2, is.nan)
#out.temp <- out
#out.temp[mask] <- 0
#out.zscores <- as.data.frame(t(apply(out.temp, 1, function(x) (x-mean(x))/sd(x))))
```

```{r}
gene.list
```


```{r}
ch.ont.samples.integrated@assays$RNA
```

```{r}
gene.list <- intersect(rownames(ch.ont.samples.integrated@assays$RNA@data), gene.list)
gene.list
```

##Getting the normalized values for each individual cell per window
```{r}

ont.raw.counts.per.cell.per.window.per.gene <- list()

for (gene in gene.list){
  
  ont.raw.counts.per.cell.per.window = lapply(define.windows, function(x){
  
    meta.temp = m.meta[c(x[1]:x[2]),]
    shared.cells <- intersect(rownames(meta.temp), rownames(ch.ont.samples.integrated@meta.data))
    message(x[1],":", x[2])

    counts <- as.data.frame(ch.ont.samples.integrated@assays$RNA@counts[gene,shared.cells])
    colnames(counts)[1] <- c("gene.counts")
    counts$total <- colSums(as.data.frame(ch.ont.samples.integrated@assays$RNA@counts[,shared.cells]))
  
    counts$norm.counts <- (counts$gene.counts/counts$total) * 10000
    counts$ln.norm.counts <- log(counts$norm.counts)
    counts$seurat.norm.counts <- ch.ont.samples.integrated@assays$RNA@data[gene,shared.cells]
    counts$genotype <- ch.ont.samples.integrated@meta.data[shared.cells,]$Genotype_1WT
    out = counts
    return(out)
  
  })
  
  ont.raw.counts.per.cell.per.window.per.gene[[gene]] <- ont.raw.counts.per.cell.per.window
}

ont.raw.counts.per.cell.per.window.per.gene$VPS35

```


```{r}

##------------------------------- ONT raw counts per window ------------------------------- ##

gene.list
ont.raw.counts.per.window = lapply(define.windows, function(x){
  
  x <- unlist(define.windows[1])
  meta.temp = m.meta[c(x[1]:x[2]),]
  shared.cells <- intersect(rownames(meta.temp), rownames(ch.ont.samples.integrated@meta.data))
  length(shared.cells)
  nrow(meta.temp)
  
  #wt <- rownames(ch.ont.samples.integrated[shared.cells,]@meta.data[which(ch.ont.samples.integrated@meta.data[shared.cells,"Genotype_1WT"] =="WT"),])
  #mut <- rownames(ch.ont.samples.integrated@meta.data[which(ch.ont.samples.integrated@meta.data[shared.cells,"Genotype_1WT"] =="MUT"),])
  
  ch.ont.samples.integrated.meta.sub <- ch.ont.samples.integrated@meta.data[shared.cells,]
  wt <- rownames(ch.ont.samples.integrated.meta.sub[which(ch.ont.samples.integrated.meta.sub$Genotype_1WT =="WT"),])
  mut <- rownames(ch.ont.samples.integrated.meta.sub[which(ch.ont.samples.integrated.meta.sub$Genotype_1WT =="MUT"),])

  message(paste(length(wt),"WT cells"))
  message(paste(length(mut),"MUT cells"))
  message(paste(length(shared.cells),"total cells"))
  
  message(x[1],":", x[2])
  bulk.counts <- as.data.frame(rowSums(as.data.frame(ch.ont.samples.integrated@assays$RNA@counts[gene.list,shared.cells])))
  colnames(bulk.counts) <- c("counts")
  bulk.counts
  out = bulk.counts
  return(out)
  
  })
  
ont.raw.counts.per.window <- as.data.frame(ont.raw.counts.per.window)
colnames(ont.raw.counts.per.window) <- as.character(1:ncol(ont.raw.counts.per.window))
ont.raw.counts.per.window

```

```{r}
ont.mut.raw.counts.per.window = lapply(define.windows, function(x){
  
  
  x <- unlist(define.windows[1])
  meta.temp = m.meta[c(x[1]:x[2]),]
  shared.cells <- intersect(rownames(meta.temp), rownames(ch.ont.samples.integrated@meta.data))
  ch.ont.samples.integrated.meta.sub <- ch.ont.samples.integrated@meta.data[shared.cells,]
  mut <- rownames(ch.ont.samples.integrated.meta.sub[which(ch.ont.samples.integrated.meta.sub$Genotype_1WT =="MUT"),])
  message(paste(length(mut),"MUT cells"))
  message(x[1],":", x[2])
  bulk.counts <- as.data.frame(rowSums(as.data.frame(ch.ont.samples.integrated@assays$RNA@counts[gene.list,mut])))
  colnames(bulk.counts) <- c("counts")
  bulk.counts
  out = bulk.counts
  return(out)
  
})

ont.mut.raw.counts.per.window <- as.data.frame(ont.mut.raw.counts.per.window)
colnames(ont.mut.raw.counts.per.window) <- as.character(1:ncol(ont.mut.raw.counts.per.window))
```


```{r}
ont.wt.raw.counts.per.window = lapply(define.windows, function(x){
  
  x <- unlist(define.windows[1])
  meta.temp = m.meta[c(x[1]:x[2]),]
  shared.cells <- intersect(rownames(meta.temp), rownames(ch.ont.samples.integrated@meta.data))
  ch.ont.samples.integrated.meta.sub <- ch.ont.samples.integrated@meta.data[shared.cells,]
  wt <- rownames(ch.ont.samples.integrated.meta.sub[which(ch.ont.samples.integrated.meta.sub$Genotype_1WT =="WT"),])
  message(paste(length(wt),"WT cells"))
  message(x[1],":", x[2])
  bulk.counts <- as.data.frame(rowSums(as.data.frame(ch.ont.samples.integrated@assays$RNA@counts[gene.list,wt])))
  colnames(bulk.counts) <- c("counts")
  bulk.counts
  out = bulk.counts
  return(out)
  
})

ont.wt.raw.counts.per.window <- as.data.frame(ont.wt.raw.counts.per.window)
colnames(ont.wt.raw.counts.per.window) <- as.character(1:ncol(ont.wt.raw.counts.per.window))
```


```{r}
ont.total.raw.counts.per.window = lapply(define.windows, function(x){
  #x <- define.windows[[1]]
  meta.temp = m.meta[c(x[1]:x[2]),]
  shared.cells <- intersect(rownames(meta.temp), rownames(ch.ont.samples.integrated@meta.data))
  message(paste(length(shared.cells),"total cells"))
  message(x[1],":", x[2])
  bulk.counts <- as.data.frame(colSums(as.data.frame(ch.ont.samples.integrated@assays$RNA@counts[,shared.cells])))
  bulk.count <- sum(bulk.counts[,1])
  out = bulk.count
  return(out)
})

ont.total.raw.counts.per.window <- unlist(ont.total.raw.counts.per.window)
```

```{r}
ont.raw.counts.per.window
ont.mut.raw.counts.per.window
ont.wt.raw.counts.per.window
```



```{r}
##------------------------------- Get ONT.expression per window for each gene involved ------------------------------- ##

gene.list <- paste(unique(intron.meta.sub$gene))

ont.expression.per.window = lapply(define.windows, function(x){
  
  x <- unlist(define.windows[10])
  meta.temp = m.meta[c(x[1]:x[2]),]
  shared.cells <- intersect(rownames(meta.temp), rownames(ch.ont.samples.integrated@meta.data))
  length(shared.cells)
  nrow(meta.temp)
  wt <- rownames(ch.ont.samples.integrated@meta.data[which(ch.ont.samples.integrated@meta.data[shared.cells,"Genotype_1WT"] =="WT"),])
  mut <- rownames(ch.ont.samples.integrated@meta.data[which(ch.ont.samples.integrated@meta.data[shared.cells,"Genotype_1WT"] =="MUT"),])

  message(paste(length(wt),"WT cells"))
  message(paste(length(mut),"MUT cells"))
  message(paste(length(shared.cells),"total cells"))
  
  message(x[1],":", x[2])

  ch.ont.samples.integrated@meta.data$window <- 0
  ch.ont.samples.integrated@meta.data[shared.cells,]$window <- 1
  #table(ch.ont.samples.integrated@meta.data$window)
  ch.ont.samples.integrated@meta.data[which(ch.ont.samples.integrated@meta.data$window ==1 & ch.ont.samples.integrated@meta.data$nCount_RNA < 1000),]$window <- 0
  #table(ch.ont.samples.integrated@meta.data$window)
  
  Idents(ch.ont.samples.integrated) <- "window"
  #expression.values <- AverageExpression(ch.ont.samples.integrated, assay = "RNA" ,features = gene.list, slot="data")
  #expression.values <- AverageExpression(ch.ont.samples.integrated, assay = "RNA" ,features = gene.list, slot="counts")
  expression.values <- AverageExpression(ch.ont.samples.integrated, features = gene.list, slot="data")
  #expression.values$RNA
  out = expression.values$integrated[,"1"]
  #out = expression.values$RNA[,"1"]
  return(out)
  
  })
  
ont.expression.per.window <- as.data.frame(ont.expression.per.window)
rownames(ont.expression.per.window) <- gene.list
colnames(ont.expression.per.window) <- as.character(1:ncol(ont.expression.per.window))

ont.expression.per.window.RNA 
ont.expression.per.window
```

```{r}
as.data.frame(ch.ont.samples.integrated@assays$RNA@data)[gene.list,]
```


```{r}
ont.mut.expression.per.window = lapply(define.windows, function(x){
  
  #x <- unlist(define.windows[1])
  meta.temp = m.meta[c(x[1]:x[2]),]
  shared.cells <- intersect(rownames(meta.temp), rownames(ch.ont.samples.integrated@meta.data))
  length(shared.cells)
  mut <- rownames(ch.ont.samples.integrated@meta.data[which(ch.ont.samples.integrated@meta.data[shared.cells,"Genotype_1WT"] =="MUT"),])
  message(paste(length(mut),"MUT cells"))
  message(x[1],":", x[2])

  ch.ont.samples.integrated@meta.data$window <- 0
  ch.ont.samples.integrated@meta.data[mut,]$window <- 1

  Idents(ch.ont.samples.integrated) <- "window"
  #expression.values <- AverageExpression(ch.ont.samples.integrated, assay = "RNA" ,features = gene.list, slot="data")
  expression.values <- AverageExpression(ch.ont.samples.integrated, assay = "RNA" ,features = gene.list, slot="counts")
  expression.values$RNA
  out = expression.values$RNA[,"1"]
  return(out)
  
  })
  
ont.mut.expression.per.window <- as.data.frame(ont.mut.expression.per.window)
rownames(ont.mut.expression.per.window) <- gene.list
colnames(ont.mut.expression.per.window) <- as.character(1:ncol(ont.mut.expression.per.window))

```



```{r}
ont.wt.expression.per.window = lapply(define.windows, function(x){
  
  #x <- unlist(define.windows[1])
  meta.temp = m.meta[c(x[1]:x[2]),]
  shared.cells <- intersect(rownames(meta.temp), rownames(ch.ont.samples.integrated@meta.data))
  length(shared.cells)
  wt <- rownames(ch.ont.samples.integrated@meta.data[which(ch.ont.samples.integrated@meta.data[shared.cells,"Genotype_1WT"] =="WT"),])
  message(paste(length(wt),"WT cells"))
  message(x[1],":", x[2])

  ch.ont.samples.integrated@meta.data$window <- 0
  ch.ont.samples.integrated@meta.data[wt,]$window <- 1

  Idents(ch.ont.samples.integrated) <- "window"
  #expression.values <- AverageExpression(ch.ont.samples.integrated, assay = "RNA" ,features = gene.list, slot="data")
  expression.values <- AverageExpression(ch.ont.samples.integrated, assay = "RNA" ,features = gene.list, slot="counts")
  expression.values$RNA
  out = expression.values$RNA[,"1"]
  return(out)
  
  })
  
ont.wt.expression.per.window <- as.data.frame(ont.wt.expression.per.window)
rownames(ont.wt.expression.per.window) <- gene.list
colnames(ont.wt.expression.per.window) <- as.character(1:ncol(ont.wt.expression.per.window))
```


```{r}
ont.expression.per.window
ont.mut.expression.per.window
ont.wt.expression.per.window
```


```{r}
as.numeric(paste(ct.fractions["EP",]))

annotation_col.new = data.frame(ep.fraction = as.numeric(paste(ct.fractions["EP",])), mep.fraction = as.numeric(paste(ct.fractions["MEP",])), imp.fraction = as.numeric(paste(ct.fractions["IMP",])) ,hsc.fraction = as.numeric(paste(ct.fractions["HSPC",])),  mean.pseudotime = pseudotime.mean)
```


```{r}

annotation_col = data.frame(ep.fraction = ct.fractions["EP"], mep.fraction = ct.fractions["MEP"] , imp.fraction = ct.fractions["IMP"] ,hsc.fraction = ct.fractions["HSPC"],  mean.pseudotime = pseudotime.mean)

#annotation_col = data.frame(ep.fraction= ct.fractions["EP"], mep.fraction= ct.fractions["MEP"] , imp.fraction = ct.fractions["IMP"] ,hsc.fraction = ct.fractions["HSPC"],  mean.pseudotime = pseudotime.mean, mut.frac= mut.fraction)

rownames(annotation_col) = paste(1:nrow(annotation_col))

ann_colors = list(
  ep.fraction = c("grey","blue"),
  mep.fraction = c("grey","green"),
  imp.fraction = c("grey","yellow"),
  hsc.fraction = c("grey","red"),
  mean.pseudotime = c("grey", "purple")
  #mut.frac = c("grey", "green")
)


to.plot

pheatmap(to.plot[complete.cases(to.plot),], annotation = annotation_col, annotation_colors = ann_colors, cluster_cols = F, cluster_rows = T, clustering_distance_rows="correlation",  color=colorRampPalette(c("navy", "white", "red"))(50))
```


##Loking at list of Junctions from Gene Cards
```{r fig.height=8, fig.width=10}

shared.junctions <- intersect(junctions.of.interest.final, rownames(out))

to.plot <- out.temp[shared.junctions,]
pheatmap(to.plot[complete.cases(to.plot),], cluster_cols = F, cluster_rows = F, annotation_col = annotation_col, annotation_colors = ann_colors)
pheatmap(to.plot[complete.cases(to.plot),], cluster_cols = F, cluster_rows = F, annotation_col = annotation_col, annotation_colors = ann_colors,color=colorRampPalette(c("navy", "white", "red"))(50))
#pheatmap(to.plot[complete.cases(to.plot),], cluster_cols = F, cluster_rows = T, annotation_col = annotation_col, annotation_colors = ann_colors, color=colorRampPalette(c("navy", "white", "red"))(50), breaks = seq(-100, +100, length.out = 50))

to.plot <- out.zscores[shared.junctions,]
pheatmap(to.plot[complete.cases(to.plot),], cluster_cols = F, cluster_rows = F, annotation_col = annotation_col, annotation_colors = ann_colors, color=colorRampPalette(c("navy", "white", "red"))(50))
```

##HSPC and EP  specific splicing events 


```{r}
annotation_row = data.frame(cryptic.junction=junctions.of.interest, cryptic.junctions.full=junctions.of.interest.final)

rownames(annotation_row) <- annotation_row$cryptic.junction
annotation_row$cell.type.site <- NA
annotation_row[junctions.of.interest.ep,]$cell.type.site <- "EP"
annotation_row[junctions.of.interest.hspc,]$cell.type.site <- "HSPC"

annotation_row

rownames(annotation_row) <- annotation_row$cryptic.junctions.full
annotation_row <- annotation_row[3]

annotation_row
to.plot
```


```{r fig.height=8, fig.width=10}


shared.junctions <- intersect(junctions.of.interest.final, rownames(out))

to.plot <- out.temp[shared.junctions,]
#pheatmap(to.plot[complete.cases(to.plot),], cluster_cols = F, cluster_rows = T, annotation_col = annotation_col, annotation_colors = ann_colors)
#pheatmap(to.plot[complete.cases(to.plot),], cluster_cols = F, cluster_rows = T, annotation_col = annotation_col, annotation_colors = ann_colors,color=colorRampPalette(c("navy", "white", "red"))(50))

pheatmap(to.plot[complete.cases(to.plot),], cluster_cols = F, cluster_rows = T, annotation_row = annotation_row , annotation_col = annotation_col, annotation_colors = ann_colors,color=colorRampPalette(c("navy", "white", "red"))(50))
pheatmap(to.plot[complete.cases(to.plot),], cluster_cols = F, cluster_rows = F, annotation_row = annotation_row , annotation_col = annotation_col, annotation_colors = ann_colors,color=colorRampPalette(c("navy", "white", "red"))(50))

to.plot <- out.zscores[shared.junctions,]
#pheatmap(to.plot[complete.cases(to.plot),], cluster_cols = F, cluster_rows = T,clustering_distance_rows="correlation", annotation_col = annotation_col, annotation_colors = ann_colors, color=colorRampPalette(c("navy", "white", "red"))(50))
pheatmap(to.plot[complete.cases(to.plot),], cluster_cols = F, cluster_rows = T, annotation_row = annotation_row, annotation_col = annotation_col, annotation_colors = ann_colors, color=colorRampPalette(c("navy", "white", "red"))(50))
pheatmap(to.plot[complete.cases(to.plot),], cluster_cols = F, cluster_rows = F, annotation_row = annotation_row, annotation_col = annotation_col, annotation_colors = ann_colors, color=colorRampPalette(c("navy", "white", "red"))(50))
```




##Loking at Ribosomal Genes and Ribosomal Poisioning 
```{r fig.height=8, fig.width=10}
shared.junctions <- intersect(junctions.of.interest.final, rownames(out))

to.plot <- out.temp[shared.junctions,]
pheatmap(to.plot[complete.cases(to.plot),], cluster_cols = F, cluster_rows = T, annotation_col = annotation_col, annotation_colors = ann_colors)
pheatmap(to.plot[complete.cases(to.plot),], cluster_cols = F, cluster_rows = T, annotation_col = annotation_col, annotation_colors = ann_colors,color=colorRampPalette(c("navy", "white", "red"))(50))

to.plot <- out.zscores[shared.junctions,]
pheatmap(to.plot[complete.cases(to.plot),], cluster_cols = F, cluster_rows = T,clustering_distance_rows="correlation", annotation_col = annotation_col, annotation_colors = ann_colors, color=colorRampPalette(c("navy", "white", "red"))(50))

pheatmap(to.plot[complete.cases(to.plot),], cluster_cols = F, cluster_rows = F, annotation_col = annotation_col, annotation_colors = ann_colors, color=colorRampPalette(c("navy", "white", "red"))(50))
```



```{r fig.height=8, fig.width=10}
shared.junctions <- intersect(junctions.of.interest.final, rownames(out))

to.plot <- out.temp[shared.junctions,]
pheatmap(to.plot[complete.cases(to.plot),], cluster_cols = F, cluster_rows = T, annotation_col = annotation_col, annotation_colors = ann_colors)

to.plot <- out.zscores[shared.junctions,]
pheatmap(to.plot[complete.cases(to.plot),], cluster_cols = F, cluster_rows = T,clustering_distance_rows="correlation", annotation_col = annotation_col, annotation_colors = ann_colors, color=colorRampPalette(c("navy", "white", "red"))(50))

pheatmap(to.plot[complete.cases(to.plot),], cluster_cols = F, cluster_rows = F, annotation_col = annotation_col, annotation_colors = ann_colors, color=colorRampPalette(c("navy", "white", "red"))(50))
```


##Generate Individual Scatter Plots 

```{r}
out
```


```{r}
load("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/16.Pseudotime.Window/CH.output/ch.259.info.Rdata")
#pseudotime.mean, ct.fractions, ct.fractions.2, mut.fraction, dPSI.per.window, ont.expression.per.window

```

```{r}
dPSI.per.window
```


```{r}
#junction <- "chrX:48793297:48793777:+" 
#gene <- "GATA1"

junction <- "chr17:44322552:44322807:-" 
gene <- "SLC25A39"

#Not well covered
#junction <- "chr14:22768171:22769776:+" 
#gene <- "OXA1L"

#Well covered
junction <- "chr3:186783639:186784428:+" 
gene <- "EIF4A2"

#junction <- "chr10:119582608:119588151:-"
#gene <- "TIAL1"

#junction <- "chr4:139072039:139073453:-"
#gene <- "ELF2"

#junction <- "chr20:35556271:35556954:+" 
#gene <- "ERGIC3"
#junction <- "chr9:136791424:136791625:+" 
#gene <- "TMEM141"
junction <- "chr1:100992754:100995109:-" 
gene <- "DPH5"

junction <- "chr3:170868201:170868297:-" 
gene <- "RPL22L1"

junction <- "chr11:64232614:64232700:+" 
gene <- "DNAJC4"

junction <- "chr1:45012285:45012856:+" 
gene <- "UROD"


junction.final <- paste0(gene,":",junction)

#dPSI info - out[junction.final,]
#ONT gene.expression info - gene.exp.matrix[gene,]

colnames <- c("window","dPSI","mut.psi","wt.psi","ONT.Expression","ONT.MUT.Expression","ONT.WT.Expression")
psi.per.window.data <- as.data.frame(matrix(nrow=ncol(dPSI.per.window), ncol=length(colnames)))
rownames(psi.per.window.data) <- as.character(1:ncol(dPSI.per.window))
colnames(psi.per.window.data) <- colnames

psi.per.window.data$window <- rownames(psi.per.window.data)
psi.per.window.data$dPSI <- as.numeric(paste(dPSI.per.window[junction.final,]))
psi.per.window.data$mut.psi <- as.numeric(paste(mut.psi.per.window[junction.final,]))
psi.per.window.data$wt.psi <- as.numeric(paste(wt.psi.per.window[junction.final,]))
psi.per.window.data$ONT.Expression <- as.numeric(paste(ont.expression.per.window.RNA[gene,]))
psi.per.window.data$ONT.MUT.Expression <- as.numeric(paste(ont.mut.expression.per.window[gene,]))
psi.per.window.data$ONT.WT.Expression <- as.numeric(paste(ont.wt.expression.per.window[gene,]))
psi.per.window.data$window <- factor(psi.per.window.data$window,levels = 1:nrow(psi.per.window.data))
psi.per.window.data
mask <- apply(psi.per.window.data, 2, is.na)
psi.per.window.data[mask] <- 0
psi.per.window.data
```

```{r}
dPSI.per.window
psi.per.window.data
ont.expression.per.window[gene,]
```

```{r}
ExpressionColor <- "red"
dPSIColor <- "blue"
```


```{r}
OldMax = max(psi.per.window.data$ONT.Expression)
OldMin = min(psi.per.window.data$ONT.Expression)
OldRange = (OldMax - OldMin)


NewMax = max(psi.per.window.data$dPSI)
NewMin = min(psi.per.window.data$dPSI)
NewRange = (NewMax - NewMin)

ggplot(psi.per.window.data, aes(x=window, group = 1)) + 
  geom_line( aes(y=dPSI), color=dPSIColor) + 
  geom_line( aes(y=(((ONT.Expression - OldMin) * NewRange) / OldRange) + NewMin), linetype = "dashed", color=ExpressionColor)+  
  geom_hline(yintercept = 0, linetype = "dotted") + 
  scale_y_continuous(
    name = "dPSI",
    sec.axis = sec_axis(~(((.-NewMin)*OldRange)/NewRange) + OldMin , name="ONT Expression \n (log-normalized counts)")) +
  theme_classic()+
  theme(plot.title = element_text(hjust = 0.5, size = 14), 
        axis.title.y = element_text(color =dPSIColor , size=13), 
        axis.title.y.right = element_text(color = ExpressionColor, size=13)) +
  ggtitle(paste0("dPSI and Expression for ", gene ," cryptic junction - ", junction)) 
```


```{r}
log1p()
log1p()
```


```{r}
OldMax = max(psi.per.window.data$ONT.Expression)
OldMin = min(psi.per.window.data$ONT.Expression)
OldRange = (OldMax - OldMin)


NewMax = max(psi.per.window.data$dPSI)
NewMin = min(psi.per.window.data$dPSI)
NewRange = (NewMax - NewMin)

ggplot(psi.per.window.data, aes(x=window, group = 1)) + 
  geom_line( aes(y=dPSI), color=dPSIColor) + 
  geom_line( aes(y=(((ONT.Expression - OldMin) * NewRange) / OldRange) + NewMin), linetype = "dashed", color=ExpressionColor)+  
  geom_hline(yintercept = 0, linetype = "dotted") + 
  scale_y_continuous(
    name = "dPSI",
    sec.axis = sec_axis(~(((.-NewMin)*OldRange)/NewRange) + OldMin , name="ONT Expression \n (log-normalized counts)")) +
  theme_classic()+
  theme(plot.title = element_text(hjust = 0.5, size = 14), 
        axis.title.y = element_text(color =dPSIColor , size=13), 
        axis.title.y.right = element_text(color = ExpressionColor, size=13)) +
  ggtitle(paste0("dPSI and Expression for ", gene ," cryptic junction - ", junction)) 
```


```{r fig.height=5, fig.width=8}

OldMax = max(psi.per.window.data$ONT.Expression)
OldMin = min(psi.per.window.data$ONT.Expression)
OldRange = (OldMax - OldMin)


NewMax = max(psi.per.window.data$dPSI)
NewMin = min(psi.per.window.data$dPSI)
NewRange = (NewMax - NewMin)

ggplot(psi.per.window.data, aes(x=window, group = 1)) + 
  geom_line( aes(y=dPSI), color=dPSIColor) + 
  geom_line( aes(y=(((ONT.Expression - OldMin) * NewRange) / OldRange) + NewMin), linetype = "dashed", color=ExpressionColor)+  
  geom_hline(yintercept = 0, linetype = "dotted") + 
  scale_y_continuous(
    name = "dPSI",
    sec.axis = sec_axis(~(((.-NewMin)*OldRange)/NewRange) + OldMin , name="ONT Expression \n (log-normalized counts)")) +
  theme_classic()+
  theme(plot.title = element_text(hjust = 0.5, size = 14), 
        axis.title.y = element_text(color =dPSIColor , size=13), 
        axis.title.y.right = element_text(color = ExpressionColor, size=13)) +
  ggtitle(paste0("dPSI and Expression for ", gene ," cryptic junction - ", junction)) 



ggplot(psi.per.window.data, aes(x=window, group = 1)) + 
  geom_point(aes(y= dPSI), color=dPSIColor) +
  stat_smooth(aes(y= dPSI), color=dPSIColor, method = "loess", formula = y ~ x, se = F, span=0.5) +
  geom_point(aes(y =(((ONT.Expression - OldMin) * NewRange) / OldRange) + NewMin), color=ExpressionColor) +
  stat_smooth(aes(y =(((ONT.Expression - OldMin) * NewRange) / OldRange) + NewMin), linetype = "dashed", color=ExpressionColor, method = "loess", formula = y ~ x, se = F, span=0.5)+
  geom_hline(yintercept = 0, linetype = "dotted") + 
  scale_y_continuous(
    name = "dPSI",
    sec.axis = sec_axis(~(((.-NewMin)*OldRange)/NewRange) + OldMin , name="ONT Expression \n (log-normalized counts)")) +
  theme_classic()+
  theme(plot.title = element_text(hjust = 0.5, size = 14), 
        axis.title.y = element_text(color =dPSIColor , size=13), 
        axis.title.y.right = element_text(color = ExpressionColor, size=13)) +
  ggtitle(paste0("dPSI and Expression for ", gene ," cryptic junction - ", junction)) 


```




###Plotting the MUT and WT expression separatly but still with dPSI 


```{r fig.height=5, fig.width=9}

OldMax = max(psi.per.window.data$ONT.MUT.Expression, psi.per.window.data$ONT.WT.Expression)
OldMin = min(psi.per.window.data$ONT.MUT.Expression, psi.per.window.data$ONT.WT.Expression)
OldRange = (OldMax - OldMin)

NewMax = max(psi.per.window.data$dPSI)
NewMin = min(psi.per.window.data$dPSI)
NewRange = (NewMax - NewMin)

ggplot(psi.per.window.data, aes(x=window, group = 1)) + 
  geom_line(aes(y=dPSI), color="blue") + 
  geom_hline(yintercept = 0, linetype = "dotted") + 
  geom_line(aes(y=(((ONT.WT.Expression - OldMin) * NewRange) / OldRange) + NewMin, linetype = "WT.Exp"),color="red") +
  geom_line(aes(y=(((ONT.MUT.Expression - OldMin) * NewRange) / OldRange) + NewMin, linetype = "MUT.Exp"),color="red") +
  scale_linetype_manual(name = "Expression Type",
                     breaks = c("WT.Exp","MUT.Exp"),
                     values = c("WT.Exp" = "solid", "MUT.Exp" = "dashed") )+
  scale_y_continuous(
    name = "dPSI",
    sec.axis = sec_axis(~(((.-NewMin)*OldRange)/NewRange) + OldMin , name="ONT Expression \n (log-normalized counts)")) +
  theme_classic() +
  theme(plot.title = element_text(hjust = 0.5, size = 14), 
        axis.title.y = element_text(color =dPSIColor , size=13), 
        axis.title.y.right = element_text(color = ExpressionColor, size=13))+
  ggtitle(paste0("dPSI and Expression for ", gene ," cryptic junction - ", junction)) 
```

##Doing one plot with mut exp + mut psi & wt exp + wt psi

```{r fig.height=5, fig.width=9}

OldMax = max(psi.per.window.data$ONT.MUT.Expression, psi.per.window.data$ONT.WT.Expression)
OldMin = min(psi.per.window.data$ONT.MUT.Expression, psi.per.window.data$ONT.WT.Expression)
OldRange = (OldMax - OldMin)

NewMax = max(psi.per.window.data$mut.psi, psi.per.window.data$wt.psi)
NewMin = min(psi.per.window.data$mut.psi, psi.per.window.data$wt.psi)
NewRange = (NewMax - NewMin)

ggplot(psi.per.window.data, aes(x=window, group = 1)) + 
  geom_line(aes(y=wt.psi, linetype = "WT"), color="blue") + 
  geom_line(aes(y=mut.psi,linetype = "MUT"), color="blue") + 
  #geom_hline(yintercept = 0, linetype = "dotted") + 
  geom_line(aes(y=(((ONT.WT.Expression - OldMin) * NewRange) / OldRange) + NewMin, linetype = "WT"),color="red") +
  geom_line(aes(y=(((ONT.MUT.Expression - OldMin) * NewRange) / OldRange) + NewMin, linetype = "MUT"),color="red") +
  scale_linetype_manual(name = "Genotype",
                     breaks = c("WT","MUT"),
                     values = c("WT" = "solid", "MUT" = "dashed") )+
  scale_y_continuous(
    name = "PSI",
    sec.axis = sec_axis(~(((.-NewMin)*OldRange)/NewRange) + OldMin , name="ONT Expression \n (log-normalized counts)")) +
  theme_classic() +
  theme(plot.title = element_text(hjust = 0.5, size = 14), 
        axis.title.y = element_text(color =dPSIColor , size=13), 
        axis.title.y.right = element_text(color = ExpressionColor, size=13))+
  ggtitle(paste0("PSI and Expression values for ", gene ," cryptic junction - ", junction)) 
```

##Doing individual Plots for mut exp + mut psi and wt exp + wt psi

```{r fig.height=5, fig.width=8}
ExpressionColor <- "red"
PSIColor <- "blue"

#MUT
OldMax = max(psi.per.window.data$ONT.MUT.Expression)
OldMin = min(psi.per.window.data$ONT.MUT.Expression)
OldRange = (OldMax - OldMin)

NewMax = max(psi.per.window.data$mut.psi)
NewMin = min(psi.per.window.data$mut.psi)
NewRange = (NewMax - NewMin)

ggplot(psi.per.window.data, aes(x=window, group = 1)) + 
  geom_line(aes(y=mut.psi), color=PSIColor) + 
  geom_line(aes(y=(((ONT.MUT.Expression - OldMin) * NewRange) / OldRange) + NewMin),linetype = "dashed",color=ExpressionColor) +
  scale_y_continuous(
    name = "mut.PSI",
    sec.axis = sec_axis(~(((.-NewMin)*OldRange)/NewRange) + OldMin , name="MUT.ONT Expression \n (log-normalized counts)")) +
  theme_classic() +
  theme(plot.title = element_text(hjust = 0.5, size = 14), 
        axis.title.y = element_text(color =PSIColor , size=13), 
        axis.title.y.right = element_text(color = ExpressionColor, size=13))+
  ggtitle(paste0("mut.psi and mut.exp values for ", gene ," cryptic junction - ", junction)) 


#WT
OldMax = max(psi.per.window.data$ONT.WT.Expression)
OldMin = min(psi.per.window.data$ONT.WT.Expression)
OldRange = (OldMax - OldMin)

NewMax = max(psi.per.window.data$wt.psi)
NewMin = min(psi.per.window.data$wt.psi)
NewRange = (NewMax - NewMin)


ggplot(psi.per.window.data, aes(x=window, group = 1)) + 
  geom_line(aes(y=wt.psi), color=PSIColor) + 
  geom_line(aes(y=(((ONT.WT.Expression - OldMin) * NewRange) / OldRange) + NewMin),linetype = "dashed",color=ExpressionColor) +
  scale_y_continuous(
    name = "wt.PSI",
    sec.axis = sec_axis(~(((.-NewMin)*OldRange)/NewRange) + OldMin , name="WT.ONT Expression \n (normalized counts)")) +
  theme_classic() +
  theme(plot.title = element_text(hjust = 0.5, size = 14), 
        axis.title.y = element_text(color =PSIColor , size=13), 
        axis.title.y.right = element_text(color = ExpressionColor, size=13))+
  ggtitle(paste0("wt.psi and wt.exp values for ", gene ," cryptic junction - ", junction)) 
```


##Get heatmap to show the cluster coverage



```{r}
win.dPSI[[1]]
```

```{r fig.height=1, fig.width=8}

extracted_cols <- lapply(win.dPSI, function(x) x[,"mut.bulk.counts"]) 
mut.counts.per.window <- as.data.frame(do.call(cbind, extracted_cols))
colnames(mut.counts.per.window) = as.character(1:ncol(mut.counts.per.window))
rownames(mut.counts.per.window) <- rownames

extracted_cols <- lapply(win.dPSI, function(x) x[,"cluster.cov.mut"]) #note the added comma!
cluster.cov.mut.per.window <- as.data.frame(do.call(cbind, extracted_cols))
colnames(cluster.cov.mut.per.window) = as.character(1:ncol(cluster.cov.mut.per.window))
rownames(cluster.cov.mut.per.window) <- rownames

extracted_cols <- lapply(win.dPSI, function(x) x[,"wt.bulk.counts"]) 
wt.counts.per.window <- as.data.frame(do.call(cbind, extracted_cols))
colnames(wt.counts.per.window) = as.character(1:ncol(wt.counts.per.window))
rownames(wt.counts.per.window) <- rownames

extracted_cols <- lapply(win.dPSI, function(x) x[,"cluster.cov.wt"]) #note the added comma!
cluster.cov.wt.per.window <- as.data.frame(do.call(cbind, extracted_cols))
colnames(cluster.cov.wt.per.window) = as.character(1:ncol(cluster.cov.wt.per.window))
rownames(cluster.cov.wt.per.window) <- rownames


raw.counts <- rbind(mut.counts.per.window[junction.final,], cluster.cov.mut.per.window[junction.final,] , wt.counts.per.window[junction.final,] , cluster.cov.wt.per.window[junction.final,])

raw.counts <- as.data.frame(sapply(raw.counts, as.integer))
rownames(raw.counts) <- c("cryptic.reads.mut", "cluster.cov.mut","cryptic.reads.wt","cluster.cov.wt")
raw.counts

#pheatmap(raw.counts, cluster_cols = F, cluster_rows = F, legend=F, display_numbers = T, number_format= "%.f",number_color = "blue",color=colorRampPalette(c("white", "white"))(50), show_rownames = T, show_colnames = F)

pheatmap(raw.counts, cluster_cols = F, cluster_rows = F, legend=F, display_numbers = T, number_format= "%.f",number_color = "blue",color=colorRampPalette(c("white", "white"))(50), show_rownames = T, show_colnames = F, gaps_row = c(2))
```

##Looking at the raw counts from the expression Data
```{r}
ont.raw.counts.per.window
ont.mut.raw.counts.per.window
ont.wt.raw.counts.per.window
```


```{r}
raw.counts["cryptic.reads.mut", "cluster.cov.mut",]
```

```{r fig.height=1, fig.width=8}
pheatmap(raw.counts[c("cryptic.reads.mut", "cluster.cov.mut"),], cluster_cols = F, cluster_rows = F, legend=F, display_numbers = T, number_format= "%.f",number_color = "blue",color=colorRampPalette(c("white", "white"))(50), show_rownames = T, show_colnames = F)
pheatmap(raw.counts[c("cryptic.reads.wt", "cluster.cov.wt"),], cluster_cols = F, cluster_rows = F, legend=F, display_numbers = T, number_format= "%.f",number_color = "blue",color=colorRampPalette(c("white", "white"))(50), show_rownames = T, show_colnames = F )
```


```{r fig.height=0.5, fig.width=8}
raw.exp.ont.counts <- rbind(ont.raw.counts.per.window[gene,],ont.mut.raw.counts.per.window[gene,],ont.wt.raw.counts.per.window[gene,])
rownames(raw.exp.ont.counts) <- c("raw.count.mut.and.wt","raw.count.mut","raw.count.wt")
raw.exp.ont.counts


pheatmap(raw.exp.ont.counts[c("raw.count.mut.and.wt"),], cluster_cols = F, cluster_rows = F, legend=F, display_numbers = T, number_format= "%.f",number_color = "red", color=colorRampPalette(c("white", "white"))(50), show_rownames = T, show_colnames = F)
```


```{r fig.height=1, fig.width=8}
pheatmap(raw.exp.ont.counts[c("raw.count.mut", "raw.count.wt"),], cluster_cols = F, cluster_rows = F, legend=F, display_numbers = T, number_format= "%.f",number_color = "red",color=colorRampPalette(c("white", "white"))(50), show_rownames = T, show_colnames = F, gaps_row = c(1))

```







```{r}
ggplot(psi.per.window.data, aes(x=window, group = 1)) + 
  geom_line( aes(y=dPSI), color=dPSIColor) + 
  geom_line( aes(y=(((ONT.Expression - OldMin) * NewRange) / OldRange) + NewMin), linetype = "dashed", color=ExpressionColor)+
  geom_hline(yintercept = 0, linetype = "dotted") + 
  scale_y_continuous(
    name = "dPSI",
    sec.axis = sec_axis(~(((.-NewMin)*OldRange)/NewRange) + OldMin , name="ONT Expression \n (log-normalized counts)")) +
  theme_classic()+
  theme(plot.title = element_text(hjust = 0.5, size = 14), 
        axis.title.y = element_text(color =dPSIColor , size=13), 
        axis.title.y.right = element_text(color = ExpressionColor, size=13)) +
  ggtitle(paste0("dPSI and Expression for ", gene ," cryptic junction - ", junction))
```


```{r}
FeaturePlot(ch.samples.integrated, features="ERGIC3", reduction = "umap")
RidgePlot(ch.samples.integrated, features="ERGIC3", group.by = "CellType")
RidgePlot(ch.ont.samples.integrated, features="ERGIC3", group.by = "Cell.Assignment", assay= "RNA")
```


##Use fig.height=8, fig.width=12 so that the buckets are spread out enough

```{r fig.height=5, fig.width=8}

ggplot(psi.per.window.data, aes(x=window, group = 1)) + 
  geom_line( aes(y=dPSI), color=dPSIColor) + 
  stat_smooth(aes(y= dPSI), method = "loess", formula = y ~ x, se = T) +
  geom_line( aes(y=(((ONT.Expression - OldMin) * NewRange) / OldRange) + NewMin), linetype = "dashed", color=ExpressionColor)+  
  stat_smooth(aes(y =(((ONT.Expression - OldMin) * NewRange) / OldRange) + NewMin), linetype = "dashed", color=ExpressionColor, method = "loess", formula = y ~ x, se = T)+
  scale_y_continuous(
    name = "dPSI",
    sec.axis = sec_axis(~(((.-NewMin)*OldRange)/NewRange) + OldMin , name="ONT Expression \n (log-normalized counts)")) +
  theme_classic()+
  theme(plot.title = element_text(hjust = 0.5, size = 14), 
        axis.title.y = element_text(color =dPSIColor , size=13), 
        axis.title.y.right = element_text(color = ExpressionColor, size=13)) +
  ggtitle(paste0("dPSI and Expression for ", gene ," cryptic junction - ", junction)) 


ggplot(psi.per.window.data, aes(x=window, group = 1)) + 
  #geom_line( aes(y=dPSI), color=dPSIColor) + 
  stat_smooth(aes(y= dPSI), method = "loess", formula = y ~ x, se = T) +
  #geom_line( aes(y=(((ONT.Expression - OldMin) * NewRange) / OldRange) + NewMin), linetype = "dashed", color=ExpressionColor)+  
  stat_smooth(aes(y =(((ONT.Expression - OldMin) * NewRange) / OldRange) + NewMin), linetype = "dashed", color=ExpressionColor, method = "loess", formula = y ~ x, se = T)+
  geom_hline(yintercept = 0, linetype = "dotted") + 
  scale_y_continuous(
    name = "dPSI",
    sec.axis = sec_axis(~(((.-NewMin)*OldRange)/NewRange) + OldMin , name="ONT Expression \n (log-normalized counts)")) +
  theme_classic()+
  theme(plot.title = element_text(hjust = 0.5, size = 14), 
        axis.title.y = element_text(color =dPSIColor , size=13), 
        axis.title.y.right = element_text(color = ExpressionColor, size=13)) +
  ggtitle(paste0("dPSI and Expression for ", gene ," cryptic junction - ", junction)) 


##Make heatmap for changed in cell type fractions per bucket 

```




```{r fig.height=2, fig.width=12}
library(pheatmap)
ct.fractions
pheatmap(ct.fractions, cluster_cols = F, cluster_rows = F, legend=F)
```


```{r}

pheatmap(ct.fractions.2, cluster_cols = F, cluster_rows = F, legend=F)

```










##Grab Normalized Expression Values from Seurat object - ONT per window (make a function for this)

```{r}
#Load CH Seurat Object:
#ch.samples.integrated@assays$RNA@data[1:20,1:20]
#Idents(ch.samples.integrated)[1:5]

#Make CellType the identities and Get average expression for gene of choice using AverageExpression Function
#The data slot of the seruat object has Log-Normalized Values for expression

Idents(mds.ont.samples.integrated) <- "Cell.Assignment"
gene <-"GATA1"
expression.values.ct <- AverageExpression(mds.ont.samples.integrated, assay = "RNA" ,features = gene, slot="data")
expression.values.ct
expression.values.ct$RNA <- as.data.frame(t(expression.values.ct$RNA))

mds.ont.samples.integrated@meta.data$All <- "ALL"
Idents(mds.ont.samples.integrated) <- "All"
expression.values.all <- AverageExpression(mds.ont.samples.integrated, assay = "RNA" ,features = gene, slot="data")
expression.values.all$RNA <- as.data.frame(t(expression.values.all$RNA))

expression.values <- rbind(expression.values.all$RNA, expression.values.ct$RNA)
psi.data$ONT.Expression <- NA
psi.data$ONT.Expression <- expression.values[rownames(psi.data),gene]

```


```{r}
trial <- FetchData(ch.ont.samples.integrated, "GATA1", slot = "data")

ch.ont.samples.integrated@meta.data$GATA1exp <- trial$GATA1


Idents(ch.ont.samples.integrated) <- "Cell.Assignment"
gene <-"GATA1"
expression.values.ct <- AverageExpression(ch.ont.samples.integrated, assay = "RNA" ,features = gene, slot="data")
expression.values.ct
expression.values.ct$RNA <- as.data.frame(t(expression.values.ct$RNA))

ggplot(ch.ont.samples.integrated@meta.data[which(ch.ont.samples.integrated@meta.data$Genotype_2WT == "MUT" | ch.ont.samples.integrated@meta.data$Genotype_2WT == "WT"),], aes(x = Cell.Assignment, y = GATA1exp, color = Genotype_2WT)) + geom_boxplot() + stat_compare_means(aes(group = Genotype_2WT), method = "wilcox.test") 
```


```{r fig.height=8, fig.width=20}
trial <- FetchData(mds.samples.integrated, "GATA1", slot = "data")

mds.samples.integrated@meta.data$GATA1exp <- trial$GATA1


ggplot(mds.samples.integrated@meta.data, aes(x = Cell.Assignment_adj, y = GATA1exp)) + geom_boxplot() 

ggplot(mds.samples.integrated@meta.data[which(mds.samples.integrated@meta.data$Genotype_2UMI == "MUT" | mds.samples.integrated@meta.data$Genotype_2UMI == "WT"),], aes(x = Cell.Assignment_adj, y = GATA1exp, color = Genotype_2UMI)) + geom_boxplot() + stat_compare_means(aes(group = Genotype_2UMI), method = "wilcox.test") 


trial <- FetchData(mds.ont.samples.integrated, "GATA1", slot = "data")

mds.ont.samples.integrated@meta.data$GATA1exp <- trial$GATA1

ggplot(mds.ont.samples.integrated@meta.data[which(mds.ont.samples.integrated@meta.data$Genotype_2WT == "MUT" | mds.ont.samples.integrated@meta.data$Genotype_2WT == "WT"),], aes(x = Cell.Assignment, y = GATA1exp, color = Genotype_2WT)) + geom_boxplot() + stat_compare_means(aes(group = Genotype_2WT), method = "wilcox.test") 
```



```{r}
mds.samples.integrated@assays$ADT
```


```{r}
##Grab the tables from the MDS object and when using it in the MDS.ONT switch the labels
```



```{r}
intron.meta <- read.table(intron.meta.file.path, sep="\t", header = T)
intron.meta$intron.junction <- paste0(intron.meta$chr,":",intron.meta$start,":",intron.meta$end,":",intron.meta$strand)
clusterIDs <- paste(intron.meta[which(intron.meta$intron.junction %in% junctions.of.interest),]$clusterID_5p)
intron.meta.sub <- intron.meta[which(intron.meta$clusterID_5p %in% clusterIDs),]
rownames(intron.meta.sub) <- intron.meta.sub$intron.junction
intron.meta.sub
junctions.sub
```


##Ordering the cell by CD71

```{r}
DefaultAssay(mds.samples.integrated)
DefaultAssay(mds.samples.integrated) <- "ADT"

cite.marker <- "ADT-CD71"
ADT_clr_expression <- FetchData(mds.samples.integrated, vars= cite.marker )

mds.samples.integrated <- AddMetaData(mds.samples.integrated, ADT_clr_expression)
cite.marker <- gsub("-", ".", cite.marker)

ord = rownames(mds.samples.integrated@meta.data)[order(mds.samples.integrated@meta.data[,cite.marker],decreasing = F)]
trial <- mds.samples.integrated@meta.data[ord,]
DefaultAssay(mds.samples.integrated) <- "RNA"

table(trial[1:10000,]$Cell.Assignment_adj)
```




