

```{r}
library(reshape2)
```
##Barplot version

```{r}
samples.integrated@meta.data[which(samples.integrated@meta.data$Genotype_1UMI_10x =="MUT"),]
```

```{r}
junctions.bulk <- read.table("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/7.Permutation_junction_output/4.Comb_patients_merge_counts/CH_combined/logOR_within_cell_type_ONLY_CRYPTIC_3P_Junctions_with_threshold_info.txt")


junctions.bulk.all <- read.table("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/7.Permutation_junction_output/4.Comb_patients_merge_counts/CH_combined/logOR_within_cell_type_ALT_3P_Junctions_with_threshold_info.txt")

junctions.bulk.all[which(junctions.bulk.all$gene == "DLST"),]

junctions.hspc.all <- read.table("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/7.Permutation_junction_output/5.Comb_patients_ind_celltype_merge_counts/CH_combined/HSPC/logOR_within_cell_type_ALT_3P_Junctions_with_threshold_info.txt")

junctions.hspc.all[which(junctions.hspc.all$gene == "CASP8"),]

junctions.hspc.all[which(junctions.hspc.all$gene == "DLST"),]
```



#Loading cell type specific results and pulling out significant events 

##For CH use the 1WT threshold ?
```{r}
##HSPC specific cryptic junctions:
junctions.hspc <- read.table("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/7.Permutation_junction_output/5.Comb_patients_ind_celltype_merge_counts/CH_combined/HSPC/logOR_within_cell_type_ONLY_CRYPTIC_3P_Junctions_with_threshold_info.txt")

##IMP specific cryptic junctions:
junctions.imp <- read.table("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/7.Permutation_junction_output/5.Comb_patients_ind_celltype_merge_counts/CH_combined/IMP/logOR_within_cell_type_ONLY_CRYPTIC_3P_Junctions_with_threshold_info.txt")

##MEP specific cryptic junctions:
junctions.mep <- read.table("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/7.Permutation_junction_output/5.Comb_patients_ind_celltype_merge_counts/CH_combined/MEP/logOR_within_cell_type_ONLY_CRYPTIC_3P_Junctions_with_threshold_info.txt")

##EP specific cryptic junctions:
junctions.ep <- read.table("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/7.Permutation_junction_output/5.Comb_patients_ind_celltype_merge_counts/CH_combined/EP/logOR_within_cell_type_ONLY_CRYPTIC_3P_Junctions_with_threshold_info.txt")

junctions.bulk <- read.table("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/7.Permutation_junction_output/4.Comb_patients_merge_counts/CH_combined/logOR_within_cell_type_ONLY_CRYPTIC_3P_Junctions_with_threshold_info.txt")

##Adding in a minimum.threshold for supportind reads from both Mut and WT cells 
dPSI.val <- 2
read.thresh <- 5
#p.val.thresh <- 0.05

junctions.hspc.sub <- junctions.hspc[which(junctions.hspc$pvalue <= 0.05 & junctions.hspc$dPSI >= dPSI.val &  junctions.hspc$three.mut.cluster.cov >= read.thresh & junctions.hspc$three.wt.cluster.cov >= read.thresh),]
rownames(junctions.hspc.sub) <- junctions.hspc.sub$intron_junction

junctions.imp.sub <- junctions.imp[which(junctions.imp$pvalue <= 0.05 & junctions.imp$dPSI >= dPSI.val & junctions.imp$three.mut.cluster.cov >= read.thresh & junctions.imp$three.wt.cluster.cov >= read.thresh),]
rownames(junctions.imp.sub) <- junctions.imp.sub$intron_junction

junctions.mep.sub <- junctions.mep[which(junctions.mep$pvalue <= 0.05 & junctions.mep$dPSI >= dPSI.val & junctions.mep$three.mut.cluster.cov >= read.thresh & junctions.mep$three.wt.cluster.cov >= read.thresh),]
rownames(junctions.mep.sub) <- junctions.mep.sub$intron_junction

junctions.ep.sub <- junctions.ep[which(junctions.ep$pvalue <= 0.05 & junctions.ep$dPSI >= dPSI.val & junctions.ep$three.mut.cluster.cov >= read.thresh & junctions.ep$three.wt.cluster.cov >= read.thresh),]
rownames(junctions.ep.sub) <- junctions.ep.sub$intron_junction


junctions.bulk.sub <- junctions.bulk[which(junctions.bulk$pvalue <= p.val.thresh & junctions.bulk$dPSI >= dPSI.val & junctions.bulk$three.mut.cluster.cov >= read.thresh & junctions.bulk$three.wt.cluster.cov >= read.thresh ),]
rownames(junctions.bulk.sub) <- junctions.bulk.sub$intron_junction

junctions.hspc.list <- paste(junctions.hspc.sub$intron_junction)
junctions.imp.list <- paste(junctions.imp.sub$intron_junction)
junctions.mep.list <- paste(junctions.mep.sub$intron_junction)
junctions.ep.list <- paste(junctions.ep.sub$intron_junction)
junctions.bulk.list <- paste(junctions.bulk.sub$intron_junction)


junctions.hspc.list.final <- paste0(junctions.hspc.sub$gene,":",junctions.hspc.sub$intron_junction)
junctions.imp.list.final <- paste0(junctions.imp.sub$gene,":",junctions.imp.sub$intron_junction)
junctions.mep.list.final <- paste0(junctions.mep.sub$gene,":",junctions.mep.sub$intron_junction)
junctions.ep.list.final <- paste0(junctions.ep.sub$gene,":",junctions.ep.sub$intron_junction)
junctions.bulk.list.final <- paste0(junctions.bulk.sub$gene,":",junctions.bulk.sub$intron_junction)

length(junctions.hspc.list)
length(junctions.imp.list)
length(junctions.mep.list)
length(junctions.ep.list)


gene <- "BAX"
gene <- "MTRF1L"
gene <- "DNAJC4"
gene <- "EIF4A2"
gene <- "SEPT2"
gene <- "DLST"
gene <- "ORAI2"
gene <- "GATA1"


junctions.hspc[which(junctions.hspc$gene == gene),]
junctions.imp[which(junctions.imp$gene == gene),]
junctions.mep[which(junctions.mep$gene == gene),]
junctions.ep[which(junctions.ep$gene == gene),]
junctions.bulk[which(junctions.bulk$gene == gene),]


##Generating junction list Junctions significant in any cell type
junctions.of.interest <- Reduce(union, list(junctions.hspc.list, junctions.imp.list, junctions.mep.list, junctions.ep.list)) 

junctions.of.interest.final <- Reduce(union, list(junctions.hspc.list.final, junctions.imp.list.final, junctions.mep.list.final, junctions.ep.list.final)) 


CH.junctions.dPSI.0.sig.5.read.thresh <- junctions.of.interest.final
CH.junctions.dPSI.2.sig.5.read.thresh <- junctions.of.interest.final

save(CH.junctions.dPSI.0.sig.5.read.thresh, CH.junctions.dPSI.2.sig.5.read.thresh, file="/gpfs/commons/groups/landau_lab/SF3B1_splice_project/17.CH.MDS.Splice.Comparison/CH.per.ct.sig.5.read.thresh.cryptic.junc.Rdata")

save(MDS.junctions.dPSI.0.sig.5.read.thresh, MDS.junctions.dPSI.2.sig.5.read.thresh, file="/gpfs/commons/groups/landau_lab/SF3B1_splice_project/17.CH.MDS.Splice.Comparison/MDS.per.ct.sig.5.read.thresh.cryptic.junc.Rdata")
```


##Saving the set of junctions to run the script on 

```{r}
gene.list <- unique(unlist(lapply(strsplit(junctions.of.interest.final,split= ":"), "[",1)))
gene.list
save(gene.list, junctions.of.interest,junctions.of.interest.final, version=2, file="/gpfs/commons/groups/landau_lab/SF3B1_splice_project/16.Pseudotime.Window/junctions.to.use.CH.dPSI2.Rdata")
```



```{r}
load("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/16.Pseudotime.Window/junctions.to.use.CH.dPSI2.Rdata")
```


```{r}
annotation_row
```

```{r}
setdiff(junctions.bulk.list.final,junctions.of.interest.final)

junctions.bulk.sub[which(junctions.bulk.sub$gene == "ORAI2"),]
```


##Creating the annotation row dataframe that also allows you to select junctions based on number of cellt ypes it is significant in.

```{r}
annotation_row[which(annotation_row$cryptic.junctions.full %in% junctions.hspc.list.final),]
```

```{r}
length(junctions.of.interest.final)
```


```{r}
annotation_row = data.frame(cryptic.junctions.full=junctions.of.interest.final)

rownames(annotation_row) <- annotation_row$cryptic.junctions.full
annotation_row$EP <- 0
annotation_row$MEP <- 0
annotation_row$IMP <- 0
annotation_row$HSPC <- 0

annotation_row[which(annotation_row$cryptic.junctions.full %in% junctions.hspc.list.final),]$HSPC <- 1
annotation_row[which(annotation_row$cryptic.junctions.full %in% junctions.imp.list.final),]$IMP <- 1
annotation_row[which(annotation_row$cryptic.junctions.full %in% junctions.mep.list.final),]$MEP <- 1
annotation_row[which(annotation_row$cryptic.junctions.full %in% junctions.ep.list.final),]$EP <- 1

#annotation_row$cell.type.site <- factor(annotation_row$cell.type.site, c("HSPC","IMP","MEP","EP"))
annotation_row <- annotation_row[-1]

annotation_row$sum <- annotation_row$EP + annotation_row$MEP + annotation_row$IMP + annotation_row$HSPC

all.examples <- rownames(annotation_row[which(annotation_row$sum > 0),])
pan.examples <- rownames(annotation_row[which(annotation_row$sum >=3),])
unique.examples <- rownames(annotation_row[which(annotation_row$sum == 1 | annotation_row$sum == 2),])

ann_colors = list(
  HSPC = c("light grey","red"),
  IMP = c("light grey","red"),
  MEP = c("light grey","red"),
  EP = c("light grey","red")
)

```


```{r}
#No expression Data for AL133410.3
setdiff(gene.list, rownames(ch.ont.samples.integrated@assays$RNA@data))
length(setdiff(gene.list, rownames(ch.ont.samples.integrated@assays$RNA@data)))
```



##Output from CH259+CH305 
```{r}
#window size of 50
#load("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/16.Pseudotime.Window/CH.output/CH259.CH305.merged.info.sig.in.any.dPSI2.bucket50.onlyMUT.Rdata")

##Use this one!!!
#load("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/16.Pseudotime.Window/CH.output/CH259.CH305.merged.info.sig.in.any.dPSI2.bucket60.onlyMUT.Rdata")

load("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/16.Pseudotime.Window/CH.output/CH259.CH305.merged.info.sig.in.any.dPSI2.bucket60.onlyMUT.wCTcounts.Rdata")

##Too few buckets - only 4
#load("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/16.Pseudotime.Window/CH.output/CH259.CH305.merged.info.sig.in.any.dPSI2.bucket75.onlyMUT.Rdata")

rownames <- junctions.of.interest.final

extracted_cols <- lapply(win.mutpsi, function(x) x[,"mut.psi"]) #note the added comma!
mut.psi.per.window <- as.data.frame(do.call(cbind, extracted_cols))
colnames(mut.psi.per.window) = as.character(1:ncol(mut.psi.per.window))
rownames(mut.psi.per.window) <- rownames

extracted_cols <- lapply(win.mutpsi, function(x) x[,"mut.bulk.counts"]) 
mut.counts.per.window <- as.data.frame(do.call(cbind, extracted_cols))
colnames(mut.counts.per.window) = as.character(1:ncol(mut.counts.per.window))
rownames(mut.counts.per.window) <- rownames

extracted_cols <- lapply(win.mutpsi, function(x) x[,"cluster.cov.mut"]) #note the added comma!
cluster.cov.mut.per.window <- as.data.frame(do.call(cbind, extracted_cols))
colnames(cluster.cov.mut.per.window) = as.character(1:ncol(cluster.cov.mut.per.window))
rownames(cluster.cov.mut.per.window) <- rownames


#mut.psi.per.window[which(rowMeans(is.na(mut.psi.per.window)) > 0.1), ]
mut.psi.per.window.zscore <- as.data.frame(t(scale(t(mut.psi.per.window))))
#mut.psi.per.window[mut.psi.per.window==0] <- .000001



```


```{r}
ch.mut.psi.per.window.8 <- mut.psi.per.window
```


##Adding Annotation Colors 

```{r fig.height=1.5, fig.width=10}

annotation_col = data.frame(ep.fraction = as.numeric(paste(ct.fractions.2["EP",])), 
                            mep.fraction = as.numeric(paste(ct.fractions.2["MEP",])), 
                            imp.fraction = as.numeric(paste(ct.fractions.2["IMP",])),
                            hsc.fraction = as.numeric(paste(ct.fractions.2["HSPC",])),  
                            mean.pseudotime = pseudotime.mean)

rownames(annotation_col) = paste(1:nrow(annotation_col))

cell.freq <- melt(t(annotation_col[,1:4]),var=c("CellType",'window'))
mycolors <- c("#980043","#4B0082","#045A8D","#313695")

#pdf("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/16.Pseudotime.Window/CH.output/pdfs/ct.sub/heatmap/CH.perc.ct.per.bin.mut.only.60.pdf", width = 20, height = 2)
ggplot(cell.freq, aes(x = window, y = value, fill = CellType)) + geom_col(position = "fill") + scale_fill_manual(values=mycolors) + theme_void() + theme(legend.key.size = unit(0.5, 'cm'), #change legend key size
        legend.key.height = unit(0.5, 'cm'), #change legend key height
        legend.key.width = unit(0.5, 'cm'), #change legend key width
        legend.title = element_text(size=10), #change legend title font size
        legend.text = element_text(size=8)) #change legend text font size
#dev.off()

```


```{r}
shared.junctions <- intersect(junctions.of.interest.final, rownames(mut.psi.per.window))
length(shared.junctions)


annotation_row$mutpsi.exp.corr <- NA


for (junction in shared.junctions){
  gene.name <- unlist(strsplit(junction,split= ":"), "[",1)[1]
  #print(junction)
  #print(gene.name)
  
  if (gene.name %in% rownames(ont.mut.expression.per.window)){
    psi.data <- as.data.frame(t(mut.psi.per.window))[,junction]
    exp.data <- as.data.frame(t(ont.mut.expression.per.window))[,gene.name]
    annotation_row[junction,]$mutpsi.exp.corr <- round(cor(psi.data, exp.data, method = c("spearman") ), 4)
    
  }

}

annotation_row
annotation_row %>% arrange(desc(mutpsi.exp.corr))
```



```{r}
annotation_row$corr.class <- 0.5
annotation_row[which(annotation_row$mutpsi.exp.corr >= 0.5),]$corr.class <- 1
annotation_row[which(annotation_row$mutpsi.exp.corr < 0 & annotation_row$mutpsi.exp.corr > -0.5),]$corr.class <- -0.5
annotation_row[which(annotation_row$mutpsi.exp.corr <= -0.5),]$corr.class <- -1

annotation_row$corr.class.2 <- 0
annotation_row[which(annotation_row$mutpsi.exp.corr >= 0.5),]$corr.class.2 <- 1
annotation_row[which(annotation_row$mutpsi.exp.corr <= -0.5),]$corr.class.2 <- -1

table(annotation_row$corr.class)
annotation_row
```



##Add to annotation_row:
#1. Delta value for change in exp (should it be first 2 and last 2 windows? Or between lowest and highest window?)
#2. Dleta value for change in psi across window
#3. Average coverage across windows value
#4. NMD verdict 
#5. distance of cryptic site 
#6. Fishers?



```{r}
library(gtools)
annotation_row$junction <- rownames(annotation_row)

#annotation_row
#mut.psi.per.window
#ont.mut.expression.per.window


##Preparing NMD.info
nmd.info <- read.csv("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/18.A3SS.NMD.Expression/logOR_within_cell_type_ONLY_CRYPTIC_3P_Junctions_with_threshold_info_w_NMD_faster_with_start_codon_correction.csv")

nmd.info$gene.and.junc <- paste0( nmd.info$gene,":",nmd.info$intron_junction)

#Addingin new columns 
#Addingin new columns 
annotation_row$psi.range <- NA
annotation_row$psi.diff <- NA
annotation_row$psi.log2fc <- NA
#annotation_row$psi.fishers <- NA
annotation_row$exp.range <- NA
annotation_row$exp.diff <- NA
annotation_row$exp.log2fc <- NA
annotation_row$avg.cov <- NA
annotation_row$NMD.junction.verdict <- NA
annotation_row$NMD.gene.verdict <- NA
annotation_row$threeprime_distance <- NA

shared.junctions <- intersect(junctions.of.interest.final, rownames(mut.psi.per.window))
num.windows <- ncol(mut.psi.per.window)

for (junction in shared.junctions){
  #junction <- "MTRF1L:chr6:152991416:152992856:-"
  annotation_row[junction,]$psi.range <- max(as.numeric(paste(mut.psi.per.window[junction,])))-min(as.numeric(paste(mut.psi.per.window[junction,])))
  
  first2.avg <- mean(c(mut.psi.per.window[junction,1], mut.psi.per.window[junction,2]))
  last2.avg <- mean(c(mut.psi.per.window[junction,(num.windows-1)], mut.psi.per.window[junction,num.windows]))
  
  annotation_row[junction,]$psi.diff <- last2.avg-first2.avg
  annotation_row[junction,]$psi.log2fc <- log2(last2.avg/first2.avg)
  
  gene <- unlist(strsplit(junction,split= ":"))[[1]]
  if (gene %in% rownames(ont.mut.expression.per.window)){
    
    annotation_row[junction,]$exp.range <- max(as.numeric(paste(ont.mut.expression.per.window[gene,])))-min(as.numeric(paste(ont.mut.expression.per.window[gene,])))
    
    first2.avg <- mean(c(ont.mut.expression.per.window[gene,1], ont.mut.expression.per.window[gene,2]))
    last2.avg <- mean(c(ont.mut.expression.per.window[gene,(num.windows-1)], ont.mut.expression.per.window[gene,num.windows]))
    
    annotation_row[junction,]$exp.diff <- last2.avg-first2.avg
    annotation_row[junction,]$exp.log2fc <- log2(last2.avg/first2.avg)
    
  }
  
  annotation_row[junction,]$avg.cov <- mean(unlist(cluster.cov.mut.per.window[junction,]))
  
  
  if (junction %in% nmd.info$gene.and.junc ){
    annotation_row[junction,]$NMD.junction.verdict <- paste(nmd.info[which(nmd.info$gene.and.junc == junction),]$NMDVerdict)
    gene <- unlist(lapply(strsplit(junction,split= ":"), "[",1))
    if(gene %in% nmd.genes){
      annotation_row[junction,]$NMD.gene.verdict <- "NMD gene"
    } else {
      annotation_row[junction,]$NMD.gene.verdict <- "not NMD gene"
    }
    
    annotation_row[junction,]$threeprime_distance <- nmd.info[which(nmd.info$gene.and.junc == junction),]$threep_distance
  }

}


annotation_row

```


##Manual adjustment of NMD verdict
```{r}

annotation_row[which(annotation_row$junction == "STX4:chr16:31034114:31034203:+" ),]$NMD.junction.verdict <- NA
annotation_row[which(annotation_row$junction == "STX4:chr16:31034114:31034203:+" ),]$NMD.gene.verdict <- NA
annotation_row[which(annotation_row$junction == "DUSP12:chr1:161750145:161751654:+" ),]$NMD.junction.verdict <- NA
annotation_row[which(annotation_row$junction == "DUSP12:chr1:161750145:161751654:+" ),]$NMD.gene.verdict <- NA


annotation_row["AP1G2:chr14:23564221:23564332:-",]$NMD.junction.verdict <- NA
annotation_row["AP1G2:chr14:23564221:23564332:-",]$NMD.gene.verdict <- NA

annotation_row[which(is.na(annotation_row$NMD.junction.verdict)),]$NMD.junction.verdict <- "NA"
annotation_row[which(is.na(annotation_row$NMD.gene.verdict)),]$NMD.gene.verdict <- "NA"
```



```{r}
ch.annotation_row.8 <- annotation_row
```


```{r}
mut.psi.per.window
shared.junctions
annotation_row
annotation_row %>% arrange(desc(abs(psi.log2fc)))
annotation_row %>% arrange(desc(abs(delta.exp)))
annotation_row %>% arrange(NMD.verdict.adj)
table(annotation_row$NMD.verdict.adj)
table(annotation_row[which(annotation_row$threeprime_distance > -10),]$NMD.verdict.adj)
table(annotation_row[which(annotation_row$threeprime_distance <= -10),]$NMD.verdict.adj)
table(annotation_row[which(annotation_row$threeprime_distance <= -10 & annotation_row$threeprime_distance >= -50),]$NMD.verdict.adj)
```

```{r}
read.cov <- 5 
psi.thresh <- 5
fc.thresh <- 0.2

annotation_row[which(annotation_row$sum >0 &annotation_row$avg.cov >= read.cov & abs(annotation_row$psi.diff) >= psi.thresh & abs(annotation_row$exp.log2fc) >= fc.thresh),]

annotation_row[which(annotation_row$sum >0 &annotation_row$avg.cov >= read.cov & abs(annotation_row$psi.diff) >= psi.thresh & abs(annotation_row$exp.log2fc) >= fc.thresh & annotation_row$NMD.junction.verdict == "not NMD" & annotation_row$NMD.gene.verdict =="not NMD gene"),]


annotation_row[which(annotation_row$sum >0 &annotation_row$avg.cov >= read.cov & abs(annotation_row$psi.diff) >= psi.thresh & abs(annotation_row$exp.log2fc) >= fc.thresh & annotation_row$NMD.junction.verdict == "not NMD" & annotation_row$NMD.gene.verdict =="NMD gene"),]

annotation_row[which(annotation_row$sum >0 &annotation_row$avg.cov >= read.cov & abs(annotation_row$psi.diff) >= psi.thresh & abs(annotation_row$exp.log2fc) >= fc.thresh & annotation_row$NMD.junction.verdict == "NMD" & annotation_row$NMD.gene.verdict =="NMD gene"),]
```



```{r}
annotation_row[which(annotation_row$sum >0 &annotation_row$avg.cov >= 10 & abs(annotation_row$psi.diff) >= 5 & abs(annotation_row$exp.log2fc) > 0.2 & !is.na(annotation_row$NMD.verdict.adj)),]

table(annotation_row[which(annotation_row$sum >0 &annotation_row$avg.cov >= 10 & abs(annotation_row$psi.diff) >= 5 & abs(annotation_row$exp.log2fc) > 0.2 & annotation_row$NMD.verdict.adj =="NMD"),]$corr.class)

table(annotation_row[which(annotation_row$sum >0 &annotation_row$avg.cov >= 10 & abs(annotation_row$psi.diff) >= 5 & abs(annotation_row$exp.log2fc) > 0.2 & annotation_row$NMD.verdict.adj =="not NMD"),]$corr.class)
```



```{r}
#table(annotation_row[which(annotation_row$threeprime_distance <= -10 & annotation_row$NMD.verdict.adj =="NMD"),]$corr.class)
#table(annotation_row[which(annotation_row$threeprime_distance <= -10 & annotation_row$NMD.verdict.adj =="not NMD"),]$corr.class)


table(annotation_row[which(annotation_row$NMD.verdict.adj =="NMD"),]$corr.class)
table(annotation_row[which(annotation_row$NMD.verdict.adj =="not NMD"),]$corr.class)


table(annotation_row[which(annotation_row$avg.cov >=20 & annotation_row$NMD.verdict.adj =="NMD"),]$corr.class)
table(annotation_row[which(annotation_row$avg.cov >=20 & annotation_row$NMD.verdict.adj =="not NMD"),]$corr.class)
```


```{r}
annotation_row[which(annotation_row$avg.cov >= 10 & abs(annotation_row$psi.log2fc) >= 0.5 & abs(annotation_row$exp.log2fc) < 0.25),] %>% arrange(desc(abs(psi.log2fc)))

#mds.heatmap.junction.list.full <- rownames(annotation_row[which(annotation_row$sum > 0),])
ch.heatmap.junction.list.full <- rownames(annotation_row[which(annotation_row$sum > 0),])
rm(ch.junctions.full)


ch.heatmap.junction.list.full
mds.heatmap.junction.list.full
all <- union(ch.heatmap.junction.list.full, mds.heatmap.junction.list.full)

length(all)

length(intersect(ch.heatmap.junction.list.full, mds.heatmap.junction.list.full))

save(mds.heatmap.junction.list.full, ch.heatmap.junction.list.full, file="/gpfs/commons/home/pchamely/cryptic.junction.Rdata")
```


```{r}
annotation_row$in.MDS <- "FALSE"


annotation_row
length(mds.heatmap.junction.list.full)
annotation_row[which(annotation_row$junction %in% mds.heatmap.junction.list.full),]$in.MDS <- "TRUE"
#annotation_row[which(annotation_row$junction %in% mds.heatmap.junction.list.sub),]

#Total - union of MDS and CH junctions 65

```



```{r}
ann_colors = list(
  #NP = c("light grey","red"),
  #IMP = c("light grey","red"),
  #HSPC = c("light grey","red"),
  #MkP = c("light grey","red"),
  #MEP = c("light grey","red"),
  #EP = c("light grey","red"),
  #corr.class = c("orange","#ffe4b3","plum", "purple")
  #corr.class.2 = c("orange","white", "purple")
  #NMD.verdict = c( "NMD" = "orange","not NMD"= "purple", "NA" = "grey"),
  in.MDS = c("FALSE" = "grey", "TRUE" = "purple")
)
```




```{r}
psi.range <- 5
annotation_row[which(annotation_row$sum >0 &annotation_row$psi.range >= psi.range ),]

table(annotation_row[which(annotation_row$sum >0 & annotation_row$psi.range >= psi.range & annotation_row$NMD.junction.verdict == "NMD" ),]$corr.class)
table(annotation_row[which(annotation_row$sum >0 & annotation_row$psi.range >= psi.range & annotation_row$NMD.junction.verdict == "not NMD" ),]$corr.class)

#psi.range <- 5
#total = 41
#p-value = 0.0278
```

```{r}

exp.range <- 0.05
annotation_row[which(annotation_row$sum >0 &annotation_row$psi.range >= psi.range & annotation_row$exp.range >= exp.range ),]

table(annotation_row[which(annotation_row$sum >0 & annotation_row$psi.range >= psi.range & annotation_row$exp.range >= exp.range & annotation_row$NMD.junction.verdict == "NMD" ),]$corr.class)
table(annotation_row[which(annotation_row$sum >0 & annotation_row$psi.range >= psi.range & annotation_row$exp.range >= exp.range & annotation_row$NMD.junction.verdict == "not NMD" ),]$corr.class)


#psi.range <- 5
#exp.range <- 0.05
#total = 30
#p-value = 0.0261

```





##Look at events with :
#1. > 10 avg coverage
#2. > 5 delta.psi
#3. > 0.1 delta exp
#4. Fishers?


##Add also in MDS as an annotation row

```{r}
annotation_row
annotation_row.sub
```


```{r fig.height=12, fig.width=10}
#shared.junctions <- intersect(junctions.of.interest.final, rownames(dPSI.per.window.sub))

##Ordering it by z-score
#to.plot <- mut.psi.per.window
to.plot <- mut.psi.per.window.zscore[rownames(mut.psi.per.window.zscore) %in% all.examples,]
#to.plot <- mut.psi.per.window.zscore[rownames(mut.psi.per.window.zscore) %in% unique.examples,]

junctions.sub <- rownames(annotation_row[which(annotation_row$sum >0 & annotation_row$avg.cov >= 10 & annotation_row$psi.range >= 2),])


#junctions.sub <- rownames(annotation_row[which(annotation_row$avg.cov >= 10 & abs(annotation_row$psi.log2fc) > 0.25 & abs(annotation_row$exp.log2fc) > 0.25 & annotation_row$NMD.verdict.adj == "not NMD"),])

junctions.sub <- rownames(annotation_row[which(annotation_row$avg.cov >= 10 & abs(annotation_row$psi.log2fc) > 0.2 & abs(annotation_row$exp.log2fc) > 0.2 & !is.na(annotation_row$NMD.verdict.adj)),])

junctions.sub <- rownames(annotation_row[which(annotation_row$threeprime_distance <= -10 & !is.na(annotation_row$NMD.verdict.adj)),])

junctions.sub <- rownames(annotation_row[which(annotation_row$threeprime_distance <= -10 & annotation_row$threeprime_distance >=-40 & annotation_row$avg.cov >= 10 & !is.na(annotation_row$NMD.verdict.adj)),])




to.plot <- to.plot[junctions.sub,]

##Reordering by highest average of every 2 bin z-scores
z.score.order <- do.call(cbind, 
                         lapply(seq(1, ncol(to.plot)-1, 1), function(i){
                           x <- as.data.frame(rowMeans(to.plot[,i:(i + 1)]))
                           colnames(x) <- c(paste0(i,".", (i+1)))
                           x
                           }))

z.score.order$max.column <- colnames(z.score.order)[apply(z.score.order,1,which.max)]
z.score.order$max.column <- as.numeric(z.score.order$max.column)
z.score.order$rownames <- rownames(z.score.order)
z.score.order <- z.score.order %>% arrange(max.column)
rownames(z.score.order) <- z.score.order$rownames

to.plot.reordered <- to.plot[rownames(z.score.order),]

paletteLength <- 50
myColor <- colorRampPalette(c("navy", "white", "red"))(paletteLength)
# length(breaks) == length(paletteLength) + 1
# use floor and ceiling to deal with even/odd length pallettelengths
myBreaks <- c(seq(min(to.plot.reordered[!is.na(to.plot.reordered)]), 0, length.out=ceiling(paletteLength/2) + 1), 
              seq(max(to.plot.reordered[!is.na(to.plot.reordered)])/paletteLength, max(to.plot.reordered[!is.na(to.plot.reordered)]), length.out=floor(paletteLength/2)))

#annotation_row.sub <- as.data.frame(annotation_row[rownames(to.plot.reordered),c(7,18)])
#colnames(annotation_row.sub) <- c("corr.class","NMD.verdict.adj")
#annotation_row.sub <- as.data.frame(annotation_row[rownames(to.plot.reordered),c(18)])
#colnames(annotation_row.sub) <- c("NMD.verdict")

annotation_row.sub <- as.data.frame(annotation_row[rownames(to.plot.reordered),c(20)])
colnames(annotation_row.sub) <- c("in.MDS")


rownames(to.plot.reordered) <- make.names(unlist(lapply(strsplit(rownames(to.plot.reordered),split= ":"), "[",1)), unique =TRUE)
make.names(unlist(lapply(strsplit(rownames(to.plot.reordered),split= ":"), "[",1)), unique =TRUE)

rownames(annotation_row.sub) <- rownames(to.plot.reordered)


pheatmap(to.plot.reordered, cluster_cols = F, cluster_rows = F, annotation_colors = ann_colors, color=myColor, border_color=NA, annotation_legend= T, breaks=myBreaks)


pdf("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/16.Pseudotime.Window/CH.output/pdfs/ct.sub/heatmap/CH.combined.heatmap.mut.psi.variable.MDS.overlab.60.v1.pdf", width = 20, height = 30)
pheatmap(to.plot.reordered, cluster_cols = F, cluster_rows = F, annotation_row = annotation_row.sub, annotation_colors = ann_colors, color=myColor, border_color=NA, annotation_legend= T, breaks=myBreaks)
dev.off()

#pheatmap(to.plot.reordered[rownames(annotation_row.sub[which(annotation_row.sub$corr.class.2 != 0),]),], cluster_cols = F, cluster_rows = F, annotation_row = annotation_row.sub[which(annotation_row.sub$corr.class.2 != 0),], annotation_colors = ann_colors, color=myColor, border_color=NA, annotation_legend= T, breaks=myBreaks)


pdf("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/16.Pseudotime.Window/CH.output/pdfs/ct.sub/heatmap/CH.combined.heatmap.mut.psi.corr.annotation.withNMD.60.v2.pdf", width = 20, height = 30)
#pheatmap(to.plot.reordered, cluster_cols = F, cluster_rows = F, color=myColor, border_color=NA, annotation_legend=F, breaks=myBreaks)
pheatmap(to.plot.reordered, cluster_cols = F, cluster_rows = F, annotation_row = annotation_row.sub, annotation_colors = ann_colors, color=myColor, border_color=NA, annotation_legend= T, breaks=myBreaks)
dev.off()
```




```{r fig.height=0.8, fig.width=25}
pdf("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/16.Pseudotime.Window/CH.output/pdfs/ct.sub/heatmap/CH.mean.pseudotime.per.bin.mut.only.60.pdf", width = 20, height = 0.8)
pheatmap(t(annotation_col[,5]), cluster_cols = F, cluster_rows = F, color=colorRampPalette(c("white","dark green" ))(50),legend=F)
dev.off()
```

##Filter out events that don't have alot of change 

```{r}

annotation_row
annotation_row[which(annotation_row$sum >0 &annotation_row$avg.cov >= 5 & abs(annotation_row$psi.diff) >= 5 & abs(annotation_row$exp.log2fc) >= 0.15 & !is.na(annotation_row$NMD.verdict.adj)),]

table(annotation_row[which(annotation_row$sum >0 &annotation_row$avg.cov >= 10 & abs(annotation_row$psi.diff) >= 5 & abs(annotation_row$exp.log2fc) > 0.15 & annotation_row$NMD.verdict.adj =="NMD"),]$corr.class)

table(annotation_row[which(annotation_row$sum >0 &annotation_row$avg.cov >= 10 & abs(annotation_row$psi.diff) >= 5 & abs(annotation_row$exp.log2fc) > 0.15 & annotation_row$NMD.verdict.adj =="not NMD"),]$corr.class)
```


##Candidates for individual events:

```{r}
#EIF4A2:chr3:186783639:186784428:+
#OXA1L:chr14:22768171:22769776:+
#SLC25A39:chr17:44322552:44322807:-
```



## ------------------------------------------ Looking at Individual Events ------------------------------------------ ##



##Good anti-correlation examples

```{r}

junction <- "chr4:5016989:5017134:-"
gene <- "CYTL1"

junction <- "chr11:114402915:114405705:+"
gene <- "RBM7"
  
junction <- "chr11:61393976:61397727:+"
gene <- "TMEM216"
  
junction <- "chr6:152991416:152992856:-"
gene <- "MTRF1L"

junction <- "chr2:74527121:74527247:-"
gene <- "AUP1"
  
junction <- "chr11:64232614:64232700:+"
gene <- "DNAJC4"

junction <- "chr22:23575085:23579984:-"
gene <- "IGLL1"

junction <- "chrX:134486530:134493487:+"
gene <- "HPRT1"

junction <- "chrX:153588454:153592505:-" 
gene <- "CCNQ"

junction <- "chr3:39406764:39407602:+" 
gene <- "RPSA"

junction <- "chr3:186783639:186784428:+" 
gene <- "EIF4A2"


junction <- "chr11:17295467:17296020:+" 
gene <- "NUCB2"

```



```{r}
junction <- "chrX:48793297:48793777:+" 
gene <- "GATA1"

#Not well covered
junction <- "chr14:22768171:22769776:+" 
gene <- "OXA1L"

#junction <- "chr10:119582608:119588151:-"
#gene <- "TIAL1"

#junction <- "chr4:139072039:139073453:-"
#gene <- "ELF2"

junction <- "chr20:35556271:35556954:+" 
gene <- "ERGIC3"

#junction <- "chr9:136791424:136791625:+" 
#gene <- "TMEM141"

junction <- "chr1:100992754:100995109:-" 
gene <- "DPH5"

junction <- "chr3:170868201:170868297:-" 
gene <- "RPL22L1"

junction <- "chr11:64232614:64232700:+" 
gene <- "DNAJC4"

junction <- "chr1:45012285:45012856:+" 
gene <- "UROD"

junction <- "chr2:169812524:169815476:-" 
gene <- "METTL5"

junction <- "chr15:58917020:58932355:-" 
gene <- "SLTM"

junction <- "chr19:11035132:11041297:+" 
gene <- "SMARCA4"

junction <- "chr2:241335212:241335958:+" 
gene <- "SEPT2"

```


##Ones that show highly dPSIs than in the buckets they are sig in
```{r}

##This one had very low coverage in the Diff Transcript usage - so it is really a trustworkthy sig event??
junction <- "chr11:65572960:65573352:+" 
gene <- "FAM89B"

junction <- "chr12:113159112:113161269:-" 
gene <- "DDX54"


junction <- "chr14:20990598:20992091:+" 
gene <- "METTL17"


junction <- "chr11:17295467:17296020:+" 
gene <- "NUCB2"


junction <- "chr10:119582608:119588151:-" 
gene <- "TIAL1"


junction <- "chr14:22768171:22769776:+" 
gene <- "OXA1L"

junction <- "chr17:44322552:44322807:-" 
gene <- "SLC25A39"


junction <- "chr19:11035132:11041297:+" 
gene <- "SMARCA4"


junction <- "chr9:35813156:35813265:-" 
gene <- "AL133410.3"



junction <- "chr1:67424322:67424887:-" 
gene <- "SERBP1"


junction <- "chr1:67424977:67425082:-" 
gene <- "SERBP1"


junction <- "chr19:48960914:48961482:+" 
gene <- "BAX"


junction <- "chr1:161750145:161751654:+" 
gene <- "DUSP12"


junction <- "chr5:139388601:139389679:-" 
gene <- "MZB1"

junction <- "chr11:77622689:77624962:-" 
gene <- "CLNS1A"


junction <- "chr11:61393976:61397727:+" 
gene <- "TMEM216"




```


```{r}

junction.final <- paste0(gene,":",junction)

colnames <- c("window","mut.psi","ONT.MUT.Expression")
psi.per.window.data <- as.data.frame(matrix(nrow=ncol(mut.psi.per.window), ncol=length(colnames)))
rownames(psi.per.window.data) <- as.character(1:ncol(mut.psi.per.window))
colnames(psi.per.window.data) <- colnames

psi.per.window.data$window <- rownames(psi.per.window.data)
psi.per.window.data$mut.psi <- as.numeric(paste(mut.psi.per.window[junction.final,]))
psi.per.window.data$ONT.MUT.Expression <- as.numeric(paste(ont.mut.expression.per.window[gene,]))
psi.per.window.data$window <- factor(psi.per.window.data$window,levels = 1:nrow(psi.per.window.data))
psi.per.window.data

mask <- apply(psi.per.window.data, 2, is.na)
psi.per.window.data[mask] <- 0
psi.per.window.data
```


```{r}

ExpressionColor <- "red"
PSIColor <- "steelblue"

OldMax = max(psi.per.window.data$ONT.MUT.Expression)
OldMin = min(psi.per.window.data$ONT.MUT.Expression)
OldRange = (OldMax - OldMin)

NewMax = max(psi.per.window.data$mut.psi)
NewMin = min(psi.per.window.data$mut.psi)
NewRange = (NewMax - NewMin)


#pdf(paste0("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/16.Pseudotime.Window/MDS.output/pdfs/ct.sub/individual.events/",gene,".mutpsi.vs.mutExpression.300.pdf"), width = 12, height = 8)

ggplot(psi.per.window.data, aes(x=window, y=mut.psi, group = 1)) + 
  geom_bar(stat="identity", fill=PSIColor) +
  geom_line(aes(y=(((ONT.MUT.Expression - OldMin) * NewRange) / OldRange) + NewMin), linetype = "dashed" , color="red") +
  geom_point(aes(y=(((ONT.MUT.Expression - OldMin) * NewRange) / OldRange) + NewMin),color="red") +
  scale_y_continuous(
    name = "mut.psi",
    #breaks=seq(min(psi.per.window.data$mut.psi), max(psi.per.window.data$mut.psi), length=5),
    #labels=as.character(round(seq(m, max(psi.per.window.data$mut.psi+m), length=5),2)),
    sec.axis = sec_axis(~(((.-NewMin)*OldRange)/NewRange) + OldMin , name="ONT Expression of MUT cells \n (average of normalized counts)")) +
  theme_classic() +
  labs(x="bin",
       title = paste0("dPSI and Expression for ", gene ," cryptic junction - ", junction),
       subtitle = paste0("Spearman Correlation : ", round(cor(psi.per.window.data$mut.psi, psi.per.window.data$ONT.MUT.Expression, method = c("spearman") ), 4))) +
  theme(plot.title = element_text(vjust = 2.5,hjust = 0.5, size = 17), 
        axis.title.y = element_text(color =PSIColor , size=15), 
        axis.title.y.right = element_text(color = ExpressionColor, size=15),
        plot.subtitle=element_text(size=12, hjust=0.5, face="italic", color="red")) 

#dev.off()

```


```{r}

psi.per.window.data
ExpressionColor <- "red"
PSIColor <- "steelblue"

m <- floor(min(psi.per.window.data$mut.psi))
psi.per.window.data$mut.psi <- psi.per.window.data$mut.psi - m
psi.per.window.data


OldMax = max(psi.per.window.data$ONT.MUT.Expression)
OldMin = min(psi.per.window.data$ONT.MUT.Expression)
OldRange = (OldMax - OldMin)

NewMax = max(psi.per.window.data$mut.psi)
NewMin = min(psi.per.window.data$mut.psi)
NewRange = (NewMax - NewMin)


#pdf(paste0("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/16.Pseudotime.Window/CH.output/pdfs/ct.sub/individual.events/",gene,".mutpsi.vs.mutExpression.300.pdf"), width = 12, height = 8)

ggplot(psi.per.window.data, aes(x=window, y=mut.psi, group = 1)) + 
  geom_bar(stat="identity", fill=PSIColor) +
  geom_line(aes(y=(((ONT.MUT.Expression - OldMin) * NewRange) / OldRange) + NewMin), linetype = "dashed" , color="red") +
  geom_point(aes(y=(((ONT.MUT.Expression - OldMin) * NewRange) / OldRange) + NewMin),color="red") +
  scale_y_continuous(
    name = "mut.psi",
    breaks=seq(0,50),
    #labels=seq(m, ceiling(max(psi.per.window.data$mut.psi+m)),length=length(seq(0,10))),
    labels=(seq(0,50) + m),
    sec.axis = sec_axis(~(((.-NewMin)*OldRange)/NewRange) + OldMin , name="ONT Expression of MUT cells \n (average of normalized counts)")) +
  theme_classic() +
  labs(x="bin",
       title = paste0("dPSI and Expression for ", gene ," cryptic junction - ", junction),
       subtitle = paste0("Spearman Correlation : ", round(cor(psi.per.window.data$mut.psi, psi.per.window.data$ONT.MUT.Expression, method = c("spearman") ), 4))) +
  theme(plot.title = element_text(vjust = 2.5,hjust = 0.5, size = 17), 
        axis.title.y = element_text(color =PSIColor , size=15), 
        axis.title.y.right = element_text(color = ExpressionColor, size=15),
        plot.subtitle=element_text(size=12, hjust=0.5, face="italic", color="red")) 

#dev.off()

#psi.per.window.data$mut.psi <- psi.per.window.data$mut.psi + m
#psi.per.window.data

```



```{r}
#FeaturePlot(mds.samples.integrated, features = gene)
FeaturePlot(samples.integrated, features = gene, reduction = "umap")

```


```{r}
rownames <- junctions.of.interest.final

##HSPC
extracted_cols <- lapply(win.ct.counts, function(x) x[,"HSPC.mut.bulk.counts"]) #note the added comma!
HSPC.mut.bulk.counts.per.window <- as.data.frame(do.call(cbind, extracted_cols))
colnames(HSPC.mut.bulk.counts.per.window) = as.character(1:ncol(HSPC.mut.bulk.counts.per.window))
rownames(HSPC.mut.bulk.counts.per.window) <- rownames

extracted_cols <- lapply(win.ct.counts, function(x) x[,"HSPC.cluster.cov.mut"]) #note the added comma!
HSPC.cluster.cov.mut.per.window <- as.data.frame(do.call(cbind, extracted_cols))
colnames(HSPC.cluster.cov.mut.per.window) = as.character(1:ncol(HSPC.cluster.cov.mut.per.window))
rownames(HSPC.cluster.cov.mut.per.window) <- rownames


##IMP
extracted_cols <- lapply(win.ct.counts, function(x) x[,"IMP.mut.bulk.counts"]) #note the added comma!
IMP.mut.bulk.counts.per.window <- as.data.frame(do.call(cbind, extracted_cols))
colnames(IMP.mut.bulk.counts.per.window) = as.character(1:ncol(IMP.mut.bulk.counts.per.window))
rownames(IMP.mut.bulk.counts.per.window) <- rownames

extracted_cols <- lapply(win.ct.counts, function(x) x[,"IMP.cluster.cov.mut"]) #note the added comma!
IMP.cluster.cov.mut.per.window <- as.data.frame(do.call(cbind, extracted_cols))
colnames(IMP.cluster.cov.mut.per.window) = as.character(1:ncol(IMP.cluster.cov.mut.per.window))
rownames(IMP.cluster.cov.mut.per.window) <- rownames


##MEP
extracted_cols <- lapply(win.ct.counts, function(x) x[,"MEP.mut.bulk.counts"]) #note the added comma!
MEP.mut.bulk.counts.per.window <- as.data.frame(do.call(cbind, extracted_cols))
colnames(MEP.mut.bulk.counts.per.window) = as.character(1:ncol(MEP.mut.bulk.counts.per.window))
rownames(MEP.mut.bulk.counts.per.window) <- rownames

extracted_cols <- lapply(win.ct.counts, function(x) x[,"MEP.cluster.cov.mut"]) #note the added comma!
MEP.cluster.cov.mut.per.window <- as.data.frame(do.call(cbind, extracted_cols))
colnames(MEP.cluster.cov.mut.per.window) = as.character(1:ncol(MEP.cluster.cov.mut.per.window))
rownames(MEP.cluster.cov.mut.per.window) <- rownames

##EP
extracted_cols <- lapply(win.ct.counts, function(x) x[,"EP.mut.bulk.counts"]) #note the added comma!
EP.mut.bulk.counts.per.window <- as.data.frame(do.call(cbind, extracted_cols))
colnames(EP.mut.bulk.counts.per.window) = as.character(1:ncol(EP.mut.bulk.counts.per.window))
rownames(EP.mut.bulk.counts.per.window) <- rownames

extracted_cols <- lapply(win.ct.counts, function(x) x[,"EP.cluster.cov.mut"]) #note the added comma!
EP.cluster.cov.mut.per.window <- as.data.frame(do.call(cbind, extracted_cols))
colnames(EP.cluster.cov.mut.per.window) = as.character(1:ncol(EP.cluster.cov.mut.per.window))
rownames(EP.cluster.cov.mut.per.window) <- rownames


junction.final
raw.counts <- rbind(HSPC.mut.bulk.counts.per.window[junction.final,], HSPC.cluster.cov.mut.per.window[junction.final,] ,
                    IMP.mut.bulk.counts.per.window[junction.final,], IMP.cluster.cov.mut.per.window[junction.final,],
                    MEP.mut.bulk.counts.per.window[junction.final,], MEP.cluster.cov.mut.per.window[junction.final,],
                    EP.mut.bulk.counts.per.window[junction.final,], EP.cluster.cov.mut.per.window[junction.final,],
                    mut.counts.per.window[junction.final,], cluster.cov.mut.per.window[junction.final,])

raw.counts <- as.data.frame(sapply(raw.counts, as.integer))
rownames(raw.counts) <- c("hspc.cryptic.reads.mut", "hspc.cluster.cov.mut",
                          "imp.cryptic.reads.mut", "imp.cluster.cov.mut",
                          "mep.cryptic.reads.mut", "mep.cluster.cov.mut",
                          "ep.cryptic.reads.mut", "ep.cluster.cov.mut",
                          "all.cryptic.reads.mut", "all.cluster.cov.mut")

pheatmap(raw.counts, cluster_cols = F, cluster_rows = F, legend=F, display_numbers = T, number_format= "%.f",number_color = "blue",color=colorRampPalette(c("white", "white"))(50), show_rownames = T, show_colnames = F, gaps_row = c(8))
```



```{r}
annotation_row[junction.final,]
```





```{r }
ExpressionColor <- "red"
dPSIColor <- "blue"

OldMax = max(psi.per.window.data$ONT.Expression)
OldMin = min(psi.per.window.data$ONT.Expression)
OldRange = (OldMax - OldMin)


NewMax = max(psi.per.window.data$dPSI)
NewMin = min(psi.per.window.data$dPSI)
NewRange = (NewMax - NewMin)

ggplot(psi.per.window.data, aes(x=window, group = 1)) + 
  geom_line( aes(y=dPSI), color=dPSIColor) + 
  geom_line( aes(y=(((ONT.Expression - OldMin) * NewRange) / OldRange) + NewMin), linetype = "dashed", color=ExpressionColor)+  
  geom_hline(yintercept = 0, linetype = "dotted") + 
  scale_y_continuous(
    name = "dPSI",
    sec.axis = sec_axis(~(((.-NewMin)*OldRange)/NewRange) + OldMin , name="ONT Expression \n (average of normalized counts)")) +
  theme_classic()+
  theme(plot.title = element_text(hjust = 0.5, size = 14), 
        axis.title.y = element_text(color =dPSIColor , size=13), 
        axis.title.y.right = element_text(color = ExpressionColor, size=13)) +
  ggtitle(paste0("dPSI and Expression for ", gene ," cryptic junction - ", junction)) 


library(grid)
library(gridExtra)

grob = grobTree(textGrob(paste("Spearman Correlation : ", round(cor(psi.per.window.data$dPSI, psi.per.window.data$ONT.Expression, method = c("spearman") ), 4)), x=0.5, y = 0.97 ,gp = gpar(col = "red", fontsize = 11)))

final.plot <- ggplot(psi.per.window.data, aes(x=window, group = 1)) + 
  geom_point(aes(y= dPSI), color=dPSIColor) +
  stat_smooth(aes(y= dPSI), color=dPSIColor, method = "loess", formula = y ~ x, se = F, span=0.5) +
  geom_point(aes(y =(((ONT.Expression - OldMin) * NewRange) / OldRange) + NewMin), color=ExpressionColor) +
  stat_smooth(aes(y =(((ONT.Expression - OldMin) * NewRange) / OldRange) + NewMin), linetype = "dashed", color=ExpressionColor, method = "loess", formula = y ~ x, se = F, span=0.5)+
  geom_hline(yintercept = 0, linetype = "dotted") + 
  scale_y_continuous(
    name = "dPSI",
    sec.axis = sec_axis(~(((.-NewMin)*OldRange)/NewRange) + OldMin , name="ONT Expression \n (average of normalized counts)")) +
  annotation_custom(grob)+
  theme_classic()+
  theme(plot.title = element_text(hjust = 0.5, size = 14), 
        axis.title.y = element_text(color =dPSIColor , size=13), 
        axis.title.y.right = element_text(color = ExpressionColor, size=13)) +
  ggtitle(paste0("dPSI and Expression for ", gene ," cryptic junction - ", junction)) 

final.plot
```








##Why is the MUT expression higher in certain windows

```{r}

thresh <- 0
MUT.info <- as.data.frame(matrix(nrow=length(win.dPSI), ncol=9))
rownames(MUT.info) <- 1:length(win.dPSI)
colnames(MUT.info) <- c("total.raw.counts","mean.exp","total.num.cells","num_norm_>_0","pct_with_gene","num_reads_<_thresh","num_reads_>=_thresh", "num_norm_>_0_reads_<_thresh","pct_reads_<_thresh")

for (i in 1:length(win.dPSI)){
  window.info <- ont.raw.counts.per.cell.per.window.per.gene[[gene]][[i]]
  MUT.info[i,"total.raw.counts"] <- sum(window.info[which(window.info$genotype =="MUT" & window.info$total >= thresh),]$gene.counts)
  MUT.info[i,"mean.exp"] <- mean(window.info[which(window.info$genotype =="MUT"),]$norm.counts)
  MUT.info[i,"total.num.cells"] <- nrow(window.info[which(window.info$genotype =="MUT"),])
  MUT.info[i,"num_norm_>_0"] <- nrow(window.info[which(window.info$genotype =="MUT" & window.info$norm.counts > 0),])
  MUT.info[i,"pct_with_gene"] <- (MUT.info[i,"num_norm_>_0"]/MUT.info[i,"total.num.cells"])*100
  MUT.info[i,"num_reads_<_thresh"] <- nrow(window.info[which(window.info$genotype =="MUT" & window.info$total < thresh),])
  MUT.info[i,"num_reads_>=_thresh"] <- nrow(window.info[which(window.info$genotype =="MUT" & window.info$total >= thresh),])
  MUT.info[i,"num_norm_>_0_reads_<_thresh"] <- nrow(window.info[which(window.info$genotype =="MUT" & window.info$norm.counts > 0 & window.info$total < thresh),])
  MUT.info[i,"pct_reads_<_thresh"] <- MUT.info[i,"num_reads_<_thresh"]/MUT.info[i,"total.num.cells"] *100
  
}

MUT.info[,c("total.raw.counts", "total.num.cells","pct_with_gene","num_reads_>=_thresh", "pct_reads_<_thresh")]

```


```{r}
WT.info <- as.data.frame(matrix(nrow=length(win.dPSI), ncol=9))
rownames(WT.info) <- 1:length(win.dPSI)
colnames(WT.info) <- c("total.raw.counts","mean.exp","total.num.cells","num_norm_>_0","pct_with_gene","num_reads_<_thresh","num_reads_>=_thresh", "num_norm_>_0_reads_<_thresh", "pct_reads_<_thresh")

for (i in 1:length(win.dPSI)){
  window.info <- ont.raw.counts.per.cell.per.window.per.gene[[gene]][[i]]
  WT.info[i,"total.raw.counts"] <- sum(window.info[which(window.info$genotype =="WT" & window.info$total >= thresh),]$gene.counts)
  WT.info[i,"mean.exp"] <- mean(window.info[which(window.info$genotype =="WT"),]$norm.counts)
  WT.info[i,"total.num.cells"] <- nrow(window.info[which(window.info$genotype =="WT"),])
  WT.info[i,"num_norm_>_0"] <- nrow(window.info[which(window.info$genotype =="WT" & window.info$norm.counts > 0),])
  WT.info[i,"pct_with_gene"] <- (WT.info[i,"num_norm_>_0"]/WT.info[i,"total.num.cells"])*100
  WT.info[i,"num_reads_<_thresh"] <- nrow(window.info[which(window.info$genotype =="WT" & window.info$total < thresh),])
  WT.info[i,"num_reads_>=_thresh"] <- nrow(window.info[which(window.info$genotype =="WT" & window.info$total >= thresh),])
  WT.info[i,"num_norm_>_0_reads_<_thresh"] <- nrow(window.info[which(window.info$genotype =="WT" & window.info$norm.counts > 0 & window.info$total < thresh),])
  WT.info[i,"pct_reads_<_thresh"] <- WT.info[i,"num_reads_<_thresh"]/WT.info[i,"total.num.cells"]*100
}

WT.info[,c("total.raw.counts", "total.num.cells","pct_with_gene","num_reads_>=_thresh", "pct_reads_<_thresh")]


```


```{r fig.height=1, fig.width=8}
raw.exp.ont.counts <- rbind(MUT.info$total.raw.counts, WT.info$total.raw.counts)
rownames(raw.exp.ont.counts) <- c("raw.count.mut","raw.count.wt")
raw.exp.ont.counts

pheatmap(raw.exp.ont.counts, cluster_cols = F, cluster_rows = F, legend=F, display_numbers = T, number_format= "%.f",number_color = "red",color=colorRampPalette(c("white", "white"))(50), show_rownames = T, show_colnames = F, gaps_row = c(1))
```



```{r fig.height=1, fig.width=8}
num.cells.per.window <- rbind(WT.info$`num_reads_>=_thresh` ,MUT.info$`num_reads_>=_thresh`)

rownames(num.cells.per.window) <- c("num.wt.cells","num.mut.cells")
num.cells.per.window

pheatmap(num.cells.per.window, cluster_cols = F, cluster_rows = F, legend=F, display_numbers = T, number_format= "%.f",number_color = "green",color=colorRampPalette(c("white", "white"))(50), show_rownames = T, show_colnames = F, gaps_row = c(1))
```


```{r}
window.info
```




```{r}
window.info <- ont.raw.counts.per.cell.per.window.per.gene[[gene]][[6]]
window.info[which(window.info$genotype =="MUT" & window.info$used ==1),]
mean(window.info[which(window.info$genotype =="MUT" & window.info$used ==1),]$norm.counts)
window.info[which(window.info$genotype =="WT" & window.info$used ==1),]
mean(window.info[which(window.info$genotype =="WT" & window.info$used ==1),]$norm.counts)

```



```{r}
window.info <- ont.raw.counts.per.cell.per.window.per.gene[[gene]][[1]]
window.info[which(window.info$genotype =="MUT"),] %>% arrange(desc(norm.counts))
window.info[which(window.info$genotype =="WT"),] %>% arrange(desc(norm.counts))

window.info[which(window.info$genotype =="MUT" & window.info$norm.counts > 10),] %>% arrange(desc(norm.counts))
window.info[which(window.info$genotype =="MUT" & window.info$norm.counts > 10 & window.info$nCount_RNA < 1000),] %>% arrange(desc(norm.counts))
window.info[which(window.info$genotype =="WT" & window.info$norm.counts > 10),] %>% arrange(desc(norm.counts))
window.info[which(window.info$genotype =="WT" & window.info$norm.counts > 10 & window.info$nCount_RNA < 1000),] %>% arrange(desc(norm.counts))

mean(window.info[which(window.info$genotype =="MUT"),]$norm.counts)
mean(window.info[which(window.info$genotype =="WT"),]$norm.counts)
```

```{r}
window.info <- ont.raw.counts.per.cell.per.window.per.gene[[gene]][[6]]
window.info[which(window.info$genotype =="MUT"),]
window.info[which(window.info$genotype =="WT"),]

window.info[which(window.info$genotype =="MUT" & window.info$norm.counts > 10),] %>% arrange(desc(norm.counts))
window.info[which(window.info$genotype =="WT" & window.info$norm.counts > 10),] %>% arrange(desc(norm.counts))

mean(window.info[which(window.info$genotype =="MUT"),]$norm.counts)
mean(window.info[which(window.info$genotype =="WT"),]$norm.counts)
```


```{r}
window.info <- ont.raw.counts.per.cell.per.window.per.gene[[gene]][[10]]
window.info[which(window.info$genotype =="MUT"),]
window.info[which(window.info$genotype =="WT"),]

window.info[which(window.info$genotype =="MUT" & window.info$norm.counts > 10),] %>% arrange(desc(norm.counts))
window.info[which(window.info$genotype =="WT" & window.info$norm.counts > 10),] %>% arrange(desc(norm.counts))

mean(window.info[which(window.info$genotype =="MUT"),]$norm.counts)
mean(window.info[which(window.info$genotype =="WT"),]$norm.counts)
```


##Typically about 10% of the cells have relatively high expression and low nCount_RNA
```{r}

```



```{r}

window.info <- ont.raw.counts.per.cell.per.window.per.gene[[gene]][[10]]
window.info[which(window.info$genotype =="MUT" & window.info$used ==1),]
mean(window.info[which(window.info$genotype =="MUT" & window.info$used ==1),]$norm.counts)
window.info[which(window.info$genotype =="WT" & window.info$used ==1),]
mean(window.info[which(window.info$genotype =="WT" & window.info$used ==1),]$norm.counts)

```



```{r}
#12-18
window.info <- ont.raw.counts.per.cell.per.window.per.gene[[gene]][[20]]
window.info[which(window.info$genotype =="MUT"),] %>% arrange(desc(norm.counts))
window.info[which(window.info$genotype =="WT"),] %>% arrange(desc(norm.counts))
```



##Scatter Plot - Is there a correlation between total read counts and normalized gene expressoin values

```{r fig.height=20, fig.width=25}

plots <- list()
plots_mut <- list()
plots_wt <- list()

for (i in 1:length(win.dPSI)){
  window.info <- ont.raw.counts.per.cell.per.window.per.gene[[gene]][[i]]
  plots[[i]] <- ggplot(window.info[which(window.info$used ==1),], aes(x=total, y=norm.counts, color = genotype)) + geom_point() + ggtitle(paste0("Window ", i))
  plots_mut[[i]] <- ggplot(window.info[which(window.info$genotype =="MUT"),], aes(x=total, y=norm.counts, color = genotype)) + geom_point() + ggtitle(paste0("Window ", i)) 
  plots_wt[[i]] <- ggplot(window.info[which(window.info$genotype =="WT"),], aes(x=total, y=norm.counts, color = genotype)) + geom_point() + ggtitle(paste0("Window ", i))
}

do.call(grid.arrange, plots)
do.call(grid.arrange, plots_mut)
do.call(grid.arrange, plots_wt)

```


```{r fig.height=5, fig.width=10}
ggplot(psi.per.window.data, aes(x=window, group = 1)) + 
  geom_point(aes(y= dPSI), color=dPSIColor) +
  stat_smooth(aes(y= dPSI), color=dPSIColor, method = "loess", formula = y ~ x, se = F, span=0.5) +
  geom_point(aes(y =(((ONT.Expression - OldMin) * NewRange) / OldRange) + NewMin), color=ExpressionColor) +
  stat_smooth(aes(y =(((ONT.Expression - OldMin) * NewRange) / OldRange) + NewMin), linetype = "dashed", color=ExpressionColor, method = "loess", formula = y ~ x, se = F, span=0.5)+
  geom_hline(yintercept = 0, linetype = "dotted") + 
  geom_rect(aes(xmin = start, xmax = end, fill = p.val), 
    ymin = -Inf, ymax = Inf, alpha = 0.2, colour = "grey50")+
  scale_y_continuous(
    name = "dPSI",
    sec.axis = sec_axis(~(((.-NewMin)*OldRange)/NewRange) + OldMin , name="ONT Expression \n (average of normalized counts)")) +
  theme_classic()+
  theme(plot.title = element_text(hjust = 0.5, size = 14), 
        axis.title.y = element_text(color =dPSIColor , size=13), 
        axis.title.y.right = element_text(color = ExpressionColor, size=13)) +
  ggtitle(paste0("dPSI and Expression for ", gene ," cryptic junction - ", junction)) 

```





##Adding in code to include cell type specific counts:
```{r}
for (ct in cluster.types) {
    
  ct.mut.cells <- meta.temp[which(meta.temp[,cluster.column] == ct),]
  intron.meta.sub.temp[,paste0(ct,".mut.bulk.counts")] <- rowSums(temp[rownames(intron.meta.sub.temp),ct.mut.cells])
  
  if (ct =="HSPC"){
    intron.meta.sub.temp <- intron.meta.sub.temp %>% group_by(clusterID_5p) %>% mutate(HSPC.cluster.cov.mut=sum(HSPC.mut.bulk.counts))
  intron.meta.sub.temp <- intron.meta.sub.temp %>% group_by(clusterID_5p) %>% mutate(HSPC.mut.psi = (HSPC.mut.bulk.counts/HSPC.cluster.cov.mut)*100)
  }
  
  else if (ct =="IMP"){
    intron.meta.sub.temp <- intron.meta.sub.temp %>% group_by(clusterID_5p) %>% mutate(IMP.cluster.cov.mut=sum(IMP.mut.bulk.counts))
  intron.meta.sub.temp <- intron.meta.sub.temp %>% group_by(clusterID_5p) %>% mutate(IMP.mut.psi = (IMP.mut.bulk.counts/IMP.cluster.cov.mut)*100)
    
  }
  
  else if (ct =="MEP"){
    intron.meta.sub.temp <- intron.meta.sub.temp %>% group_by(clusterID_5p) %>% mutate(MEP.cluster.cov.mut=sum(MEP.mut.bulk.counts))
  intron.meta.sub.temp <- intron.meta.sub.temp %>% group_by(clusterID_5p) %>% mutate(MEP.mut.psi = (MEP.mut.bulk.counts/MEP.cluster.cov.mut)*100)
    
  }
  
  else if (ct =="EP"){
    intron.meta.sub.temp <- intron.meta.sub.temp %>% group_by(clusterID_5p) %>% mutate(EP.cluster.cov.mut=sum(EP.mut.bulk.counts))
  intron.meta.sub.temp <- intron.meta.sub.temp %>% group_by(clusterID_5p) %>% mutate(EP.mut.psi = (EP.mut.bulk.counts/EP.cluster.cov.mut)*100)
    
  }
  
}
```

