

```{r}
library(pheatmap)
```

```{r}
junctions.bulk.all <- read.table("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/7.Permutation_junction_output/6.Comb_patients_2WT/5_read_threshold/MDS_P5_P6/logOR_within_cell_type_ALT_3P_Junctions_with_threshold_and_exon.skipping_info.txt")

junctions.bulk <- read.table("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/7.Permutation_junction_output/6.Comb_patients_2WT/5_read_threshold/MDS_P5_P6/logOR_within_cell_type_ONLY_CRYPTIC_3P_Junctions_with_threshold_info.txt")

junctions.bulk[which(junctions.bulk$gene =="STAU1"),]
```

```{r}
junctions.bulk.all[which(junctions.bulk.all$gene =="GYPA"),]
```



##Collection junctions of interest
```{r}
##HSPC specific cryptic junctions:
junctions.hspc <- read.table("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/7.Permutation_junction_output/7.Comb_patients_ind_celltype_merge_counts_2WT/MDS_P5_P6/HSPC/logOR_within_cell_type_ONLY_CRYPTIC_3P_Junctions_with_threshold_info.txt")

##IMP specific cryptic junctions:
junctions.imp <- read.table("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/7.Permutation_junction_output/7.Comb_patients_ind_celltype_merge_counts_2WT/MDS_P5_P6/IMP/logOR_within_cell_type_ONLY_CRYPTIC_3P_Junctions_with_threshold_info.txt")

junctions.mkp <- read.table("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/7.Permutation_junction_output/7.Comb_patients_ind_celltype_merge_counts_2WT/MDS_P5_P6/MkP/logOR_within_cell_type_ONLY_CRYPTIC_3P_Junctions_with_threshold_info.txt")

##MEP specific cryptic junctions:
junctions.mep <- read.table("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/7.Permutation_junction_output/7.Comb_patients_ind_celltype_merge_counts_2WT/MDS_P5_P6/MEP/logOR_within_cell_type_ONLY_CRYPTIC_3P_Junctions_with_threshold_info.txt")

##EP specific cryptic junctions:
junctions.ep <- read.table("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/7.Permutation_junction_output/7.Comb_patients_ind_celltype_merge_counts_2WT/MDS_P5_P6/EP/logOR_within_cell_type_ONLY_CRYPTIC_3P_Junctions_with_threshold_info.txt")

junctions.np <- read.table("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/7.Permutation_junction_output/7.Comb_patients_ind_celltype_merge_counts_2WT/MDS_P5_P6/NP/logOR_within_cell_type_ONLY_CRYPTIC_3P_Junctions_with_threshold_info.txt")


p.val.thresh <- 0.05
dPSI.val <- 2
read.thresh <- 0

junctions.hspc.sub <- junctions.hspc[which(junctions.hspc$pvalue <= p.val.thresh & junctions.hspc$dPSI >= dPSI.val &  junctions.hspc$three.mut.cluster.cov >= read.thresh & junctions.hspc$three.wt.cluster.cov >= read.thresh),]
rownames(junctions.hspc.sub) <- junctions.hspc.sub$intron_junction

junctions.imp.sub <- junctions.imp[which(junctions.imp$pvalue <= p.val.thresh & junctions.imp$dPSI >= dPSI.val & junctions.imp$three.mut.cluster.cov >= read.thresh & junctions.imp$three.wt.cluster.cov >= read.thresh),]
rownames(junctions.imp.sub) <- junctions.imp.sub$intron_junction

junctions.mkp.sub <- junctions.mkp[which(junctions.mkp$pvalue <= p.val.thresh & junctions.mkp$dPSI >= dPSI.val & junctions.mkp$three.mut.cluster.cov >= read.thresh & junctions.mkp$three.wt.cluster.cov >= read.thresh),]
rownames(junctions.mkp.sub) <- junctions.mkp.sub$intron_junction

junctions.mep.sub <- junctions.mep[which(junctions.mep$pvalue <= p.val.thresh & junctions.mep$dPSI >= dPSI.val & junctions.mep$three.mut.cluster.cov >= read.thresh & junctions.mep$three.wt.cluster.cov >= read.thresh),]
rownames(junctions.mep.sub) <- junctions.mep.sub$intron_junction

junctions.ep.sub <- junctions.ep[which(junctions.ep$pvalue <= p.val.thresh & junctions.ep$dPSI >= dPSI.val & junctions.ep$three.mut.cluster.cov >= read.thresh & junctions.ep$three.wt.cluster.cov >= read.thresh),]
rownames(junctions.ep.sub) <- junctions.ep.sub$intron_junction

junctions.np.sub <- junctions.np[which(junctions.np$pvalue <= p.val.thresh & junctions.np$dPSI >= dPSI.val & junctions.np$three.mut.cluster.cov >= read.thresh & junctions.np$three.wt.cluster.cov >= read.thresh),]
rownames(junctions.np.sub) <- junctions.np.sub$intron_junction


junctions.hspc.list <- paste(junctions.hspc.sub$intron_junction)
junctions.imp.list <- paste(junctions.imp.sub$intron_junction)
junctions.mkp.list <- paste(junctions.mkp.sub$intron_junction)
junctions.mep.list <- paste(junctions.mep.sub$intron_junction)
junctions.ep.list <- paste(junctions.ep.sub$intron_junction)
junctions.np.list <- paste(junctions.np.sub$intron_junction)


junctions.hspc.list.final <- paste0(junctions.hspc.sub$gene,":",junctions.hspc.sub$intron_junction)
junctions.imp.list.final <- paste0(junctions.imp.sub$gene,":",junctions.imp.sub$intron_junction)
junctions.mkp.list.final <- paste0(junctions.mkp.sub$gene,":",junctions.mkp.sub$intron_junction)
junctions.mep.list.final <- paste0(junctions.mep.sub$gene,":",junctions.mep.sub$intron_junction)
junctions.ep.list.final <- paste0(junctions.ep.sub$gene,":",junctions.ep.sub$intron_junction)
junctions.np.list.final <- paste0(junctions.np.sub$gene,":",junctions.np.sub$intron_junction)


length(junctions.hspc.list)
length(junctions.imp.list)
length(junctions.mkp.list)
length(junctions.mep.list)
length(junctions.ep.list)
length(junctions.np.list)


##Generating junction list Junctions significant in ery lineage cts
junctions.of.interest <- Reduce(union, list(junctions.hspc.list, junctions.mep.list, junctions.ep.list, junctions.mkp.list)) 

junctions.of.interest.final <- Reduce(union, list(junctions.hspc.list.final, junctions.mep.list.final, junctions.ep.list.final, junctions.mkp.list.final)) 

gene.list <- unique(unlist(lapply(strsplit(junctions.of.interest.final,split= ":"), "[",1)))

save(gene.list, junctions.of.interest,junctions.of.interest.final, version=2, file="/gpfs/commons/groups/landau_lab/SF3B1_splice_project/16.Pseudotime.Window/junctions.to.use.MDS.full.ct.sub.Rdata")


##Generating junction list Junctions significant in all cell type
junctions.of.interest <- Reduce(union, list(junctions.hspc.list, junctions.imp.list, junctions.mep.list, junctions.ep.list, junctions.mkp.list,junctions.np.list)) 

junctions.of.interest.final <- Reduce(union, list(junctions.hspc.list.final, junctions.imp.list.final, junctions.mep.list.final, junctions.ep.list.final, junctions.mkp.list.final, junctions.np.list.final)) 

gene.list <- unique(unlist(lapply(strsplit(junctions.of.interest.final,split= ":"), "[",1)))

#save(gene.list, junctions.of.interest,junctions.of.interest.final, version=2, file="/gpfs/commons/groups/landau_lab/SF3B1_splice_project/16.Pseudotime.Window/junctions.to.use.MDS.full.ct.all.Rdata")

save(gene.list, junctions.of.interest,junctions.of.interest.final, version=2, file="/gpfs/commons/groups/landau_lab/SF3B1_splice_project/16.Pseudotime.Window/junctions.to.use.MDS.noSigThresh.ct.sub.Rdata")

```


```{r}
junctions.np
```


```{r}
shared <- intersect(junctions.ep.list, junctions.mep.list)
length(shared)

junctions.np[which(junctions.np$intron_junction %in% shared),]
```

```{r}
RPS27A
RPL22L1
METTL5
#EIF4A2 - good example - goes from -dPSI to +dPSI
C17orf49
SLC3A2
EIF3E
SERBP1
TPT1
STAU1
```


```{r}
"SRSF11"
"STAU1"

```



```{r}
gene <- "GTF2F2"
junctions.hspc[which(junctions.hspc$gene == gene),]
junctions.imp[which(junctions.imp$gene == gene),]
junctions.mkp[which(junctions.mkp$gene == gene),]
junctions.mep[which(junctions.mep$gene == gene),]
junctions.ep[which(junctions.ep$gene == gene),]
junctions.np[which(junctions.np$gene == gene),]

FeaturePlot(mds.samples.integrated, features = gene)
```




```{r}
length(junctions.ep.list.final)
length(unique(junctions.of.interest.final))
length(intersect(junctions.of.interest.final,junctions.ep.list.final) )
```

```{r}
junctions.ep.list.final
"SF3B1:chr2:197403059:197403584:-" %in% junctions.ep.list.final
"SF3B1:chr2:197403059:197403584:-" %in% junctions.of.interest.final
```




```{r}
annotation_row = data.frame(cryptic.junctions.full=junctions.of.interest.final)

rownames(annotation_row) <- annotation_row$cryptic.junctions.full

annotation_row$EP <- 0
annotation_row$MEP <- 0
annotation_row$MkP <- 0
annotation_row$HSPC <- 0
#annotation_row$IMP <- 0
#annotation_row$NP <- 0

#annotation_row[which(annotation_row$cryptic.junctions.full %in% junctions.hspc.list.final),]$HSPC <- 1
#annotation_row[which(annotation_row$cryptic.junctions.full %in% junctions.imp.list.final),]$IMP <- 1
annotation_row[which(annotation_row$cryptic.junctions.full %in% junctions.mep.list.final),]$MEP <- 1
annotation_row[which(annotation_row$cryptic.junctions.full %in% junctions.ep.list.final),]$EP <- 1
annotation_row[which(annotation_row$cryptic.junctions.full %in% junctions.mkp.list.final),]$MkP <- 1
annotation_row[which(annotation_row$cryptic.junctions.full %in% junctions.np.list.final),]$NP <- 1

#annotation_row$cell.type.site <- factor(annotation_row$cell.type.site, c("HSPC","IMP","MEP","EP"))
annotation_row <- annotation_row[-1]
annotation_row

annotation_row$sum <- rowSums(annotation_row)

all.examples <- rownames(annotation_row[which(annotation_row$sum > 0),])
pan.examples <- rownames(annotation_row[which(annotation_row$sum >=3),])
unique.examples <- rownames(annotation_row[which(annotation_row$sum == 1),])

#ep.unique.examples <- rownames(annotation_row[which(annotation_row$sum == 1 & annotation_row$EP ==1),])

ann_colors = list(
  #NP = c("light grey","red"),
  #IMP = c("light grey","red"),
  HSPC = c("light grey","red"),
  MkP = c("light grey","red"),
  MEP = c("light grey","red"),
  EP = c("light grey","red")
)

```




```{r}
#unique.examples
#pan.examples

length(all.examples)
nrow(annotation_row)
length(unique.examples)
length(pan.examples)

```

```{r}

```


## ----------- After running junctions through pipeline ----------- ##


##Output from MDSP5+MDSP6 and window size of 100
```{r}
#Using CITE 
#load("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/16.Pseudotime.Window/MDS.output/MDSP5.MDSP6.merged.info.sig.in.any.bucket.300.usingCITE.ct.all.Rdata")
#load("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/16.Pseudotime.Window/MDS.output/MDSP5.MDSP6.merged.info.sig.in.any.bucket.400.usingCITE.ct.all.Rdata")


load("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/16.Pseudotime.Window/MDS.output/MDSP5.MDSP6.merged.info.sig.in.any.bucket.300.usingCITE.ct.sub.Rdata")

load("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/16.Pseudotime.Window/MDS.output/MDSP5.MDSP6.merged.info.sig.in.any.bucket.400.usingCITE.ct.sub.Rdata")


load("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/16.Pseudotime.Window/MDS.output/MDSP5.MDSP6.merged.info.no.sig.thresh.bucket.300.usingCITE.ct.sub.Rdata")


rownames <- junctions.of.interest.final

extracted_cols <- lapply(win.dPSI, function(x) x[,"dPSI"]) #note the added comma!
dPSI.per.window <- as.data.frame(do.call(cbind, extracted_cols))
colnames(dPSI.per.window) = as.character(1:ncol(dPSI.per.window))
rownames(dPSI.per.window) <- rownames

extracted_cols <- lapply(win.dPSI, function(x) x[,"mut.psi"]) #note the added comma!
mut.psi.per.window <- as.data.frame(do.call(cbind, extracted_cols))
colnames(mut.psi.per.window) = as.character(1:ncol(mut.psi.per.window))
rownames(mut.psi.per.window) <- rownames

extracted_cols <- lapply(win.dPSI, function(x) x[,"wt.psi"]) #note the added comma!
wt.psi.per.window <- as.data.frame(do.call(cbind, extracted_cols))
colnames(wt.psi.per.window) = as.character(1:ncol(wt.psi.per.window))
rownames(wt.psi.per.window) <- rownames

#dPSI.per.window[which(rowMeans(is.na(dPSI.per.window)) > 0.1), ] #(removed multiple NA's)
#dPSI.per.window[which(rowMeans(is.na(dPSI.per.window)) <= 0.1), ]

dPSI.per.window.sub <- dPSI.per.window[which(rowMeans(is.na(dPSI.per.window)) <= 0.1), ]
dPSI.per.window.sub.zscores <- as.data.frame(t(scale(t(dPSI.per.window.sub))))
#dPSI.per.window.sub.zscores

#mask <- apply(dPSI.per.window.sub, 2, is.nan)
#dPSI.per.window.sub.temp <- dPSI.per.window.sub
#dPSI.per.window.sub.temp[mask] <- 0
#dPSI.per.window.sub.zscores <- as.data.frame(t(apply(dPSI.per.window.sub.temp, 1, function(x) (x-mean(x))/sd(x))))

```


```{r}


#write.table(dPSI.per.window, file="/gpfs/commons/groups/landau_lab/SF3B1_splice_project/16.Pseudotime.Window/psi.matricies/dPSIperCD71bin.crypticsite.significant.in.atleast.1ct.HSPC.MkP.MEP.EP.binsize3000cells.txt", quote = F)

write.table(dPSI.per.window, file="/gpfs/commons/groups/landau_lab/SF3B1_splice_project/16.Pseudotime.Window/psi.matricies/dPSIperCD71bin.crypticsite.no.sig.thresh.in.cts.HSPC.MkP.MEP.EP.binsize3000cells.txt", quote = F)


```


## ------------------------------------------ Generating heatmaps ------------------------------------------ ##

##Adding Annotation Colors 
```{r fig.height=1.5, fig.width=10}

#annotation_col = data.frame(ep.fraction = as.numeric(paste(ct.fractions.2["EP",])), mep.fraction = as.numeric(paste(ct.fractions.2["MEP",])), imp.fraction = as.numeric(paste(ct.fractions.2["IMP",])) ,hsc.fraction = as.numeric(paste(ct.fractions.2["HSPC",])),  mean.pseudotime = pseudotime.mean, mut.frac= mut.fraction)

#Only using some celltypes
annotation_col = data.frame(ep.fraction = as.numeric(paste(ct.fractions.2["EP",])), 
                            mep.fraction = as.numeric(paste(ct.fractions.2["MEP",])),
                            mkp.fraction = as.numeric(paste(ct.fractions.2["MkP",])),
                            hsc.fraction = as.numeric(paste(ct.fractions.2["HSPC",])),
                            mean.CD71.exp = cite.expression.mean, 
                            mut.frac= mut.fraction)

#annotation_col = data.frame(ep.fraction = as.numeric(paste(ct.fractions.2["EP",])), 
#                            mep.fraction = as.numeric(paste(ct.fractions.2["MEP",])), 
#                            mkp.fraction = as.numeric(paste(ct.fractions.2["MkP",])),
#                            hsc.fraction = as.numeric(paste(ct.fractions.2["HSPC",])), 
#                            imp.fraction = as.numeric(paste(ct.fractions.2["IMP",])),
#                            np.fraction = as.numeric(paste(ct.fractions.2["NP",])), 
#                            mono.fraction = as.numeric(paste(ct.fractions.2["Mat_Mono",])),
#                            mean.CD71.exp = cite.expression.mean, 
#                            mut.frac= mut.fraction)

annotation_col

rownames(annotation_col) = paste(1:nrow(annotation_col))

cell.freq <- melt(t(annotation_col[,1:4]),var=c("CellType",'window'))
mycolors <- c("#980043","#4B0082","#DD3497","#313695")


#cell.freq <- melt(t(annotation_col[,1:7]),var=c("CellType",'window'))
#mycolors <- c("#980043","#4B0082","#DD3497","#313695","#045A8D","#006D2C","lightskyblue4")


pdf("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/16.Pseudotime.Window/MDS.output/pdfs/ct.sub/heatmap/perc.ct.per.bin.pdf", width = 20, height = 2)
ggplot(cell.freq, aes(x = window, y = value, fill = CellType)) + geom_col(position = "fill") + scale_fill_manual(values=mycolors) + theme_void() + theme(legend.key.size = unit(0.5, 'cm'), #change legend key size
        legend.key.height = unit(0.5, 'cm'), #change legend key height
        legend.key.width = unit(0.5, 'cm'), #change legend key width
        legend.title = element_text(size=10), #change legend title font size
        legend.text = element_text(size=8)) #change legend text font size
dev.off()

```



```{r}
"SF3B1:chr2:197403059:197403584:-" %in% rownames(dPSI.per.window.sub)
"SF3B1:chr2:197403059:197403584:-" %in% rownames(dPSI.per.window.sub.zscores)
```

```{r}
to.plot.reordered
```

```{r}
#fig.height=20, fig.width=25
```


```{r }
#shared.junctions <- intersect(junctions.of.interest.final, rownames(dPSI.per.window.sub))

##Ordering it by z-score
to.plot <- dPSI.per.window.sub.zscores
to.plot <- dPSI.per.window.sub.zscores[rownames(dPSI.per.window.sub.zscores) %in% all.examples,]
to.plot <- dPSI.per.window.sub.zscores[rownames(dPSI.per.window.sub.zscores) %in% unique.examples,]

ncols <- ncol(to.plot)
to.plot$max.column <- colnames(to.plot)[apply(to.plot,1,which.max)]
to.plot$max.column <- as.numeric(to.plot$max.column)
to.plot$rownames <- rownames(to.plot)
to.plot.reordered <- to.plot %>% arrange(max.column)
rownames(to.plot.reordered) <- to.plot.reordered$rownames
to.plot.reordered <- to.plot.reordered[,1:ncols]
to.plot.reordered

paletteLength <- 50
myColor <- colorRampPalette(c("navy", "white", "red"))(paletteLength)
# length(breaks) == length(paletteLength) + 1
# use floor and ceiling to deal with even/odd length pallettelengths
myBreaks <- c(seq(min(to.plot.reordered[!is.na(to.plot.reordered)]), 0, length.out=ceiling(paletteLength/2) + 1), 
              seq(max(to.plot.reordered[!is.na(to.plot.reordered)])/paletteLength, max(to.plot.reordered[!is.na(to.plot.reordered)]), length.out=floor(paletteLength/2)))


#pheatmap(to.plot.reordered, cluster_cols = F, cluster_rows = F, annotation_row = annotation_row[,-5] , annotation_col = annotation_col, annotation_colors = ann_colors, color=myColor, border_color=NA, annotation_legend=F, breaks=myBreaks)

#pheatmap(to.plot.reordered, cluster_cols = F, cluster_rows = F, annotation_row = annotation_row[rownames(to.plot.reordered),-5] , annotation_col = annotation_col[,8:9], annotation_colors = ann_colors, color=myColor, border_color=NA, annotation_legend=F, breaks=myBreaks)

pheatmap(to.plot.reordered, cluster_cols = F, cluster_rows = F, annotation_row = annotation_row[rownames(to.plot.reordered),-4], annotation_colors = ann_colors, color=myColor, border_color=NA, annotation_legend=F, breaks=myBreaks)



pdf("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/16.Pseudotime.Window/MDS.output/pdfs/ct.sub/heatmap/MDS.combined.heatmap.pdf", width = 20, height = 30)
pheatmap(to.plot.reordered, cluster_cols = F, cluster_rows = F, color=myColor, border_color=NA, annotation_legend=F, breaks=myBreaks)
dev.off()
```

```{r}
annotation_col
```


```{r fig.height=0.8, fig.width=25}

pdf("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/16.Pseudotime.Window/MDS.output/pdfs/ct.sub/heatmap/mean.CD71.expression.per.bin.pdf", width = 20, height = 0.8)
pheatmap(t(annotation_col[,5]), cluster_cols = F, cluster_rows = F, color=colorRampPalette(c("white","dark green" ))(50),legend=F)

#pheatmap(t(annotation_col[,8]), cluster_cols = F, cluster_rows = F, color=colorRampPalette(c("white","dark green" ))(50),legend=F)
dev.off()


pdf("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/16.Pseudotime.Window/MDS.output/pdfs/ct.sub/heatmap/mut.frac.per.bin.pdf", width = 20, height = 0.8)
pheatmap(t(annotation_col[,6]), cluster_cols = F, cluster_rows = F, color=colorRampPalette(c("white", "#9370DB"))(50), legend=F)
#pheatmap(t(annotation_col[,9]), cluster_cols = F, cluster_rows = F, color=colorRampPalette(c("white", "#9370DB"))(50), legend=F)
dev.off()

```

## ------------------------------------------ Looking at Individual Events ------------------------------------------ ##

```{r}

extracted_cols <- lapply(win.dPSI, function(x) x[,"cluster.cov.mut"]) #note the added comma!
cluster.cov.mut.per.window <- as.data.frame(do.call(cbind, extracted_cols))
colnames(cluster.cov.mut.per.window) = as.character(1:ncol(cluster.cov.mut.per.window))
rownames(cluster.cov.mut.per.window) <- rownames

extracted_cols <- lapply(win.dPSI, function(x) x[,"cluster.cov.wt"]) #note the added comma!
cluster.cov.wt.per.window <- as.data.frame(do.call(cbind, extracted_cols))
colnames(cluster.cov.wt.per.window) = as.character(1:ncol(cluster.cov.wt.per.window))
rownames(cluster.cov.wt.per.window) <- rownames

## Datatables to use
#dPSI.per.window
#cluster.cov.mut.per.window
#cluster.cov.wt.per.window

##If the cluster coverage is 0 in either wt or mut, remove that junction from the list 

```

##Candidates for individual events:

```{r}
to.plot.reordered
```




```{r}
junction <- "chr4:144116976:144119685:-" 
gene <- "GYPA"

junction <- "chr19:55455378:55455635:-" 
gene <- "ISOC2"

junction <- "chr20:49124605:49135832:-" 
gene <- "STAU1"


  
```

```{r}
junction <- "chr17:58005617:58005800:-" 
gene <- "SRSF1"

junction <- "chr7:80661210:80662972:+" 
gene <- "CD36"

junction <- "chr7:80672040:80672754:+" 
gene <- "CD36"

junction <- "chr11:57426389:57426587:-" 
gene <- "SLC43A3"

junction <- "chr2:85621579:85623625:+" 
gene <- "USP39"

junction <- "chr14:77737095:77738777:-" 
gene <- "SNW1"

junction <- "chr1:70246907:70249930:+" 
gene <- "SRSF11"



```




```{r}
junction <- "chrX:48793297:48793777:+" 
gene <- "GATA1"

junction <- "chr17:44322552:44322807:-" 
gene <- "SLC25A39"

#Not well covered
junction <- "chr14:22768171:22769776:+" 
gene <- "OXA1L"

#Well covered
junction <- "chr3:186783639:186784428:+" 
gene <- "EIF4A2"

#junction <- "chr10:119582608:119588151:-"
#gene <- "TIAL1"

junction <- "chr4:139072039:139073453:-"
gene <- "ELF2"

junction <- "chr20:35556271:35556954:+" 
gene <- "ERGIC3"

#junction <- "chr9:136791424:136791625:+" 
#gene <- "TMEM141"

junction <- "chr1:100992754:100995109:-" 
gene <- "DPH5"

junction <- "chr3:170868201:170868297:-" 
gene <- "RPL22L1"

junction <- "chr11:64232614:64232700:+" 
gene <- "DNAJC4"




junction <- "chr1:45012285:45012856:+" 
gene <- "UROD"

junction <- "chr19:48960914:48961482:+" 
gene <- "BAX"

junction <- "chr1:161167486:161168412:+" 
gene <- "PPOX"

junction <- "chr2:197403059:197403584:-" 
gene <- "SF3B1"


junction <- "chr17:7017014:7017256:+" 
gene <- "C17orf49"

junction <- "chr1:67424322:67424887:-" 
gene <- "SERBP1"


junction <- "chr3:129171280:129171445:-" 
gene <- "CNBP"


junction <- "chr11:8683265:8683989:+" 
gene <- "RPL27A"


####VERY GOOD GLOBAL EXAMPLE
junction <- "chr14:20990598:20992091:+" 
gene <- "METTL17"
  

junction <- "chr1:100992754:100995109:-" 
gene <- "DPH5"

junction <- "chr2:169812524:169815476:-" 
gene <- "METTL5"

junction <- "chr1:67424977:67425082:-" 
gene <- "SERBP1"

junction <- "chr13:95725252:95757624:+" 
gene <- "DNAJC3"


junction <- "chr20:35556271:35556954:+" 
gene <- "ERGIC3"

```

```{r}
gene
junction
```


```{r}



junction <- "chrX:48793297:48793777:+" 
gene <- "GATA1"

junction <- "chr14:74889349:74889877:+" 
gene <- "DLST"



junction.final %in% rownames(to.plot.reordered)


##Genes to look at METTL5, SERBP1, ERGIC3, DPH5, GATA1 , (DLST?)
```

```{r}

junction.final <- paste0(gene,":",junction)

#dPSI info - out[junction.final,]
#ONT gene.expression info - gene.exp.matrix[gene,]

colnames <- c("window","dPSI","mut.psi","wt.psi","ONT.Expression","ONT.MUT.Expression","ONT.WT.Expression")
psi.per.window.data <- as.data.frame(matrix(nrow=ncol(dPSI.per.window), ncol=length(colnames)))
rownames(psi.per.window.data) <- as.character(1:ncol(dPSI.per.window))
colnames(psi.per.window.data) <- colnames

psi.per.window.data$window <- rownames(psi.per.window.data)
psi.per.window.data$dPSI <- as.numeric(paste(dPSI.per.window[junction.final,]))
psi.per.window.data$mut.psi <- as.numeric(paste(mut.psi.per.window[junction.final,]))
psi.per.window.data$wt.psi <- as.numeric(paste(wt.psi.per.window[junction.final,]))
psi.per.window.data$ONT.Expression <- as.numeric(paste(ont.expression.per.window[gene,]))
psi.per.window.data$ONT.MUT.Expression <- as.numeric(paste(ont.mut.expression.per.window[gene,]))
psi.per.window.data$ONT.WT.Expression <- as.numeric(paste(ont.wt.expression.per.window[gene,]))
psi.per.window.data$window <- factor(psi.per.window.data$window,levels = 1:nrow(psi.per.window.data))
psi.per.window.data
mask <- apply(psi.per.window.data, 2, is.na)
psi.per.window.data[mask] <- 0
psi.per.window.data
```


```{r }
ExpressionColor <- "red"
dPSIColor <- "blue"

OldMax = max(psi.per.window.data$ONT.Expression)
OldMin = min(psi.per.window.data$ONT.Expression)
OldRange = (OldMax - OldMin)


NewMax = max(psi.per.window.data$dPSI)
NewMin = min(psi.per.window.data$dPSI)
NewRange = (NewMax - NewMin)

ggplot(psi.per.window.data, aes(x=window, group = 1)) + 
  geom_line( aes(y=dPSI), color=dPSIColor) + 
  geom_line( aes(y=(((ONT.Expression - OldMin) * NewRange) / OldRange) + NewMin), linetype = "dashed", color=ExpressionColor)+  
  geom_hline(yintercept = 0, linetype = "dotted") + 
  scale_y_continuous(
    name = "dPSI",
    sec.axis = sec_axis(~(((.-NewMin)*OldRange)/NewRange) + OldMin , name="ONT Expression \n (average of normalized counts)")) +
  theme_classic()+
  theme(plot.title = element_text(hjust = 0.5, size = 14), 
        axis.title.y = element_text(color =dPSIColor , size=13), 
        axis.title.y.right = element_text(color = ExpressionColor, size=13)) +
  ggtitle(paste0("dPSI and Expression for ", gene ," cryptic junction - ", junction)) 


library(grid)
library(gridExtra)

grob = grobTree(textGrob(paste("Spearman Correlation : ", round(cor(psi.per.window.data$dPSI, psi.per.window.data$ONT.Expression, method = c("spearman") ), 4)), x=0.5, y = 0.97 ,gp = gpar(col = "red", fontsize = 11)))

final.plot <- ggplot(psi.per.window.data, aes(x=window, group = 1)) + 
  geom_point(aes(y= dPSI), color=dPSIColor) +
  stat_smooth(aes(y= dPSI), color=dPSIColor, method = "loess", formula = y ~ x, se = F, span=0.5) +
  geom_point(aes(y =(((ONT.Expression - OldMin) * NewRange) / OldRange) + NewMin), color=ExpressionColor) +
  stat_smooth(aes(y =(((ONT.Expression - OldMin) * NewRange) / OldRange) + NewMin), linetype = "dashed", color=ExpressionColor, method = "loess", formula = y ~ x, se = F, span=0.5)+
  geom_hline(yintercept = 0, linetype = "dotted") + 
  scale_y_continuous(
    name = "dPSI",
    sec.axis = sec_axis(~(((.-NewMin)*OldRange)/NewRange) + OldMin , name="ONT Expression \n (average of normalized counts)")) +
  annotation_custom(grob)+
  theme_classic()+
  theme(plot.title = element_text(hjust = 0.5, size = 14), 
        axis.title.y = element_text(color =dPSIColor , size=13), 
        axis.title.y.right = element_text(color = ExpressionColor, size=13)) +
  ggtitle(paste0("dPSI and Expression for ", gene ," cryptic junction - ", junction)) 

final.plot

```



```{r}
psi.per.window.data
```






```{r}
#mds.samples.integrated@meta.data

FeaturePlot(mds.samples.integrated, features = gene)
#FeaturePlot(mds.samples.integrated, features = "SERBP1")
#FeaturePlot(mds.samples.integrated, features = "ADT-CD71")


```


```{r fig.height=1, fig.width=8}
extracted_cols <- lapply(win.dPSI, function(x) x[,"mut.bulk.counts"]) 
mut.counts.per.window <- as.data.frame(do.call(cbind, extracted_cols))
colnames(mut.counts.per.window) = as.character(1:ncol(mut.counts.per.window))
rownames(mut.counts.per.window) <- rownames

extracted_cols <- lapply(win.dPSI, function(x) x[,"cluster.cov.mut"]) #note the added comma!
cluster.cov.mut.per.window <- as.data.frame(do.call(cbind, extracted_cols))
colnames(cluster.cov.mut.per.window) = as.character(1:ncol(cluster.cov.mut.per.window))
rownames(cluster.cov.mut.per.window) <- rownames

extracted_cols <- lapply(win.dPSI, function(x) x[,"wt.bulk.counts"]) 
wt.counts.per.window <- as.data.frame(do.call(cbind, extracted_cols))
colnames(wt.counts.per.window) = as.character(1:ncol(wt.counts.per.window))
rownames(wt.counts.per.window) <- rownames

extracted_cols <- lapply(win.dPSI, function(x) x[,"cluster.cov.wt"]) #note the added comma!
cluster.cov.wt.per.window <- as.data.frame(do.call(cbind, extracted_cols))
colnames(cluster.cov.wt.per.window) = as.character(1:ncol(cluster.cov.wt.per.window))
rownames(cluster.cov.wt.per.window) <- rownames

raw.counts <- rbind(mut.counts.per.window[junction.final,], cluster.cov.mut.per.window[junction.final,] , wt.counts.per.window[junction.final,] , cluster.cov.wt.per.window[junction.final,])

raw.counts <- as.data.frame(sapply(raw.counts, as.integer))
rownames(raw.counts) <- c("cryptic.reads.mut", "cluster.cov.mut","cryptic.reads.wt","cluster.cov.wt")

pheatmap(raw.counts, cluster_cols = F, cluster_rows = F, legend=F, display_numbers = T, number_format= "%.f",number_color = "blue",color=colorRampPalette(c("white", "white"))(50), show_rownames = T, show_colnames = F, gaps_row = c(2))


psi.per.window.data$p.val <- NA

for (window in 1:ncol(raw.counts)){
  contingency.table <- cbind(raw.counts[1:2,window], raw.counts[3:4,window])
  contingency.table[2,] <- contingency.table[2,] - contingency.table[1,]
  result <- fisher.test(contingency.table)
  psi.per.window.data[window,]$p.val <- result$p.value
}


#psi.per.window.data
#psi.per.window.data$window <- as.numeric(psi.per.window.data$window)
#psi.per.window.data$start <- psi.per.window.data$window - 0.5
#psi.per.window.data$end <- psi.per.window.data$window + 0.5

```


```{r}
psi.per.window.data$p.val.label <- ""
psi.per.window.data[which(psi.per.window.data$p.val <= 0.05 ),]$p.val.label <- "*"
psi.per.window.data[which(psi.per.window.data$p.val <= 0.01 ),]$p.val.label <- "**"
psi.per.window.data[which(psi.per.window.data$p.val <= 0.001 ),]$p.val.label <- "***"
psi.per.window.data
```


```{r}
ExpressionColor <- "red"
dPSIColor <- "steelblue"

ggplot(psi.per.window.data, aes(x=window, y=dPSI ,  group = 1)) + 
  geom_bar(stat="identity", fill=dPSIColor) +
  geom_text(aes(label=p.val.label), vjust=-0.3, size=4 ) +
  #geom_text(aes(data=subset(psi.per.window.data, dPSI >= 0 ) , label=p.val.label), vjust=-0.5, size=3.5 ) +
  geom_point(aes(y =(((ONT.Expression - OldMin) * NewRange) / OldRange) + NewMin), color=ExpressionColor) +
  stat_smooth(aes(y =(((ONT.Expression - OldMin) * NewRange) / OldRange) + NewMin), linetype = "dashed", color=ExpressionColor, method = "loess", formula = y ~ x, se = F, span=0.5)+
  geom_hline(yintercept = 0, linetype = "dotted") + 
  scale_y_continuous(
    name = "dPSI",
    sec.axis = sec_axis(~(((.-NewMin)*OldRange)/NewRange) + OldMin , name="ONT Expression \n (average of normalized counts)")) +
  #annotation_custom(grob)+
  theme_classic()+
  labs(x="bin",
       title = paste0("dPSI and Expression for ", gene ," cryptic junction - ", junction),
       subtitle = paste0("Spearman Correlation : ", round(cor(psi.per.window.data$dPSI, psi.per.window.data$ONT.Expression, method = c("spearman") ), 4))) +
  theme(plot.title = element_text(vjust = 2.5,hjust = 0.5, size = 14), 
        axis.title.y = element_text(color =dPSIColor , size=13), 
        axis.title.y.right = element_text(color = ExpressionColor, size=13),
        plot.subtitle=element_text(size=10, hjust=0.5, face="italic", color="red")) 
```


```{r}

ExpressionColor <- "red"
dPSIColor <- "steelblue"

pdf(paste0("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/16.Pseudotime.Window/MDS.output/pdfs/ct.sub/individual.events/",gene,".dPSI.vs.Expression.pdf"), width = 12, height = 8)

ggplot(psi.per.window.data, aes(x=window, y=dPSI ,  group = 1)) + 
  geom_bar(stat="identity", fill=dPSIColor) +
  geom_text(aes(label=p.val.label), vjust=-0.2, size=8) +
  #geom_text(aes(data=subset(psi.per.window.data, dPSI >= 0 ) , label=p.val.label), vjust=-0.5, size=3.5 ) +
  geom_point(aes(y =(((ONT.Expression - OldMin) * NewRange) / OldRange) + NewMin), color=ExpressionColor) +
  stat_smooth(aes(y =(((ONT.Expression - OldMin) * NewRange) / OldRange) + NewMin), linetype = "dashed", color=ExpressionColor, method = "loess", formula = y ~ x, se = F, span=0.5)+
  geom_hline(yintercept = 0, linetype = "dotted") + 
  scale_y_continuous(
    name = "dPSI",
    sec.axis = sec_axis(~(((.-NewMin)*OldRange)/NewRange) + OldMin , name="ONT Expression \n (average of normalized counts)")) +
  #annotation_custom(grob)+
  theme_classic()+
  labs(x="bin",
       title = paste0("dPSI and Expression for ", gene ," cryptic junction - ", junction),
       subtitle = paste0("Spearman Correlation : ", round(cor(psi.per.window.data$dPSI, psi.per.window.data$ONT.Expression, method = c("spearman") ), 4))) +
  theme(plot.title = element_text(vjust = 2.5,hjust = 0.5, size = 17), 
        axis.title.y = element_text(color =dPSIColor , size=15), 
        axis.title.y.right = element_text(color = ExpressionColor, size=15),
        plot.subtitle=element_text(size=12, hjust=0.5, face="italic", color="red")) 

dev.off()
  
```


```{r fig.height=1, fig.width=15}
paletteLength <- 50
myColor <- colorRampPalette(c("navy", "white", "red"))(paletteLength)
# length(breaks) == length(paletteLength) + 1
# use floor and ceiling to deal with even/odd length pallettelengths
myBreaks <- c(seq(min(psi.per.window.data$p.val), 0.05, length.out=ceiling(paletteLength/2) + 1), 
              seq(max(psi.per.window.data$p.val)/paletteLength, max(psi.per.window.data$p.val), length.out=floor(paletteLength/2)))

pheatmap(t(psi.per.window.data$p.val), cluster_cols = F, cluster_rows = F,  breaks=myBreaks, display_numbers = T,fontsize_number=12,  number_format= "%.3f",number_color = "black")
```


##Work on bars for dSI values with expression behind??

```{r}

```


##Doing one plot with mut exp + mut psi & wt exp + wt psi

```{r fig.height=5, fig.width=9}

OldMax = max(psi.per.window.data$ONT.MUT.Expression, psi.per.window.data$ONT.WT.Expression)
OldMin = min(psi.per.window.data$ONT.MUT.Expression, psi.per.window.data$ONT.WT.Expression)
OldRange = (OldMax - OldMin)

NewMax = max(psi.per.window.data$mut.psi, psi.per.window.data$wt.psi)
NewMin = min(psi.per.window.data$mut.psi, psi.per.window.data$wt.psi)
NewRange = (NewMax - NewMin)

ggplot(psi.per.window.data, aes(x=window, group = 1)) + 
  geom_line(aes(y=wt.psi, linetype = "WT"), color="blue") + 
  geom_line(aes(y=mut.psi, linetype = "MUT"), color="blue") + 
  #geom_hline(yintercept = 0, linetype = "dotted") + 
  geom_line(aes(y=(((ONT.WT.Expression - OldMin) * NewRange) / OldRange) + NewMin, linetype = "WT"),color="red") +
  geom_point(aes(y=(((ONT.WT.Expression - OldMin) * NewRange) / OldRange) + NewMin),color="red") +
  geom_line(aes(y=(((ONT.MUT.Expression - OldMin) * NewRange) / OldRange) + NewMin, linetype = "MUT"),color="red") +
  geom_point(aes(y=(((ONT.MUT.Expression - OldMin) * NewRange) / OldRange) + NewMin),color="red") +
  scale_linetype_manual(name = "Genotype",
                     breaks = c("WT","MUT"),
                     values = c("WT" = "solid", "MUT" = "dashed") )+
  scale_y_continuous(
    name = "PSI",
    sec.axis = sec_axis(~(((.-NewMin)*OldRange)/NewRange) + OldMin , name="ONT Expression \n (average of normalized counts)")) +
  theme_classic() +
  theme(plot.title = element_text(hjust = 0.5, size = 14), 
        axis.title.y = element_text(color =dPSIColor , size=13), 
        axis.title.y.right = element_text(color = ExpressionColor, size=13))+
  ggtitle(paste0("PSI and Expression values for ", gene ," cryptic junction - ", junction)) 
```


##looking only at mut.psi and mut.exp

```{r}

ExpressionColor <- "red"
dPSIColor <- "steelblue"

OldMax = max(psi.per.window.data$ONT.MUT.Expression)
OldMin = min(psi.per.window.data$ONT.MUT.Expression)
OldRange = (OldMax - OldMin)

NewMax = max(psi.per.window.data$mut.psi)
NewMin = min(psi.per.window.data$mut.psi)
NewRange = (NewMax - NewMin)

ggplot(psi.per.window.data, aes(x=window, group = 1)) + 
  geom_line(aes(y=mut.psi), color="blue") + 
  geom_point(aes(y=mut.psi), color="blue") +
  geom_hline(yintercept = 0, linetype = "dotted") + 
  geom_line(aes(y=(((ONT.MUT.Expression - OldMin) * NewRange) / OldRange) + NewMin), linetype = "dashed" , color="red") +
  geom_point(aes(y=(((ONT.MUT.Expression - OldMin) * NewRange) / OldRange) + NewMin),color="red") +
  scale_y_continuous(
    name = "PSI",
    sec.axis = sec_axis(~(((.-NewMin)*OldRange)/NewRange) + OldMin , name="ONT Expression of MUT cells \n (average of normalized counts)")) +
  theme_classic() +
  theme(plot.title = element_text(hjust = 0.5, size = 14), 
        axis.title.y = element_text(color =dPSIColor , size=13), 
        axis.title.y.right = element_text(color = ExpressionColor, size=13))+
  ggtitle(paste0("mut.psi vs mut.exp for ", gene ," cryptic junction - ", junction)) 
```



```{r}
annotation_row[junction.final,]
```



```{r}
contingency.table <- cbind(c(118,1350), c(8,98))
contingency.table[2,] <- contingency.table[2,] - contingency.table[1,]
result <- fisher.test(contingency.table)
contingency.table
result$p.value



contingency.table <- cbind(c(16,457), c(0,35))
contingency.table[2,] <- contingency.table[2,] - contingency.table[1,]
result <- fisher.test(contingency.table)
contingency.table
result$p.value
```



##Why is the MUT expression higher in certain windows

```{r}

thresh <- 0
MUT.info <- as.data.frame(matrix(nrow=length(win.dPSI), ncol=9))
rownames(MUT.info) <- 1:length(win.dPSI)
colnames(MUT.info) <- c("total.raw.counts","mean.exp","total.num.cells","num_norm_>_0","pct_with_gene","num_reads_<_thresh","num_reads_>=_thresh", "num_norm_>_0_reads_<_thresh","pct_reads_<_thresh")

for (i in 1:length(win.dPSI)){
  window.info <- ont.raw.counts.per.cell.per.window.per.gene[[gene]][[i]]
  MUT.info[i,"total.raw.counts"] <- sum(window.info[which(window.info$genotype =="MUT" & window.info$total >= thresh),]$gene.counts)
  MUT.info[i,"mean.exp"] <- mean(window.info[which(window.info$genotype =="MUT"),]$norm.counts)
  MUT.info[i,"total.num.cells"] <- nrow(window.info[which(window.info$genotype =="MUT"),])
  MUT.info[i,"num_norm_>_0"] <- nrow(window.info[which(window.info$genotype =="MUT" & window.info$norm.counts > 0),])
  MUT.info[i,"pct_with_gene"] <- (MUT.info[i,"num_norm_>_0"]/MUT.info[i,"total.num.cells"])*100
  MUT.info[i,"num_reads_<_thresh"] <- nrow(window.info[which(window.info$genotype =="MUT" & window.info$total < thresh),])
  MUT.info[i,"num_reads_>=_thresh"] <- nrow(window.info[which(window.info$genotype =="MUT" & window.info$total >= thresh),])
  MUT.info[i,"num_norm_>_0_reads_<_thresh"] <- nrow(window.info[which(window.info$genotype =="MUT" & window.info$norm.counts > 0 & window.info$total < thresh),])
  MUT.info[i,"pct_reads_<_thresh"] <- MUT.info[i,"num_reads_<_thresh"]/MUT.info[i,"total.num.cells"] *100
  
}

MUT.info[,c("total.raw.counts", "total.num.cells","pct_with_gene","num_reads_>=_thresh", "pct_reads_<_thresh")]

```


```{r}
WT.info <- as.data.frame(matrix(nrow=length(win.dPSI), ncol=9))
rownames(WT.info) <- 1:length(win.dPSI)
colnames(WT.info) <- c("total.raw.counts","mean.exp","total.num.cells","num_norm_>_0","pct_with_gene","num_reads_<_thresh","num_reads_>=_thresh", "num_norm_>_0_reads_<_thresh", "pct_reads_<_thresh")

for (i in 1:length(win.dPSI)){
  window.info <- ont.raw.counts.per.cell.per.window.per.gene[[gene]][[i]]
  WT.info[i,"total.raw.counts"] <- sum(window.info[which(window.info$genotype =="WT" & window.info$total >= thresh),]$gene.counts)
  WT.info[i,"mean.exp"] <- mean(window.info[which(window.info$genotype =="WT"),]$norm.counts)
  WT.info[i,"total.num.cells"] <- nrow(window.info[which(window.info$genotype =="WT"),])
  WT.info[i,"num_norm_>_0"] <- nrow(window.info[which(window.info$genotype =="WT" & window.info$norm.counts > 0),])
  WT.info[i,"pct_with_gene"] <- (WT.info[i,"num_norm_>_0"]/WT.info[i,"total.num.cells"])*100
  WT.info[i,"num_reads_<_thresh"] <- nrow(window.info[which(window.info$genotype =="WT" & window.info$total < thresh),])
  WT.info[i,"num_reads_>=_thresh"] <- nrow(window.info[which(window.info$genotype =="WT" & window.info$total >= thresh),])
  WT.info[i,"num_norm_>_0_reads_<_thresh"] <- nrow(window.info[which(window.info$genotype =="WT" & window.info$norm.counts > 0 & window.info$total < thresh),])
  WT.info[i,"pct_reads_<_thresh"] <- WT.info[i,"num_reads_<_thresh"]/WT.info[i,"total.num.cells"]*100
}

WT.info[,c("total.raw.counts", "total.num.cells","pct_with_gene","num_reads_>=_thresh", "pct_reads_<_thresh")]


```



```{r fig.height=1, fig.width=8}
num.cells.per.window <- rbind(WT.info$`num_reads_>=_thresh` ,MUT.info$`num_reads_>=_thresh`)

rownames(num.cells.per.window) <- c("num.wt.cells","num.mut.cells")
num.cells.per.window

pheatmap(num.cells.per.window, cluster_cols = F, cluster_rows = F, legend=F, display_numbers = T, number_format= "%.f",number_color = "dark green",color=colorRampPalette(c("white", "white"))(50), show_rownames = T, show_colnames = F, gaps_row = c(1))
```





```{r}
contingency.table <- cbind(c(2, 50) ,c(0,50) )
contingency.table
fisher.test(contingency.table)
```

```{r fig.height=1, fig.width=8}
raw.exp.ont.counts <- rbind(MUT.info$total.raw.counts, WT.info$total.raw.counts)
rownames(raw.exp.ont.counts) <- c("raw.count.mut","raw.count.wt")
raw.exp.ont.counts

pheatmap(raw.exp.ont.counts, cluster_cols = F, cluster_rows = F, legend=F, display_numbers = T, number_format= "%.f",number_color = "red",color=colorRampPalette(c("white", "white"))(50), show_rownames = T, show_colnames = F, gaps_row = c(1))
```
