
##Looking at overlapp between our Genes and other Bulk MDS studies 

##Load in the reference list of genes from a bulk study 
```{r fig.height=5, fig.width=5}
pallegatti.gene.reference <- read.csv("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/11.Alt3p_splicing_references/Pallegatti_Suppl3_SF3B1.csv")

pallegatti.alt3p.gene.reference.cryptic <- unique(pallegatti.gene.reference[which(pallegatti.gene.reference$Event.type == "A3SS"),]$Gene)

omar.brd9.alt3p.gene.reference <- read.table("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/11.Alt3p_splicing_references/Omar.BRD9.paper.final.list.a3ss.txt")

omar.brd9.alt3p.gene.reference.crytpic <- unique(omar.brd9.alt3p.gene.reference$V1)

dolatshad.alt3p.gene.reference <- read.csv("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/11.Alt3p_splicing_references/Dolatshad.csv")

pallegatti.alt3p.gene.reference.cryptic
omar.brd9.alt3p.gene.reference.crytpic
intersect(pallegatti.alt3p.gene.reference.cryptic,omar.brd9.alt3p.gene.reference.crytpic)

alt3p.reference.union.gene.list.cryptic <- omar.brd9.alt3p.gene.reference.crytpic
#alt3p.reference.union.gene.list.cryptic <- union(pallegatti.alt3p.gene.reference.cryptic,omar.brd9.alt3p.gene.reference.crytpic)

length(alt3p.reference.union.gene.list.cryptic)

```




```{r fig.height=5, fig.width=5}
library(VennDiagram)
area1 = length(pallegatti.alt3p.gene.reference.cryptic)
area2 = length(omar.brd9.alt3p.gene.reference.crytpic)
cross.area = length(intersect(pallegatti.alt3p.gene.reference.cryptic,omar.brd9.alt3p.gene.reference.crytpic))

grid.newpage()
draw.pairwise.venn(area1, area2, cross.area, category = c("pellagatti", "Inoue"), lty = rep("blank", 2), fill = c("light blue", "pink"), alpha = rep(0.5, 2), cat.pos = c(0, 0), cat.dist = rep(0.025, 2))

```

##Load in out alt3p splice event results file
```{r}
#CH_combined
#alt3p_findings_1UMI <- read.table("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/7.Permutation_junction_output/4.Comb_patients_merge_counts/pre_pseudocounts/CH_combined/logOR_within_cell_type_ALT_3P_Junctions_with_threshold_info.txt")

alt3p_findings_1UMI <- read.table("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/7.Permutation_junction_output/4.Comb_patients_merge_counts/CH_combined/logOR_within_cell_type_ALT_3P_Junctions_with_threshold_info.txt")

#alt3p_findings_2UMI <- read.table("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/7.Permutation_junction_output/6.Comb_patients_2WT/5_read_threshold/CH_combined/logOR_within_cell_type_ALT_3P_Junctions_with_threshold_info.txt")


#MDS_combined
#alt3p_findings_1UMI <- read.table("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/7.Permutation_junction_output/4.Comb_patients_merge_counts/pre_pseudocounts/MDS_P5_P6/logOR_within_cell_type_ALT_3P_Junctions_with_threshold_info.txt")

#alt3p_findings_1UMI <- read.table("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/7.Permutation_junction_output/4.Comb_patients_merge_counts/MDS_P5_P6/logOR_within_cell_type_ALT_3P_Junctions_with_threshold_info.txt")

#alt3p_findings_2UMI <- read.table("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/7.Permutation_junction_output/6.Comb_patients_2WT/5_read_threshold/MDS_P5_P6/logOR_within_cell_type_ALT_3P_Junctions_with_threshold_info.txt")

alt3p_findings <- alt3p_findings_1UMI
alt3p.findings.gene.list <- alt3p_findings$gene

alt3p.findings.gene.list.cryptic <- alt3p_findings[which(alt3p_findings$Final_Verdict =="Cryptic_threeprime" & alt3p_findings$dPSI > 0),]$gene

alt3p.findings.gene.list.cryptic.sig <- alt3p_findings[which(alt3p_findings$Final_Verdict =="Cryptic_threeprime" & alt3p_findings$dPSI > 0 & alt3p_findings$pvalue <= 0.05),]$gene


length(alt3p.findings.gene.list.cryptic)
length(alt3p.findings.gene.list.cryptic.sig)

```


```{r}
alt3p_findings_1UMI[which(alt3p_findings_1UMI$gene =="OXA1L"),]
```

```{r}
alt3p_findings_1UMI[which(alt3p_findings_1UMI$gene %in% c("ERGIC3","RPL22L1","OXA1L")& alt3p_findings_1UMI$Final_Verdict =="Cryptic_threeprime"),] %>% arrange(intron_junction)
alt3p_findings_2UMI[which(alt3p_findings_2UMI$gene %in% c("ERGIC3","RPL22L1","OXA1L")& alt3p_findings_2UMI$Final_Verdict =="Cryptic_threeprime"),] %>% arrange(intron_junction)
```


##Looking at things that are different between the 2 groups
```{r}
length(setdiff(alt3p.reference.gene.list.cryptic,alt3p.findings.gene.list.cryptic))
setdiff(alt3p.reference.gene.list.cryptic,alt3p.findings.gene.list.cryptic)

length(unique(alt3p_findings[which((alt3p_findings$gene %in% setdiff(alt3p.reference.gene.list.cryptic, alt3p.findings.gene.list.cryptic)) & alt3p_findings$dPSI >0),]$gene))
unique(alt3p_findings[which((alt3p_findings$gene %in% setdiff(alt3p.reference.gene.list.cryptic,alt3p.findings.gene.list.cryptic)) & alt3p_findings$dPSI > 0),]$gene)

alt3p_findings[which( (alt3p_findings$gene %in% setdiff(alt3p.reference.gene.list.cryptic,alt3p.findings.gene.list.cryptic)) & alt3p_findings$dPSI >0),]
```



##--------------------  Making VennDiagrams of overlap  --------------------##

##Regardless of significance 
```{r fig.height=5, fig.width=5}

area1 = length(unique(alt3p.findings.gene.list.cryptic))
area2 = length(unique(alt3p.reference.union.gene.list.cryptic))
cross.area = length(intersect(alt3p.findings.gene.list.cryptic, alt3p.reference.union.gene.list.cryptic))

grid.newpage()
#draw.pairwise.venn(area1, area2, cross.area, category = c("CH.combined all MUT vs WT", "Bulk SF3B1mut vs WT"), lty = rep("blank", 2), fill = c("light blue", "pink"), alpha = rep(0.5, 2), cat.pos = c(0, 0), cat.dist = rep(0.025, 2))

draw.pairwise.venn(area1, area2, cross.area, category = c("MDS.untreated.combined all MUT vs WT", "Bulk SF3B1mut vs WT"), lty = rep("blank", 2), fill = c("light blue", "pink"), alpha = rep(0.5, 2), cat.pos = c(0, 0), cat.dist = rep(0.025, 2))
```


```{r}
intersect(alt3p.findings.gene.list.cryptic, alt3p.reference.union.gene.list.cryptic)
```



##Looking only at those that are significant 
```{r fig.height=5, fig.width=5}
area1 = length(unique(alt3p.findings.gene.list.cryptic.sig))
area2 = length(unique(alt3p.reference.union.gene.list.cryptic))
cross.area = length(intersect(alt3p.findings.gene.list.cryptic.sig, alt3p.reference.union.gene.list.cryptic))

grid.newpage()
draw.pairwise.venn(area1, area2, cross.area, category = c("CH.combined all MUT vs WT", "Bulk SF3B1mut vs WT"), lty = rep("blank", 2), fill = c("light blue", "pink"), alpha = rep(0.5, 2), cat.pos = c(0, 0), cat.dist = rep(0.025, 2))

#draw.pairwise.venn(area1, area2, cross.area, category = c("MDS.untreated.combined all MUT vs WT", "Bulk SF3B1mut vs WT"), lty = rep("blank", 2), fill = c("light blue", "pink"), alpha = rep(0.5, 2), cat.pos = c(0, 0), cat.dist = rep(0.025, 2))
```

```{r}
#MDS-2UMI
intersect(alt3p.findings.gene.list.cryptic.sig, alt3p.reference.union.gene.list.cryptic)
```


```{r}
#MDS-1UMI
intersect(alt3p.findings.gene.list.cryptic.sig, alt3p.reference.union.gene.list.cryptic)
```


```{r}
#CH-2UMI
intersect(alt3p.findings.gene.list.cryptic.sig, alt3p.reference.union.gene.list.cryptic)
```


```{r}
#CH-1UMI
intersect(alt3p.findings.gene.list.cryptic.sig, alt3p.reference.union.gene.list.cryptic)
```


##WT coverage gets super low

```{r}
alt3p_findings[which(alt3p_findings$Final_Verdict =="Cryptic_threeprime"  & alt3p_findings$gene %in% c("ORAI2","UXS1")),]

alt3p_findings[which(alt3p_findings$Final_Verdict =="Cryptic_threeprime"  & alt3p_findings$gene %in% intersect(alt3p.findings.gene.list.cryptic.sig, alt3p.reference.union.gene.list.cryptic)),]

alt3p_findings[which(alt3p_findings$Final_Verdict =="Cryptic_threeprime" & alt3p_findings$dPSI > 0 & alt3p_findings$gene %in% intersect(alt3p.findings.gene.list.cryptic.sig, alt3p.reference.union.gene.list.cryptic)),]
```



##Looking into those that are lost when you add in the significance threshold 
```{r}


gene.set1 <- intersect(alt3p.findings.gene.list.cryptic, alt3p.reference.union.gene.list.cryptic)
gene.set2 <- intersect(alt3p.findings.gene.list.cryptic.sig, alt3p.reference.union.gene.list.cryptic)
gene.set2

gene.to.check <- setdiff(gene.set1, gene.set2)

alt3p_findings[which(alt3p_findings$Final_Verdict =="Cryptic_threeprime" & alt3p_findings$dPSI > 0 & alt3p_findings$gene %in% gene.to.check),]

```


```{r}
alt3p_findings[which(alt3p_findings$Final_Verdict =="Cryptic_threeprime" & alt3p_findings$dPSI > 0 & alt3p_findings$gene %in% gene.to.check),]

alt3p_findings[which(alt3p_findings$Final_Verdict =="Cryptic_threeprime" & alt3p_findings$dPSI > 0 & alt3p_findings$pvalue <= 0.05 & alt3p_findings$gene %in% gene.set2),]


library(reshape2)
to.plot.1 <- melt(alt3p_findings[which(alt3p_findings$Final_Verdict =="Cryptic_threeprime" & alt3p_findings$dPSI > 0 & alt3p_findings$gene %in% gene.to.check),c("three.mut.cluster.cov",
"three.wt.cluster.cov")]) 

to.plot.2 <- melt(alt3p_findings[which(alt3p_findings$Final_Verdict =="Cryptic_threeprime" & alt3p_findings$dPSI > 0 & alt3p_findings$pvalue <= 0.05 & alt3p_findings$gene %in% gene.set2),c("three.mut.cluster.cov",
"three.wt.cluster.cov")]) 



to.plot.2 

```


```{r}

ggplot(alt3p_findings[which(alt3p_findings$Final_Verdict =="Cryptic_threeprime" & alt3p_findings$dPSI > 0 & alt3p_findings$gene %in% gene.to.check),], aes(three.mut.cluster.cov)) + geom_histogram(binwidth = 100) + xlim(0,8000)

ggplot(alt3p_findings[which(alt3p_findings$Final_Verdict =="Cryptic_threeprime" & alt3p_findings$dPSI > 0 & alt3p_findings$pvalue <= 0.05 & alt3p_findings$gene %in% gene.set2),], aes(three.mut.cluster.cov)) + geom_histogram(binwidth = 100) + xlim(0,8000)


ggplot(alt3p_findings[which(alt3p_findings$Final_Verdict =="Cryptic_threeprime" & alt3p_findings$dPSI > 0 & alt3p_findings$gene %in% gene.to.check),], aes(three.wt.cluster.cov)) + geom_histogram(binwidth = 20) + xlim(0,1000)

ggplot(alt3p_findings[which(alt3p_findings$Final_Verdict =="Cryptic_threeprime" & alt3p_findings$dPSI > 0 & alt3p_findings$pvalue <= 0.05 & alt3p_findings$gene %in% gene.set2),], aes(three.wt.cluster.cov)) + geom_histogram(binwidth = 20) + xlim(0,1000)


```

##-------------------- Correlate the wtPSI, mutPSI, dPSI, and p-value with findings from Omar's paper -------------------- ##



```{r}
alt3p_findings[which(alt3p_findings$gene %in% intersect(alt3p.findings.gene.list.cryptic.sig, alt3p.reference.union.gene.list.cryptic)  & alt3p_findings$Final_Verdict == "Cryptic_threeprime" & alt3p_findings$pvalue <= 0.05),]
```

##Referenece Info 
```{r}

omar.reference_1 <- read.csv("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/11.Alt3p_splicing_references/OmarBRD9.K562.csv")
omar.reference_2 <- read.csv("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/11.Alt3p_splicing_references/OmarBRD9.MEL270.csv")
omar.reference_1 <- omar.reference_1[which(omar.reference_1$Type == "a3ss" & omar.reference_1$deltaPsi > 0),]
omar.reference_2 <- omar.reference_2[which(omar.reference_2$Type == "a3ss" & omar.reference_2$deltaPsi > 0),]

##Better to filter by event or gene?? - Doesn't matter
omar.reference_2[which(!(omar.reference_2$Gene.Name %in% omar.reference_1$Gene.Name)),]
omar.reference_2[which(!(omar.reference_2$Event %in% omar.reference_1$Event)),]


omar.reference <- rbind(omar.reference_1, omar.reference_2[which(!(omar.reference_2$Event %in% omar.reference_1$Event)),])
length(unique(omar.reference$Gene.Name))
omar.reference 
```

##CH Genotype 1UMI 
```{r}

Ref <- omar.reference[which(omar.reference$Gene.Name %in% intersect(alt3p.findings.gene.list.cryptic.sig, alt3p.reference.union.gene.list.cryptic) & omar.reference$Type == "a3ss"),]  %>% arrange(Gene.Name)

Obs <- alt3p_findings[which(alt3p_findings$gene %in% intersect(alt3p.findings.gene.list.cryptic.sig, alt3p.reference.union.gene.list.cryptic)  & alt3p_findings$Final_Verdict == "Cryptic_threeprime" & alt3p_findings$pvalue <= 0.05 & alt3p_findings$total.reads.per.junction >= 10),] %>% arrange(gene)

gene.list <- omar.reference[which(omar.reference$Gene.Name %in% intersect(alt3p.findings.gene.list.cryptic.sig, alt3p.reference.union.gene.list.cryptic) & omar.reference$Type == "a3ss"),]$Gene.Name

Ref
Obs
```

##CH Genotype 2UMI 
```{r}

Ref <- omar.reference[which(omar.reference$Gene.Name %in% intersect(alt3p.findings.gene.list.cryptic.sig, alt3p.reference.union.gene.list.cryptic) & omar.reference$Type == "a3ss"),]

Obs <- alt3p_findings[which(alt3p_findings$gene %in% intersect(alt3p.findings.gene.list.cryptic.sig, alt3p.reference.union.gene.list.cryptic)  & alt3p_findings$Final_Verdict == "Cryptic_threeprime" & alt3p_findings$pvalue <= 0.05 & alt3p_findings$dPSI >= 2),]

gene.list <- omar.reference[which(omar.reference$Gene.Name %in% intersect(alt3p.findings.gene.list.cryptic.sig, alt3p.reference.union.gene.list.cryptic) & omar.reference$Type == "a3ss"),]$Gene.Name

Ref$Gene.Name <- paste(Ref$Gene.Name)
Obs$gene <- paste(Obs$gene)

Ref <- Ref %>% arrange(Gene.Name)
Obs <- Obs %>% arrange(gene)

Ref
Obs
```

```{r}
alt3p_findings[which(alt3p_findings$gene == "TPR"),]
alt3p_findings[which(is.na(alt3p_findings$pvalue)),]
alt3p_findings[which(alt3p_findings$three.obs.logOR.ratio == "-Inf"),]
alt3p_findings

```


#MDS Genotype 1 UMI
```{r}
Ref <- omar.reference[which(omar.reference$Gene.Name %in% intersect(alt3p.findings.gene.list.cryptic.sig, alt3p.reference.union.gene.list.cryptic) & omar.reference$Type == "a3ss" & !(omar.reference$Gene.Name %in% c("ARFGAP2","FDPS","TPR"))),]

Obs <- alt3p_findings[which(alt3p_findings$gene %in% intersect(alt3p.findings.gene.list.cryptic.sig, alt3p.reference.union.gene.list.cryptic)  & alt3p_findings$Final_Verdict == "Cryptic_threeprime" & alt3p_findings$pvalue <= 0.05& alt3p_findings$dPSI >1 & alt3p_findings$gene != "TPR" ),]

gene.list <- omar.reference[which(omar.reference$Gene.Name %in% intersect(alt3p.findings.gene.list.cryptic.sig, alt3p.reference.union.gene.list.cryptic) & omar.reference$Type == "a3ss"),]$Gene.Name

Ref$Gene.Name <- paste(Ref$Gene.Name)
Obs$gene <- paste(Obs$gene)

Ref <- Ref %>% arrange(Gene.Name)
Obs <- Obs %>% arrange(gene)

Ref 
Obs
```

#MDS Genotype 2 UMI
```{r}
Ref <- omar.reference[which(omar.reference$Gene.Name %in% intersect(alt3p.findings.gene.list.cryptic.sig, alt3p.reference.union.gene.list.cryptic) & omar.reference$Type == "a3ss" & !(omar.reference$Gene.Name %in% c("ARFGAP2","FDPS","TPR"))),]

Obs <- alt3p_findings[which(alt3p_findings$gene %in% intersect(alt3p.findings.gene.list.cryptic.sig, alt3p.reference.union.gene.list.cryptic)  & alt3p_findings$Final_Verdict == "Cryptic_threeprime" & alt3p_findings$pvalue <= 0.05& alt3p_findings$dPSI >1 & alt3p_findings$gene != "TPR" ),]

gene.list <- omar.reference[which(omar.reference$Gene.Name %in% intersect(alt3p.findings.gene.list.cryptic.sig, alt3p.reference.union.gene.list.cryptic) & omar.reference$Type == "a3ss"),]$Gene.Name

Ref$Gene.Name <- paste(Ref$Gene.Name)
Obs$gene <- paste(Obs$gene)

Ref <- Ref %>% arrange(Gene.Name)
Obs <- Obs %>% arrange(gene)

Ref 
Obs
```


##Create a table to look at mut and wt PSIs separatly 

```{r}

ref.vs.obs.junc.comparison <- as.data.frame(matrix(ncol=3,nrow=length(gene.list)*2))

colnames(ref.vs.obs.junc.comparison) <- c("Ref","Obs","Group")
rownames(ref.vs.obs.junc.comparison)[1:length(gene.list)] <- paste0(Ref$Gene.Name,".wt.psi")
rownames(ref.vs.obs.junc.comparison)[(length(gene.list)+1):(length(Ref$Gene.Name)*2)] <- paste0(Ref$Gene.Name,".mut.psi")

ref.vs.obs.junc.comparison[1:length(gene.list),]$Group <- "WT"
ref.vs.obs.junc.comparison[(length(gene.list)+1):(length(gene.list)*2),]$Group <- "MUT"

ref.vs.obs.junc.comparison$Ref <- c(Ref$psi_SF3B1_WT*100,Ref$psi_SF3B1_K700E*100)
ref.vs.obs.junc.comparison$Obs<- c(Obs$wt.psi,Obs$mut.psi)

ref.vs.obs.junc.comparison
```



```{r }
library("ggpubr")

ggscatter(ref.vs.obs.junc.comparison, x = "Ref", y = "Obs", add = "reg.line", conf.int = FALSE, color = "Group", xlab = paste("Reference PSIs"), ylab = paste0("Observed PSIs"), label = rownames(ref.vs.obs.junc.comparison), repel = TRUE) +  stat_cor(aes(color = Group), label.x = 3) + stat_cor(aes(color = Group), method = "spearman", cor.coef.name = "rho",label.x=30) + ylim(0,100) + xlim(0,100)

#+  stat_cor(method = "pearson", aes(color = Group), label.x = 3) + ylim(0,100) + xlim(0,100)

ggscatter(ref.vs.obs.junc.comparison, x = "Ref", y = "Obs", add = "reg.line", add.params = list(color = "blue", fill = "lightgray"),conf.int = FALSE, xlab = paste("Reference PSIs"), ylab = paste0("Observed PSIs"), label = rownames(ref.vs.obs.junc.comparison), repel = TRUE) +  stat_cor(label.x = 3, label.y = 90) + stat_cor(method = "spearman", cor.coef.name = "rho",label.x = 35, label.y = 90) + ylim(0,100) + xlim(0,100)

#stat_cor(method = "pearson", label.x = 3)

#ggscatter(ref.vs.obs.junc.comparison, x = "Ref", y = "Obs", add = "reg.line", conf.int = FALSE, color = "Group", xlab = paste("Reference PSIs"), ylab = paste0("Observed PSIs"), label = rownames(ref.vs.obs.junc.comparison), repel = TRUE) +  stat_cor(method = "spearman", aes(color = Group), label.x = 3) + ylim(0,80)

```

##Looking at dPSI

```{r}
ref.vs.obs.junc.dPSI.comparison <- as.data.frame(matrix(ncol=2,nrow=length(gene.list)))

colnames(ref.vs.obs.junc.dPSI.comparison) <- c("Ref","Obs")
rownames(ref.vs.obs.junc.dPSI.comparison)<- Ref$Gene.Name

ref.vs.obs.junc.dPSI.comparison$Ref <- Ref$deltaPsi * 100
ref.vs.obs.junc.dPSI.comparison$Obs<- Obs$dPSI

```

```{r}
ref.vs.obs.junc.dPSI.comparison[-3,]

ggscatter(ref.vs.obs.junc.dPSI.comparison[-3,], x = "Ref", y = "Obs", add = "reg.line", add.params = list(color = "blue"), conf.int = FALSE, xlab = paste("Reference dPSIs"), ylab = paste0("Observed dPSIs"), label = rownames(ref.vs.obs.junc.dPSI.comparison[-3,]), repel = TRUE) +  stat_cor(method = "pearson", label.x = 3)

ggscatter(ref.vs.obs.junc.dPSI.comparison, x = "Ref", y = "Obs", add = "reg.line", add.params = list(color = "blue"), conf.int = FALSE, xlab = paste("Reference dPSIs"), ylab = paste0("Observed dPSIs"), label = rownames(ref.vs.obs.junc.dPSI.comparison), repel = TRUE) +  stat_cor(method = "pearson", label.x = 3)
```



##Correlation of PSI's regardless of significance 
```{r}

##CH 1UMI
Ref <- omar.reference[which(omar.reference$Gene.Name %in% intersect(alt3p.findings.gene.list.cryptic, alt3p.reference.union.gene.list.cryptic) & omar.reference$Type == "a3ss"),]  %>% arrange(Gene.Name)

Obs <- alt3p_findings[which(alt3p_findings$gene %in% intersect(alt3p.findings.gene.list.cryptic, alt3p.reference.union.gene.list.cryptic)  & alt3p_findings$Final_Verdict == "Cryptic_threeprime"),] %>% arrange(gene)

gene.list <- omar.reference[which(omar.reference$Gene.Name %in% intersect(alt3p.findings.gene.list.cryptic.sig, alt3p.reference.union.gene.list.cryptic) & omar.reference$Type == "a3ss"),]$Gene.Name

Ref
Obs
```

```{r}

require(data.table)

#  Turn data.frame into data.table, which looks like..
dt <- data.table(Ref)
dt$Event <- paste(dt$Event)
dt$Event <- unlist(lapply(strsplit(dt$Event,split= "@"), "[",2))
dt

trial <- dt[ , list( Event = unlist( strsplit( Event , "\\|" ) ) ) , by = Event ]
colnames(trial)[2] <- "Event_Collapsed"
trial$chr <- unlist(lapply(strsplit(trial$Event_Collapsed,split= ":"), "[",1))
trial$start <- unlist(lapply(strsplit(trial$Event_Collapsed,split= ":"), "[",2))
trial$end <- as.numeric(unlist(lapply(strsplit(trial$Event_Collapsed,split= ":"), "[",3)))
trial$strand <- unlist(lapply(strsplit(trial$Event_Collapsed,split= ":"), "[",4))
trial$end <- (trial$end-1)
trial$hg19 <- paste0("chr",trial$chr,":",trial$start,"-",trial$end)
trial
```


```{r}
write.table(trial$hg19,file="/gpfs/commons/groups/landau_lab/SF3B1_splice_project/11.Alt3p_splicing_references/junctions.hg19.txt", quote = F, row.names = F, col.names = F)
```


```{r}
trial
```


```{r}
junctions.hg38 <- read.table("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/11.Alt3p_splicing_references/junctions.hg38.txt")
trial$hg38 <- paste0(junctions.hg38$V1,":",trial$strand)

trial
Obs
```

```{r}

intersect(trial$hg38, Obs$intron_junction)

```

```{r}
Ref$Event <- paste(Ref$Event)
events <- trial[which(trial$hg38 %in% intersect(trial$hg38, Obs$intron_junction)),]$Event
Ref$Event_2 <- unlist(lapply(strsplit(Ref$Event,split= "@"), "[",2))

Ref$Gene.Name <- paste(Ref$Gene.Name)
Obs$gene <- paste(Obs$gene)


Ref <- Ref[which(Ref$Event_2 %in% events),] %>% arrange(Gene.Name)
Obs <- Obs[which(Obs$intron_junction %in% intersect(trial$hg38, Obs$intron_junction)),]  %>% arrange(gene)

Ref <- Ref[-14,]
Ref
Obs

gene.list <- Ref$Gene.Name

```



```{r}

ref.vs.obs.junc.comparison <- as.data.frame(matrix(ncol=3,nrow=length(gene.list)*2))

colnames(ref.vs.obs.junc.comparison) <- c("Ref","Obs","Group")
rownames(ref.vs.obs.junc.comparison)[1:length(gene.list)] <- paste0(Ref$Gene.Name,".wt.psi")
rownames(ref.vs.obs.junc.comparison)[(length(gene.list)+1):(length(Ref$Gene.Name)*2)] <- paste0(Ref$Gene.Name,".mut.psi")

ref.vs.obs.junc.comparison[1:length(gene.list),]$Group <- "WT"
ref.vs.obs.junc.comparison[(length(gene.list)+1):(length(gene.list)*2),]$Group <- "MUT"

ref.vs.obs.junc.comparison$Ref <- c(Ref$psi_SF3B1_WT*100,Ref$psi_SF3B1_K700E*100)
ref.vs.obs.junc.comparison$Obs<- c(Obs$wt.psi,Obs$mut.psi)

ref.vs.obs.junc.comparison
```


```{r}
library("ggpubr")

ggscatter(ref.vs.obs.junc.comparison, x = "Ref", y = "Obs", add = "reg.line", conf.int = FALSE, color = "Group", xlab = paste("Reference PSIs"), ylab = paste0("Observed PSIs"), label = rownames(ref.vs.obs.junc.comparison), repel = TRUE) +  stat_cor(aes(color = Group), label.x = 3) + stat_cor(aes(color = Group), method = "spearman", cor.coef.name = "rho",label.x=30) + ylim(0,100) + xlim(0,100)

#+  stat_cor(method = "pearson", aes(color = Group), label.x = 3) + ylim(0,100) + xlim(0,100)

ggscatter(ref.vs.obs.junc.comparison, x = "Ref", y = "Obs", add = "reg.line", add.params = list(color = "blue", fill = "lightgray"),conf.int = FALSE, xlab = paste("Reference PSIs"), ylab = paste0("Observed PSIs"), label = rownames(ref.vs.obs.junc.comparison), repel = TRUE) +  stat_cor(label.x = 3, label.y = 90) + stat_cor(method = "spearman", cor.coef.name = "rho",label.x = 35, label.y = 90) + ylim(0,100) + xlim(0,100)
```

```{r}
ref.vs.obs.junc.dPSI.comparison

```


##Looking at dPSI

```{r}
ref.vs.obs.junc.dPSI.comparison <- as.data.frame(matrix(ncol=2,nrow=length(gene.list)))

colnames(ref.vs.obs.junc.dPSI.comparison) <- c("Ref","Obs")
rownames(ref.vs.obs.junc.dPSI.comparison)<- Ref$Gene.Name

ref.vs.obs.junc.dPSI.comparison$Ref <- Ref$deltaPsi * 100
ref.vs.obs.junc.dPSI.comparison$Obs<- Obs$dPSI

ggscatter(ref.vs.obs.junc.dPSI.comparison, x = "Ref", y = "Obs", add = "reg.line", add.params = list(color = "blue"), conf.int = FALSE, xlab = paste("Reference dPSIs"), ylab = paste0("Observed dPSIs"), label = rownames(ref.vs.obs.junc.dPSI.comparison), repel = TRUE) +  stat_cor(method = "pearson", label.x = 3)

```








# ----------------- Looking at overlap with Pellagati et al Junctions regardless of significance ----------------- ##

##Converting Pellagati et al Junctions into hg38 and proper format

```{r}
pellagati.reference <-read_csv("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/11.Alt3p_splicing_references/Pallegatti_Suppl3_SF3B1.csv")

pellagati.reference.a3ss <- pellagati.reference[which(pellagati.reference$`Event type` == "A3SS"),]
pellagati.reference.a3ss
```

```{r}
pellagati.reference.a3ss$hg19 <- paste0("chr",pellagati.reference.a3ss$Chr,":" ,pellagati.reference.a3ss$`Start position`,"-" ,pellagati.reference.a3ss$`End Position`)

pellagati.reference.a3ss[which(pellagati.reference.a3ss$Gene == "DPH5"),]
```

```{r}
write.table(pellagati.reference.a3ss$hg19,file="/gpfs/commons/groups/landau_lab/SF3B1_splice_project/11.Alt3p_splicing_references/junctions.hg19.pellagati.txt", quote = F, row.names = F, col.names = F)
```


```{r}
junctions.hg38 <- read.table("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/11.Alt3p_splicing_references/junctions.hg38.pellagati.txt")

junctions.hg38
pellagati.reference.a3ss
pellagati.reference.a3ss$hg38 <- paste0(junctions.hg38$V1)


pellagati.reference.a3ss$start.hg38 <- unlist(lapply(strsplit(pellagati.reference.a3ss$hg38, split= "-"), "[",1))
pellagati.reference.a3ss$start.hg38 <- unlist(lapply(strsplit(pellagati.reference.a3ss$start.hg38, split= ":"), "[",2))
pellagati.reference.a3ss$end.hg38 <- unlist(lapply(strsplit(pellagati.reference.a3ss$hg38,split= "-"), "[",2))
```


```{r}
pellagati.reference.a3ss
pellagati.reference.a3ss[,c("Gene","Event type","Strand","hg19","hg38")]
```

```{r}
Ref <- pellagati.reference.a3ss[which(pellagati.reference.a3ss$Gene %in% intersect(alt3p.findings.gene.list, pellagati.reference.a3ss$Gene)),] %>% arrange(Gene)
Ref <- pellagati.reference.a3ss

##Add in fiveprime and threeprim like we have 
Ref = Ref %>% mutate(five_prime = ifelse(Strand=="+", start.hg38, end.hg38),
                     three_prime = ifelse(Strand=="+", end.hg38, start.hg38))

Ref$cryptic.junction <- paste0("chr",Ref$Chr,":",Ref$five_prime)
Ref$dPSI <- (Ref$`Inclusion Level_WTMDS` * 100)

Ref[which(Ref$Gene =="SEPT6"),c("Gene","Event type","Strand","hg19","hg38","cryptic.junction")]

```

```{r}
alt3p_findings[which(alt3p_findings$intron_junction =="chr5:140485732:140486957:+"),]
alt3p_findings[which(alt3p_findings$gene =="SEPT6"),]

```


## -------- CH.combined 1UMI -------- ##
```{r}
alt3p_findings_1UMI <- read.table("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/7.Permutation_junction_output/4.Comb_patients_merge_counts/CH_combined/logOR_within_cell_type_ALT_3P_Junctions_with_threshold_info.txt")
alt3p_findings <- alt3p_findings_1UMI

#Obs <- alt3p_findings[which(alt3p_findings$gene %in% intersect(alt3p.findings.gene.list.cryptic, pellagati.reference.a3ss$Gene) & alt3p_findings$Final_Verdict == "Cryptic_threeprime"),] %>% arrange(gene)

Obs <- alt3p_findings
Obs$gene <- paste(Obs$gene)
Obs$cryptic.junction <- paste0(Obs$chr,":",Obs$three_prime)

##Final tables
Ref.final <- Ref[which(Ref$cryptic.junction %in% intersect(Ref$cryptic.junction,Obs$cryptic.junction)),] %>% arrange(cryptic.junction)
Obs.final <- Obs[which(Obs$cryptic.junction %in% intersect(Ref$cryptic.junction,Obs$cryptic.junction)),] %>% arrange(cryptic.junction)

Ref.final <- Ref.final[which(!(Ref.final$Gene %in% c("LETMD1","TMEM91") )),]
Obs.final <- Obs.final[which(!(Obs.final$gene %in% c("LETMD1","TMEM91") )),]

Ref.final
Obs.final

Obs.final <- Obs.final[which(!(Obs.final$five_prime_ID %in% c("clu_76186"))),]
Ref.final
Obs.final
gene.list <- Obs.final$gene

```


## ----------- CH.combined 2UMI ----------- ##
```{r}

alt3p_findings_2UMI <- read.table("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/7.Permutation_junction_output/6.Comb_patients_2WT/5_read_threshold/CH_combined/logOR_within_cell_type_ALT_3P_Junctions_with_threshold_info.txt")
alt3p_findings <- alt3p_findings_2UMI

#Obs <- alt3p_findings[which(alt3p_findings$gene %in% intersect(alt3p.findings.gene.list.cryptic, pellagati.reference.a3ss$Gene) & alt3p_findings$Final_Verdict == "Cryptic_threeprime"),] %>% arrange(gene)

Obs <- alt3p_findings
Obs$gene <- paste(Obs$gene)
Obs$cryptic.junction <- paste0(Obs$chr,":",Obs$three_prime)
Obs
Ref

##Final tables
Ref.final <- Ref[which(Ref$cryptic.junction %in% intersect(Ref$cryptic.junction,Obs$cryptic.junction)),] %>% arrange(cryptic.junction)
Obs.final <- Obs[which(Obs$cryptic.junction %in% intersect(Ref$cryptic.junction,Obs$cryptic.junction)),] %>% arrange(cryptic.junction)

Ref.final
Obs.final

Ref.final <- Ref.final[which(!(Ref.final$Gene %in% c("TMEM91") )),]
Obs.final <- Obs.final[which(!(Obs.final$gene %in% c("TMEM91") )),]

Obs.final <- Obs.final[which(!(Obs.final$five_prime_ID %in% c("clu_76186"))),]

Ref.final
Obs.final

gene.list <- Obs.final$gene

```



## ----------- MDS.combined 1UMI ----------- ##
```{r}
alt3p_findings_1UMI <- read.table("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/7.Permutation_junction_output/4.Comb_patients_merge_counts/MDS_P5_P6/logOR_within_cell_type_ALT_3P_Junctions_with_threshold_info.txt")
alt3p_findings <- alt3p_findings_1UMI


#Obs <- alt3p_findings[which(alt3p_findings$Final_Verdict == "Cryptic_threeprime"),]
Obs <- alt3p_findings
Obs$gene <- paste(Obs$gene)
Obs$cryptic.junction <- paste0(Obs$chr,":",Obs$three_prime)
Obs


##Final tables
Ref.final <- Ref[which(Ref$cryptic.junction %in% intersect(Ref$cryptic.junction,Obs$cryptic.junction)),] %>% arrange(cryptic.junction)
Obs.final <- Obs[which(Obs$cryptic.junction %in% intersect(Ref$cryptic.junction,Obs$cryptic.junction)),] %>% arrange(cryptic.junction)

Ref.final
Obs.final

Ref.final <- Ref.final[which(!(Ref.final$Gene %in% c("TMEM91") )),]
Obs.final <- Obs.final[which(!(Obs.final$gene %in% c("TMEM91") )),]

Obs.final <- Obs.final[which(!(Obs.final$five_prime_ID %in% c("clu_86720","clu_82654","clu_68133","clu_68132","clu_47488","clu_35713","clu_23648","clu_17093","clu_17091"))),]


Ref.final
Obs.final

gene.list <- Obs.final$gene


```


## ----------- MDS.combined 2UMI ----------- ##
```{r}

alt3p_findings_2UMI <- read.table("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/7.Permutation_junction_output/6.Comb_patients_2WT/5_read_threshold/MDS_P5_P6/logOR_within_cell_type_ALT_3P_Junctions_with_threshold_info.txt")
alt3p_findings <- alt3p_findings_2UMI


#Obs <- alt3p_findings[which(alt3p_findings$gene %in% intersect(alt3p.findings.gene.list.cryptic, pellagati.reference.a3ss$Gene) & alt3p_findings$Final_Verdict == "Cryptic_threeprime"),] %>% arrange(gene)

Obs <- alt3p_findings
Obs$gene <- paste(Obs$gene)
Obs$cryptic.junction <- paste0(Obs$chr,":",Obs$three_prime)

##Final tables
Ref.final <- Ref[which(Ref$cryptic.junction %in% intersect(Ref$cryptic.junction,Obs$cryptic.junction)),] %>% arrange(cryptic.junction)
Obs.final <- Obs[which(Obs$cryptic.junction %in% intersect(Ref$cryptic.junction,Obs$cryptic.junction)),] %>% arrange(cryptic.junction)

Ref.final
Obs.final

Ref.final <- Ref.final[which(!(Ref.final$Gene %in% c("TMEM91") )),]
Obs.final <- Obs.final[which(!(Obs.final$gene %in% c("TMEM91") )),]

Obs.final <- Obs.final[which(!(Obs.final$five_prime_ID %in% c("clu_86720","clu_82654","clu_68133","clu_68132","clu_47488","clu_35713","clu_23648","clu_17093","clu_17091"))),]


Ref.final
Obs.final

gene.list <- Obs.final$gene
```

## Create table to compare dPSIs ##

```{r}
ref.vs.obs.junc.dPSI.comparison <- as.data.frame(matrix(ncol=2,nrow=length(gene.list)))

colnames(ref.vs.obs.junc.dPSI.comparison) <- c("Ref","Obs")
rownames(ref.vs.obs.junc.dPSI.comparison)<- gene.list

ref.vs.obs.junc.dPSI.comparison$Ref <- Ref.final$dPSI
ref.vs.obs.junc.dPSI.comparison$Obs <- Obs.final$dPSI
```


```{r}
ref.vs.obs.junc.dPSI.comparison
```


##Overlap the dPSIs CH 1UMI - not filtering by "Cryptic_threeprime""
```{r}

ggscatter(ref.vs.obs.junc.dPSI.comparison, x = "Ref", y = "Obs", add = "reg.line", add.params = list(color = "blue"), conf.int = FALSE, xlab = paste("Reference dPSIs"), ylab = paste0("Observed dPSIs"), label = rownames(ref.vs.obs.junc.dPSI.comparison), repel = TRUE) + stat_cor(label.x = -15, label.y = 50) + stat_cor(method = "spearman", cor.coef.name = "rho",label.x = -15, label.y = 45)

```


##Overlap the dPSIs CH 2UMI - not filtering by "Cryptic_threeprime""
```{r}

ggscatter(ref.vs.obs.junc.dPSI.comparison, x = "Ref", y = "Obs", add = "reg.line", add.params = list(color = "blue"), conf.int = FALSE, xlab = paste("Reference dPSIs"), ylab = paste0("Observed dPSIs"), label = rownames(ref.vs.obs.junc.dPSI.comparison), repel = TRUE) + stat_cor(label.x = -15, label.y = 50) + stat_cor(method = "spearman", cor.coef.name = "rho",label.x = -15, label.y = 45)

```


##Overlap the dPSIs MDS 1UMI - not filtering by "Cryptic_threeprime""
```{r}

ggscatter(ref.vs.obs.junc.dPSI.comparison, x = "Ref", y = "Obs", add = "reg.line", add.params = list(color = "blue"), conf.int = FALSE, xlab = paste("Reference dPSIs"), ylab = paste0("Observed dPSIs"), label = rownames(ref.vs.obs.junc.dPSI.comparison), repel = TRUE) + stat_cor(label.x = -15, label.y = 50) + stat_cor(method = "spearman", cor.coef.name = "rho",label.x = -15, label.y = 45)

```

##Overlap the dPSIs MDS 2UMI - filter by "Cryptic_threeprime"
```{r}

ref.vs.obs.junc.dPSI.comparison[-21,]

ggscatter(ref.vs.obs.junc.dPSI.comparison[-21,], x = "Ref", y = "Obs", add = "reg.line", add.params = list(color = "blue"), conf.int = FALSE, xlab = paste("Reference dPSIs"), ylab = paste0("Observed dPSIs"), label = rownames(ref.vs.obs.junc.dPSI.comparison[-21,]), repel = TRUE) +  stat_cor(method = "pearson", label.x = 3)

```

##Overlap the dPSIs MDS 2UMI - not filtering by "Cryptic_threeprime""
```{r}
ref.vs.obs.junc.dPSI.comparison
ggscatter(ref.vs.obs.junc.dPSI.comparison[-34,], x = "Ref", y = "Obs", add = "reg.line", add.params = list(color = "blue"), conf.int = FALSE, xlab = paste("Reference dPSIs"), ylab = paste0("Observed dPSIs"), label = rownames(ref.vs.obs.junc.dPSI.comparison[-34,]), repel = TRUE) + stat_cor(label.x = -15, label.y = 50) + stat_cor(method = "spearman", cor.coef.name = "rho",label.x = -15, label.y = 45)

```



##Overlap the dPSIs
```{r}

ggscatter(ref.vs.obs.junc.dPSI.comparison, x = "Ref", y = "Obs", add = "reg.line", add.params = list(color = "blue"), conf.int = FALSE, xlab = paste("Reference dPSIs"), ylab = paste0("Observed dPSIs"), label = rownames(ref.vs.obs.junc.dPSI.comparison), repel = TRUE) +  stat_cor(method = "pearson", label.x = 3)

```



##Correlate the p-values
```{r}
ref.vs.obs.junc.pval.comparison <- as.data.frame(matrix(ncol=2,nrow=length(gene.list)))

colnames(ref.vs.obs.junc.pval.comparison) <- c("Ref","Obs")
rownames(ref.vs.obs.junc.pval.comparison)<- gene.list

ref.vs.obs.junc.pval.comparison$Ref <- Ref$FDR_WTMDS
ref.vs.obs.junc.pval.comparison$Obs<- Obs$pvalue

ref.vs.obs.junc.pval.comparison

ggscatter(ref.vs.obs.junc.pval.comparison, x = "Ref", y = "Obs", add = "reg.line", add.params = list(color = "blue"), conf.int = FALSE, xlab = paste("Reference p-val"), ylab = paste0("Observed p-val"), label = rownames(ref.vs.obs.junc.dPSI.comparison), repel = TRUE) +  stat_cor(method = "pearson", label.x = 0.001)
```

