

```{r}
intron.meta.data.merged

trial <- read.table("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/12.Bulk.PSI/output/CH259.CH305.all.intron.metadata.with.bulkPSI.ONT.tsv" ,sep="\t")

trial <- read.table("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/12.Bulk.PSI/output/MDSP5.MDSP6.all.intron.metadata.with.bulkPSI.ONT.tsv" ,sep="\t")


intron.meta.data.merged[which(intron.meta.data.merged$gene == "ERGIC3"),]
trial[which(trial$gene == "ERGIC3"),] %>% arrange(intron.junction)
```

##Looking at DNMT3A sample
```{r}
DNMT3A.bulk.psi.all <- read.table("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/12.Bulk.PSI/output/CH502_DNMT3A.all.intron.metadata.with.bulkPSI.ONT.tsv" ,sep="\t")
rownames(DNMT3A.bulk.psi.all) <- DNMT3A.bulk.psi.all$intron.junction

#WT only
DNMT3A.bulk.psi <- read.table("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/12.Bulk.PSI/output/CH502_DNMT3A.WT.only.all.intron.metadata.with.bulkPSI.ONT.tsv" ,sep="\t")
DNMT3A.bulk.psi

DNMT3A.bulk.psi <- DNMT3A.bulk.psi[which(DNMT3A.bulk.psi$cluster.coverage >=5),]
rownames(DNMT3A.bulk.psi) <- DNMT3A.bulk.psi$intron.junction
#DNMT3A.bulk.psi[rownames(psi.values),]

DNMT3A.bulk.psi.all[rownames(psi.values),] %>% arrange(desc(bulk.psi))
DNMT3A.bulk.psi[rownames(psi.values),] %>% arrange(desc(bulk.psi))

```


##Comparing bulk PSIs to separate mut and wt PSIs (using the Alt3p cryptically spliced junctions )
```{r}
#1WT
CH.combined.cryptic.3p.junctions <- read.table("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/7.Permutation_junction_output/4.Comb_patients_merge_counts/CH_combined/logOR_within_cell_type_ONLY_CRYPTIC_3P_Junctions_with_threshold_info.txt")


#2WT
CH.combined.cryptic.3p.junctions <- read.table("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/7.Permutation_junction_output/6.Comb_patients_2WT/5_read_threshold/CH_combined/logOR_within_cell_type_ONLY_CRYPTIC_3P_Junctions_with_threshold_info.txt")

CH.combined.cryptic.3p.junctions
CH.combined.cryptic.3p.junctions[which(CH.combined.cryptic.3p.junctions$pvalue <= 0.05 & abs(CH.combined.cryptic.3p.junctions$dPSI) >=2),]
```


#Create Table of significant cryptic junctions with dPSI > 2 and P <=0.05
```{r}
psi.values <- CH.combined.cryptic.3p.junctions[which(CH.combined.cryptic.3p.junctions$pvalue <= 0.05 & CH.combined.cryptic.3p.junctions$dPSI >=2),c("gene","intron_junction","wt.psi","mut.psi")]

psi.values <- CH.combined.cryptic.3p.junctions[which(CH.combined.cryptic.3p.junctions$pvalue <= 0.05 & CH.combined.cryptic.3p.junctions$dPSI <=-2),c("gene","intron_junction","wt.psi","mut.psi")]

rownames(psi.values) <- psi.values$intron_junction

psi.values$bulk.psi <- trial[rownames(psi.values),]$bulk.psi.CH259.CH305
psi.values$dnmt3a.psi <- DNMT3A.bulk.psi[rownames(psi.values),]$bulk.psi

psi.values <- psi.values[which(!is.na(psi.values$bulk.psi)),]
psi.values <- psi.values[which(!is.na(psi.values$dnmt3a.psi)),]

psi.values %>% arrange(desc(dnmt3a.psi))
DNMT3A.bulk.psi[rownames(psi.values),] %>% arrange(desc(bulk.psi))
CH.combined.cryptic.3p.junctions[which(CH.combined.cryptic.3p.junctions$intron_junction %in% rownames(psi.values)),]


```


#Create Table of significant cryptic junctions with dPSI > 2 regardless of significance 
```{r}
psi.values <- CH.combined.cryptic.3p.junctions[which(CH.combined.cryptic.3p.junctions$dPSI >=2) ,c("gene","intron_junction","wt.psi","mut.psi")]

psi.values <- CH.combined.cryptic.3p.junctions[which(CH.combined.cryptic.3p.junctions$dPSI <=-2) ,c("gene","intron_junction","wt.psi","mut.psi")]

rownames(psi.values) <- psi.values$intron_junction

psi.values$bulk.psi <- trial[rownames(psi.values),]$bulk.psi.CH259.CH305
psi.values$dnmt3a.psi <- DNMT3A.bulk.psi[rownames(psi.values),]$bulk.psi

psi.values <- psi.values[which(!is.na(psi.values$bulk.psi)),]
psi.values <- psi.values[which(!is.na(psi.values$dnmt3a.psi)),]

psi.values %>% arrange(desc(dnmt3a.psi))
DNMT3A.bulk.psi[rownames(psi.values),] %>% arrange(desc(bulk.psi))
CH.combined.cryptic.3p.junctions[which(CH.combined.cryptic.3p.junctions$intron_junction %in% rownames(psi.values)),]
```

##Correlate the WT and DNMT3A PSI's
```{r}
ggplot(psi.values, aes(x=wt.psi, y=dnmt3a.psi)) + geom_point() + geom_smooth(method=lm) + theme(legend.position="none", axis.line=element_blank(), axis.text=element_blank()) + theme_classic(base_size = 17)


library("ggpubr")

ggscatter(psi.values, x = "wt.psi", y = "dnmt3a.psi", add = "reg.line", add.params = list(color = "blue", fill = "lightgray"), conf.int = T, xlab = paste("wt.psi of cryptic 3p site"), ylab = paste0("dnmt3a.psi of cryptic 3p site")) +  stat_cor(method = "pearson", label.x = 3) 
#+ ylim(0,20) + xlim(0,20)


ggscatter(psi.values, x = "mut.psi", y = "dnmt3a.psi", add = "reg.line", add.params = list(color = "blue", fill = "lightgray"), conf.int = T, xlab = paste("mut.psi of cryptic 3p site"), ylab = paste0("dnmt3a.psi of cryptic 3p site")) +  stat_cor(method = "pearson", label.x = 3) 

#+ ylim(0,20) + xlim(0,20)
```



```{r}
MDS.combined.cryptic.3p.junctions <- read.table("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/7.Permutation_junction_output/6.Comb_patients_2WT/5_read_threshold/MDS_P5_P6/logOR_within_cell_type_ONLY_CRYPTIC_3P_Junctions_with_threshold_info.txt")

psi.values <- MDS.combined.cryptic.3p.junctions[which(MDS.combined.cryptic.3p.junctions$dPSI >= 2) ,c("gene","intron_junction","wt.psi","mut.psi")]

#psi.values <- MDS.combined.cryptic.3p.junctions[which(MDS.combined.cryptic.3p.junctions$dPSI <= -2) ,c("gene","intron_junction","wt.psi","mut.psi")]

#for dPSI <= -2
psi.values[duplicated(psi.values$intron_junction), ]
psi.values[which(psi.values$intron_junction == "chr15:79907706:79923451:-"),]
psi.values <- psi.values[which(psi.values$gene != "ST20-MTHFS"),]

rownames(psi.values) <- psi.values$intron_junction
psi.values$bulk.psi <- trial[rownames(psi.values),]$bulk.psi
psi.values$dnmt3a.psi <- DNMT3A.bulk.psi[rownames(psi.values),]$bulk.psi

psi.values <- psi.values[which(!is.na(psi.values$bulk.psi)),]
psi.values <- psi.values[which(!is.na(psi.values$dnmt3a.psi)),]

psi.values %>% arrange(desc(dnmt3a.psi))
DNMT3A.bulk.psi[rownames(psi.values),] %>% arrange(desc(bulk.psi))
MDS.combined.cryptic.3p.junctions[which(CH.combined.cryptic.3p.junctions$intron_junction %in% rownames(psi.values)),]

```


```{r}
MDS.combined.cryptic.3p.junctions <- read.table("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/7.Permutation_junction_output/6.Comb_patients_2WT/5_read_threshold/MDS_P5_P6/logOR_within_cell_type_ONLY_CRYPTIC_3P_Junctions_with_threshold_info.txt")

psi.values <- MDS.combined.cryptic.3p.junctions[which(MDS.combined.cryptic.3p.junctions$pvalue <= 0.05 & MDS.combined.cryptic.3p.junctions$dPSI >= 2) ,c("gene","intron_junction","wt.psi","mut.psi")]

#psi.values <- MDS.combined.cryptic.3p.junctions[which(MDS.combined.cryptic.3p.junctions$pvalue <= 0.05 & MDS.combined.cryptic.3p.junctions$dPSI <= -2) ,c("gene","intron_junction","wt.psi","mut.psi")]

#for dPSI <= -2
psi.values[duplicated(psi.values$intron_junction), ]
psi.values[which(psi.values$intron_junction == "chr15:79907706:79923451:-"),]
psi.values <- psi.values[which(psi.values$gene != "ST20-MTHFS"),]

rownames(psi.values) <- psi.values$intron_junction
psi.values$bulk.psi <- trial[rownames(psi.values),]$bulk.psi
psi.values$dnmt3a.psi <- DNMT3A.bulk.psi[rownames(psi.values),]$bulk.psi

psi.values <- psi.values[which(!is.na(psi.values$bulk.psi)),]
psi.values <- psi.values[which(!is.na(psi.values$dnmt3a.psi)),]

psi.values %>% arrange(desc(dnmt3a.psi))
DNMT3A.bulk.psi[rownames(psi.values),] %>% arrange(desc(bulk.psi))
MDS.combined.cryptic.3p.junctions[which(CH.combined.cryptic.3p.junctions$intron_junction %in% rownames(psi.values)),]

```


```{r}
library(reshape2)
psi.values.ggplot <- melt(psi.values, id.vars=c('intron_junction', 'gene'),var='psi.type')
colnames(psi.values.ggplot)[4] <- "psi.value"

ggplot(psi.values.ggplot, aes(x=psi.type, y=psi.value, color=psi.type)) + geom_boxplot(outlier.shape = NA) + geom_point(pch = 21, position = position_jitterdodge()) + theme_classic(base_size = 17)

ggplot(psi.values.ggplot[which(psi.values.ggplot$psi.type %in% c("wt.psi","mut.psi")),], aes(x=psi.type, y=psi.value, color=psi.type)) + geom_boxplot(outlier.shape = NA) + geom_point(pch = 21, position = position_jitterdodge()) + stat_compare_means(aes(group = psi.type), method = "wilcox.test") + theme_classic(base_size = 17)

ggplot(psi.values.ggplot[which(psi.values.ggplot$psi.type %in% c("wt.psi","bulk.psi")),], aes(x=psi.type, y=psi.value, color=psi.type)) + geom_boxplot(outlier.shape = NA) + geom_point(pch = 21, position = position_jitterdodge()) + stat_compare_means(aes(group = psi.type), method = "wilcox.test") + theme_classic(base_size = 17)

ggplot(psi.values.ggplot[which(psi.values.ggplot$psi.type %in% c("wt.psi","dnmt3a.psi")),], aes(x=psi.type, y=psi.value, color=psi.type)) + geom_boxplot(outlier.shape = NA) + geom_point(pch = 21, position = position_jitterdodge()) + stat_compare_means(aes(group = psi.type), method = "wilcox.test") + theme_classic(base_size = 17)

ggplot(psi.values.ggplot[which(psi.values.ggplot$psi.type %in% c("mut.psi","bulk.psi")),], aes(x=psi.type, y=psi.value, color=psi.type)) + geom_boxplot(outlier.shape = NA) + geom_point(pch = 21, position = position_jitterdodge()) + stat_compare_means(aes(group = psi.type), method = "wilcox.test") + theme_classic(base_size = 17)

ggplot(psi.values.ggplot[which(psi.values.ggplot$psi.type %in% c("mut.psi","dnmt3a.psi")),], aes(x=psi.type, y=psi.value, color=psi.type)) + geom_boxplot(outlier.shape = NA) + geom_point(pch = 21, position = position_jitterdodge()) + stat_compare_means(aes(group = psi.type), method = "wilcox.test") + theme_classic(base_size = 17)

ggplot(psi.values.ggplot[which(psi.values.ggplot$psi.type %in% c("bulk.psi","dnmt3a.psi")),], aes(x=psi.type, y=psi.value, color=psi.type)) + geom_boxplot(outlier.shape = NA) + geom_point(pch = 21, position = position_jitterdodge()) + stat_compare_means(aes(group = psi.type), method = "wilcox.test") + theme_classic(base_size = 17)
```






```{r}
library(reshape2)
psi.values.ggplot <- melt(psi.values, id.vars=c('intron_junction', 'gene'),var='psi.type')
colnames(psi.values.ggplot)[4] <- "psi.value"

ggplot(psi.values.ggplot, aes(x=psi.type, y=psi.value, color=psi.type)) + geom_boxplot(outlier.shape = NA) + geom_point(pch = 21, position = position_jitterdodge()) + theme_classic(base_size = 17)

ggplot(psi.values.ggplot[which(psi.values.ggplot$psi.type %in% c("wt.psi","mut.psi")),], aes(x=psi.type, y=psi.value, color=psi.type)) + geom_boxplot(outlier.shape = NA) + geom_point(pch = 21, position = position_jitterdodge()) + stat_compare_means(aes(group = psi.type), method = "wilcox.test") + theme_classic(base_size = 17)

ggplot(psi.values.ggplot[which(psi.values.ggplot$psi.type %in% c("wt.psi","bulk.psi")),], aes(x=psi.type, y=psi.value, color=psi.type)) + geom_boxplot(outlier.shape = NA) + geom_point(pch = 21, position = position_jitterdodge()) + stat_compare_means(aes(group = psi.type), method = "wilcox.test") + theme_classic(base_size = 17)

ggplot(psi.values.ggplot[which(psi.values.ggplot$psi.type %in% c("wt.psi","dnmt3a.psi")),], aes(x=psi.type, y=psi.value, color=psi.type)) + geom_boxplot(outlier.shape = NA) + geom_point(pch = 21, position = position_jitterdodge()) + stat_compare_means(aes(group = psi.type), method = "wilcox.test") + theme_classic(base_size = 17)

ggplot(psi.values.ggplot[which(psi.values.ggplot$psi.type %in% c("mut.psi","bulk.psi")),], aes(x=psi.type, y=psi.value, color=psi.type)) + geom_boxplot(outlier.shape = NA) + geom_point(pch = 21, position = position_jitterdodge()) + stat_compare_means(aes(group = psi.type), method = "wilcox.test") + theme_classic(base_size = 17)

ggplot(psi.values.ggplot[which(psi.values.ggplot$psi.type %in% c("mut.psi","dnmt3a.psi")),], aes(x=psi.type, y=psi.value, color=psi.type)) + geom_boxplot(outlier.shape = NA) + geom_point(pch = 21, position = position_jitterdodge()) + stat_compare_means(aes(group = psi.type), method = "wilcox.test") + theme_classic(base_size = 17)

ggplot(psi.values.ggplot[which(psi.values.ggplot$psi.type %in% c("bulk.psi","dnmt3a.psi")),], aes(x=psi.type, y=psi.value, color=psi.type)) + geom_boxplot(outlier.shape = NA) + geom_point(pch = 21, position = position_jitterdodge()) + stat_compare_means(aes(group = psi.type), method = "wilcox.test") + theme_classic(base_size = 17)




ggplot(psi.values.ggplot[which(psi.values.ggplot$psi.type != "mut.psi"),], aes(x=psi.type, y=psi.value, color=psi.type)) + geom_boxplot(outlier.shape = NA) + geom_point(pch = 21, position = position_jitterdodge()) + stat_compare_means(aes(group = psi.type), method = "wilcox.test") + theme_classic(base_size = 17)

ggplot(psi.values.ggplot[which(psi.values.ggplot$psi.type != "bulk.psi"),], aes(x=psi.type, y=psi.value, color=psi.type)) + geom_boxplot(outlier.shape = NA) + geom_point(pch = 21, position = position_jitterdodge()) + stat_compare_means(aes(group = psi.type), method = "wilcox.test") + theme_classic(base_size = 17)
```



```{r}
library(reshape2)
psi.values.ggplot <- melt(psi.values, id.vars=c('intron_junction', 'gene'),var='psi.type')
colnames(psi.values.ggplot)[4] <- "psi.value"

ggplot(psi.values.ggplot[which(psi.values.ggplot$psi.type != "mut.psi"),], aes(x=psi.type, y=psi.value, color=psi.type)) + geom_boxplot(outlier.shape = NA) + geom_point(pch = 21, position = position_jitterdodge()) + stat_compare_means(aes(group = psi.type), method = "wilcox.test") + theme_classic(base_size = 17)

```


```{r}
test <- wilcox.test(psi.values$mut.psi,psi.values$bulk.psi, alternative = "less")

test
```

#Create Table of significant cryptic junctions with dPSI > 2 regardless of significance 
```{r}
psi.values <- CH.combined.cryptic.3p.junctions[which(CH.combined.cryptic.3p.junctions$dPSI <= -2) ,c("gene","intron_junction","wt.psi","mut.psi")]

rownames(psi.values) <- psi.values$intron_junction

psi.values$bulk.psi <- trial[rownames(psi.values),]$bulk.psi.CH259.CH305

psi.values <- psi.values[which(!is.na(psi.values$bulk.psi)),]

psi.values


```


```{r}

```


```{r}

```

##Compare the WT PSIs to the DNMT3A PSI-are they higher than expected? Is there a correlation

```{r}

```


##Are the cryptic sites in the DNMT3A sample with really high PSIs real??
```{r}
DNMT3A.bulk.psi[rownames(psi.values),] %>% arrange(desc(bulk.psi))
```


##Subsetting the count matrix to include only the WT cells 
```{r}
#counts.file <- parsed_args$args[1]

cell.metadata.file <- read.table("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/9.Splice_pipeline_example_run/CH502_DNMT3A/output_files/leafcutter_outputs/CH502.metadata.txt")


table(cell.metadata.file$Genotype_Approx_Match_No_Gene)
cell.metadata.file[which(cell.metadata.file[,"Genotype_Approx_Match_No_Gene"] == "WT"),]$BC

ncol(cell.metadata.file)

celltype.column.name <- parsed_args$args[3]
celltype <- parsed_args$args[4]
intron.meta.data.file <- parsed_args$args[5]

```

#What is the dPSI between the DNMT3A sample and our bulk sample
```{r}

```



