
##Script to generate all final figures

## -------------------------- Load in all libraries -------------------------- ##
```{r}
library(ggplot2)
library(ggrepel)
library(ggpubr)
```


## -------------------------- VAF GoT vs MutFrac Correlation -------------------------- ##

```{r }
##Create a table with data
vaf.mut.frac <- as.data.frame(matrix(ncol=2,nrow=8))

colnames(vaf.mut.frac) <- c("VAF","MutFrac/2")
#rownames(vaf.mut.frac) <- c("MDSP1","MDSP2","MDSP3","CH259","CH305","MDSP4","MDSP5","MDSP6")
rownames(vaf.mut.frac) <- c("MDSP1","MDSP2","MDSP3","CH01","CH02","MDS01","MDS02","MDS03")

vaf.mut.frac$VAF <- c(0.11,0.32,0.21,0.221,0.154,0.395,0.376,0.386)
vaf.mut.frac$`MutFrac/2`<- c(0.18,0.30,0.26,0.14,0.17,0.316,0.398,0.413)
vaf.mut.frac$MutFrac <- vaf.mut.frac$`MutFrac/2`*2

vaf.mut.frac

##Generate Scatterplot
ggscatter(vaf.mut.frac, x = "VAF", y = "MutFrac/2", add = "reg.line", add.params = list(color = "blue", fill = "lightgray"), conf.int = TRUE,  cor.coef = TRUE, cor.coeff.args = list(method = "pearson", label.x = 0.25, label.sep = "\n"), xlab = "VAF", ylab = "MutFrac / 2", label = rownames(vaf.mut.frac), repel = TRUE)

pdf('/gpfs/commons/home/pchamely/vaf.mut.frac.1.pdf', width = 10, height = 7)
ggscatter(vaf.mut.frac[(4:8),], x = "VAF", y = "MutFrac/2", add = "reg.line", add.params = list(color = "blue", linetype = "dashed" , fill = "lightgray"), conf.int = TRUE,  cor.coef = TRUE, cor.coeff.args = list(method = "pearson", label.x = 0.15,label.y = 0.45 ,label.sep = "\n"), xlab = "Bulk VAF in unsorted stem cell grafts", ylab = "Observed GoT VAF", label = rownames(vaf.mut.frac[(4:8),]), repel = TRUE)
dev.off()
```



## -------------------------- Comparing Bulk refernce dPSIs to our observed dPSIs-------------------------- ##


```{r}

```



## -------------------------- Denisty Plot showing distance of cryptic sites -------------------------- ##

```{r}

##Load in the table with Differential Usage (can adjust for either MDS/CH and 1WT vs 2WT)
diff_usage_info <- read.table("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/7.Permutation_junction_output/6.Comb_patients_2WT/5_read_threshold/MDS_P5_P6/logOR_within_cell_type_ALT_3P_Junctions_with_threshold_info.txt")

##Grab the relevant infomation for all cryptic junctions 
density_plot_info <- diff_usage_info[which(diff_usage_info$Final_Verdict =="Cryptic_threeprime"), c("obs.wt","obs.mut","total.reads.per.junction","dPSI","pvalue","threep_distance")]

##Label events as being more highly used in WT vs MUT based on the dPSI
density_plot_info$Genotype <- NA
density_plot_info[which(density_plot_info$dPSI >= 2),]$Genotype <- "MUT"
density_plot_info[which(density_plot_info$dPSI <= -2),]$Genotype <- "WT"

##Looking at the distance of cryptic sites for both significant MUT and WT events 
ggplot(density_plot_info[which(!is.na(density_plot_info$Genotype) & density_plot_info$pvalue <= 0.05),], aes(x=threep_distance, color = Genotype)) + geom_density() + theme(legend.position="none", axis.line=element_blank(), axis.text=element_blank()) + theme_classic(base_size = 17) +scale_color_manual(values=c("red","blue"))  + ylim(0,0.07) 

##Looking at the distance of cryptic sites for only significant  MUT events
ggplot(density_plot_info[which(!is.na(density_plot_info$Genotype) & density_plot_info$pvalue <= 0.05 & density_plot_info$Genotype =="MUT" ),], aes(x=threep_distance, color = Genotype)) + geom_density() + theme(legend.position="none", axis.line=element_blank(), axis.text=element_blank()) + theme_classic(base_size = 17) +scale_color_manual(values=c("red","blue"))  + ylim(0,0.07)

#Generate final PDF
pdf('/gpfs/commons/home/pchamely/leafcutter_scripts/6.cryptic3pSS_analyses/density.plots/MDS.untreated.combined.density.plot.MUTonly.pdf', width = 10, height = 6)
ggplot(density_plot_info[which(!is.na(density_plot_info$Genotype) & density_plot_info$pvalue <= 0.05 & density_plot_info$Genotype =="MUT" ),], aes(x=threep_distance, color = Genotype)) + geom_density() + theme(legend.position="none", axis.line=element_blank(), axis.text=element_blank()) + theme_classic(base_size = 17) +scale_color_manual(values=c("red","blue"))  + ylim(0,0.07)
dev.off()
```


## -------------------------- MDS cell type specific splicing heatmap - Figure 3C -------------------------- ##


##2UMI threshold
```{r}
ep = read.table("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/7.Permutation_junction_output/7.Comb_patients_ind_celltype_merge_counts_2WT/MDS_P5_P6/EP/logOR_within_cell_type_ALT_3P_Junctions_with_threshold_info.txt")
ep.cryptic = ep %>% filter(Final_Verdict == "Cryptic_threeprime")

hspc = read.table("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/7.Permutation_junction_output/7.Comb_patients_ind_celltype_merge_counts_2WT/MDS_P5_P6/HSPC/logOR_within_cell_type_ALT_3P_Junctions_with_threshold_info.txt")
hspc.cryptic = hspc %>% filter(Final_Verdict == "Cryptic_threeprime")

imp = read.table("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/7.Permutation_junction_output/7.Comb_patients_ind_celltype_merge_counts_2WT/MDS_P5_P6/IMP/logOR_within_cell_type_ALT_3P_Junctions_with_threshold_info.txt")
imp.cryptic = imp %>% filter(Final_Verdict == "Cryptic_threeprime")

mep = read.table("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/7.Permutation_junction_output/7.Comb_patients_ind_celltype_merge_counts_2WT/MDS_P5_P6/MEP/logOR_within_cell_type_ALT_3P_Junctions_with_threshold_info.txt")
mep.cryptic = mep %>% filter(Final_Verdict == "Cryptic_threeprime")

mkp = read.table("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/7.Permutation_junction_output/7.Comb_patients_ind_celltype_merge_counts_2WT/MDS_P5_P6/MkP/logOR_within_cell_type_ALT_3P_Junctions_with_threshold_info.txt")
mkp.cryptic = mkp %>% filter(Final_Verdict == "Cryptic_threeprime")

np = read.table("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/7.Permutation_junction_output/7.Comb_patients_ind_celltype_merge_counts_2WT/MDS_P5_P6/NP/logOR_within_cell_type_ALT_3P_Junctions_with_threshold_info.txt")
np.cryptic = np %>% filter(Final_Verdict == "Cryptic_threeprime")


##Filter on significant events with at least 5 reads in either WT and MUT covering that junction

ep.cryptic$junction = paste(ep.cryptic$gene, ep.cryptic$intron_junction, sep = ":")
hspc.cryptic$junction = paste(hspc.cryptic$gene, hspc.cryptic$intron_junction, sep = ":")
imp.cryptic$junction = paste(imp.cryptic$gene, imp.cryptic$intron_junction, sep = ":")
mep.cryptic$junction = paste(mep.cryptic$gene, mep.cryptic$intron_junction, sep = ":")
mkp.cryptic$junction = paste(mkp.cryptic$gene, mkp.cryptic$intron_junction, sep = ":")
np.cryptic$junction = paste(np.cryptic$gene, np.cryptic$intron_junction, sep = ":")


ep.cryptic.heat = ep.cryptic %>% filter(pvalue <= 0.05) %>% filter(three.wt.cluster.cov >= 5) %>% filter(three.mut.cluster.cov >= 5)  %>% filter(dPSI >= 2) %>% select(junction, gene ,wt.psi, mut.psi, dPSI, pvalue)
#ep.cryptic.heat$celltype = "EP"

hspc.cryptic.heat = hspc.cryptic %>% filter(pvalue <= 0.05) %>% filter(three.wt.cluster.cov >= 5) %>% filter(three.mut.cluster.cov >= 5) %>% filter(dPSI >= 2) %>% select(junction, gene, wt.psi, mut.psi, dPSI, pvalue)
#hspc.cryptic.heat$celltype = "HSPC"

imp.cryptic.heat = imp.cryptic %>% filter(pvalue <= 0.05) %>% filter(three.wt.cluster.cov >= 5) %>% filter(three.mut.cluster.cov >= 5) %>% filter(dPSI >= 2) %>% select(junction, gene, wt.psi, mut.psi, dPSI, pvalue)
#imp.cryptic.heat$celltype = "IMP"

mep.cryptic.heat = mep.cryptic %>% filter(pvalue <= 0.05) %>% filter(three.wt.cluster.cov >= 5) %>% filter(three.mut.cluster.cov >= 5) %>% filter(dPSI >= 2) %>% select(junction, gene, wt.psi, mut.psi, dPSI, pvalue)
#mep.cryptic.heat$celltype = "MEP"

mkp.cryptic.heat = mkp.cryptic %>% filter(pvalue <= 0.05) %>% filter(three.wt.cluster.cov >= 5) %>% filter(three.mut.cluster.cov >= 5) %>% filter(dPSI >= 2) %>% select(junction, gene, wt.psi, mut.psi, dPSI, pvalue)
#mkp.cryptic.heat$celltype = "MkP"

np.cryptic.heat = np.cryptic %>% filter(pvalue <= 0.05) %>% filter(three.wt.cluster.cov >= 5) %>% filter(three.mut.cluster.cov >= 5)  %>% filter(dPSI >= 2) %>% select(junction, gene, wt.psi, mut.psi, dPSI, pvalue)
#np.cryptic.heat$celltype = "NP"

#all.cell.types = rbind(ep.cryptic.heat, hspc.cryptic.heat, imp.cryptic.heat, mep.cryptic.heat, mkp.cryptic.heat, np.cryptic.heat)
#all.cell.types

colnames(ep.cryptic.heat) = c("junction","gene", "EP.wt.psi", "EP.mut.psi", "EP.dPSI", "EP.pvalue")
colnames(hspc.cryptic.heat) = c("junction","gene", "HSPC.wt.psi", "HSPC.mut.psi", "HSPC.dPSI", "HSPC.pvalue")
colnames(imp.cryptic.heat) = c("junction", "gene","IMP.wt.psi", "IMP.mut.psi", "IMP.dPSI", "IMP.pvalue")
colnames(mep.cryptic.heat) = c("junction", "gene","MEP.wt.psi", "MEP.mut.psi", "MEP.dPSI", "MEP.pvalue")
colnames(mkp.cryptic.heat) = c("junction","gene", "MkP.wt.psi", "MkP.mut.psi", "MkP.dPSI", "MkP.pvalue")
colnames(np.cryptic.heat) = c("junction", "gene","NP.wt.psi", "NP.mut.psi", "NP.dPSI", "NP.pvalue")

all.cell.types.filt = Reduce(full_join, list(np.cryptic.heat, imp.cryptic.heat, hspc.cryptic.heat,mkp.cryptic.heat, mep.cryptic.heat,ep.cryptic.heat))

rownames(all.cell.types.filt) = make.names(all.cell.types.filt$gene, unique = TRUE)

all.cell.types.filt
```

#Generate heatmap
```{r fig.height=30, fig.width=20}
all.cell.types.heat = all.cell.types.filt[,grep("dPSI", colnames(all.cell.types.filt))]
all.cell.types.heat = as.matrix(all.cell.types.heat)
all.cell.types.heat[is.na(all.cell.types.heat)] = 0

pheatmap(all.cell.types.heat, scale = "row", cluster_cols = FALSE, cutree_rows = 6, fontsize_row = 8, fontsize_col = 20, show_rownames = T, labels_col = c("NP","IMP","HSPC","MkP","MEP","EP"), border_color=NA,color = colorRampPalette(c("light grey", "white", "red"))(25))

pheatmap(all.cell.types.heat, scale = "row", cluster_cols = FALSE, cluster_rows = F, fontsize_row = 5, fontsize_col = 20, show_rownames = T, labels_col = c("NP","IMP","HSPC","MkP","MEP","EP"), border_color=NA, color = colorRampPalette(c("light grey", "white", "red"))(25))


pdf('/gpfs/commons/groups/landau_lab/SF3B1_splice_project/scripts/FinalFigures/MDS.per.celltype.sig.cryptic3p.dPSI.2.cluster.thresh.5.pdf', width = 10, height = 20)
pheatmap(all.cell.types.heat, scale = "row", cluster_cols = FALSE, cutree_rows = 6, fontsize_row = 5, fontsize_col = 20, show_rownames = T, labels_col = c("NP","IMP","HSPC","MkP","MEP","EP"), border_color=NA, color = colorRampPalette(c("light grey", "white", "red"))(25))
dev.off()

```



##Add colors for pathways

```{r}

#MetabolismRNA <- c("CHTOP","ERCC3","IGF2BP2","TPR","TRNT1","URM1","ZBTB8OS","NUP98","PRPF38A","SMNDC1","THOC1")
            
#Splicing <- c("GTF2F2","PCBP2","SF3B1","SNW1","SRSF11","U2AF1L4","USP39","HNRNPUL1","SYF2")

RNA_processing <- c("CHTOP","ERCC3","IGF2BP2","TPR","TRNT1","URM1","ZBTB8OS","NUP98","PRPF38A","SMNDC1","THOC1","GTF2F2","PCBP2","SF3B1","SNW1","SRSF11","U2AF1L4","USP39","HNRNPUL1","SYF2" )

CellCycle<-c("ACD","DNAJC3","MCM4","NKTR","ORC1","TRMT2A","CENPT","CEP63","HAUS4","HAUS5","NUP98","PCM1","SMC2")

EP_diff<-c("GATA1","CD36", "PPOX","GYPA","B9D1","CAPRIN2","CD36","CDCA2","CTSE","ERI2","EXO1","FOXRED1","FXYD3","GTF2F2","IGF2BP2","ISOC2","KLHL18","KRT13","MCM4","MECR","MPDU1","ORC1","PARVB","SNW1","SSX2IP","STAU1","TBPL1","TTK","WAPL","ADD2","ASH2L","BCL11A","CD164",
                 "CNN2",
                 "CTNNB1",
                 "FOG1",
                 "GFI1B",
                 "GLRX5",
                 "HIPK1",
                 "IKZF1",
                 "MKNK2",
                 "PICALM",
                 "RAF1",
                 "ROGD1",
                 "RUNX1",
                 "SP3",
                 "SPI1",
                 "STAT2",
                 "TAL1",
                 "DICER1",
                 "PPP2CA",
                 "PPP2R1A",
                 "ROD1",
                 "SOX9",
                 "UNCX",
                 "GATA2",
                 "NKX2-5",
                 "TAZ",
                 "ZBTB16",
                 "CDK6",
                 "KIT",
                 "RPA1",
                 "SP1",
                 "TACC3",
                 "FES",
                 "LYN",
                 "SIK1",
                 "SPOCK2",
                 "EBP",
                 "SOD2",
                 "SOX6",
                 "TIPARP",
                 "TTC7A",
                 "TXNRD2",
                 "STK3",
                 "STK4",
                 "THOC5",
                 "ACIN1",
                 "ADD1",
                 "ALAS2",
                 "CITED2",
                 "KLF1",
                 "KLF2",
                 "MAEA",
                 "MLL5",
                 "RCOR1",
                 "RPS14",
                 "RPS19",
                 "SFXN1",
                 "SLC25A38",
                 "TIMP1",
                 "ZFPM1",
                 "STAT5A",
                 "EPOR",
                 "EPB42",
                 "MED1",
                 "RPS6",
                 "SLC11A2",
                 "SRF",
                 "BCL6",
                 "BPGM",
                 "DYRK3",
                 "ERCC2",
                 "HCLS1",
                 "HIPK2",
                 "JAK2",
                 "JMJD6",
                 "NCKAP1L",
                 "PTPN6",
                 "TRIM10")
Heme_metabolism<- c("ALAD","BLVRB","CPOX","FECH","FXN","HMBS","SLC25A39",
                 "SPTA1",
                 "TMEM14C",
                 "TSPO",
                 "UROD",
                 "UROS",
                 "BLVRA",
                 "ALAS1",
                 "EIF2AK1",
                 "HMOX2",
                 "TFRC",
                 "CYBRD1",
                 "CIAO1",
                 "FBXL5",
                 "FTL",
                 "ISCA1",
                 "ISCA2",
                 "ISCU",
                 "LCN2",
                 "SLC11A1",
                 "SLC22A17",
                 "SLC25A28",
                 "SLC25A37",
                 "ID2",
                 "RB1")

Oxygen<-c("ACTN4","ADRBK1","AGTRAP","ATPIF1", "ARNT","CAT", "CREBBP","CRYAA","CST3","DUSP1","ECE1",
                           "ENG",
                           "EP300",
                           "GAB1",
                           "GPX1",
                           "GPX4",
                           "HIF1A",
                           "HSP90B1",
                           "HYOU1",
                           "LIAS",
                           "MAP2K1",
                           "MLH1",
                           "MMP8",
                           "MMP9",
                           "MT3",
                           "NAPRT1",
                           "NDUFA12",
                           "NDUFA6",
                           "NDUFS2",
                           "NDUFS8",
                           "NOL3",
                           "NPPC",
                           "PEBP1",
                           "PML",
                           "PNKP",
                           "PPP1R15B",
                           "PRDX3",
                           "PRDX6",
                           "PRNP",
                           "SCAP",
                           "SHC1",
                           "STAT5B",
                           "STK25",
                           "TSC2",
                           "TXNIP",
                           "UBQLN1",
                           "UCP2",
                           "USF1",
                           "XRCC1",
                           "WRN",
                           "BTG3",
                           "DDIT3",
                           "ERCC3",
                           "NUDT1",
                           "SMAD3",
                           "UCN",
                           "ERCC1",
                           "PKLR",
                           "TXN2",
                           "LONP1",
                           "EPX",
                           "GCLC",
                           "OXSR1",
                           "PAK1",
                           "PRDX2",
                           "PRDX5",
                           "SMAD4",
                           "TGFB1",
                           "ADM",
                           "EGLN1",
                           "EGLN2",
                           "IPCEF1",
                           "MSRA",
                           "MTF1",
                           "NFKB1",
                           "NR4A2",
                           "NRF1",
                           "PARK7",
                           "PKM2",
                           "PTGS2",
                           "PTK2B",
                           "RGS14",
                           "SERPINA1",
                           "SIRT1",
                           "THBS1",
                           "TNFRSF1A",
                           "TPM4",
                           "UBE2B",
                           "VEGFA",
                           "VNN1")






library(ComplexHeatmap)
all.cell.types.filt$col <- "NA"
all.cell.types.filt[which(all.cell.types.filt$gene %in% Oxygen),]$col <- "black"
#all.cell.types.filt[which(all.cell.types.filt$gene %in% Splicing),]$col <- "red"
#all.cell.types.filt[which(all.cell.types.filt$gene %in% MetabolismRNA),]$col <- "blue"
all.cell.types.filt[which(all.cell.types.filt$gene %in% RNA_processing),]$col <- "red"
all.cell.types.filt[which(all.cell.types.filt$gene %in% CellCycle),]$col <- "red4"
all.cell.types.filt[which(all.cell.types.filt$gene %in% Heme_metabolism),]$col <- "green3"
all.cell.types.filt[which(all.cell.types.filt$gene %in% EP_diff),]$col <- "orange"


hm <- t(scale(t(all.cell.types.heat), center = TRUE, scale = TRUE))

rownames(all.cell.types.filt) <- 1:nrow(all.cell.types.filt)

save(all.cell.types.filt, hm, version=2,file="/gpfs/commons/groups/landau_lab/SF3B1_splice_project/scripts/FinalFigures/for.cell.type.splicing.heatmap.Rdata")

##--------- Generate it in local R with the following code ---------##


to.mark <- as.numeric(rownames(all.cell.types.filt[which(all.cell.types.filt$col != "NA"),]))
genes <- all.cell.types.filt[which(all.cell.types.filt$col != "NA"),]$gene
colors <- all.cell.types.filt[which(all.cell.types.filt$col != "NA"),]$col

ha = rowAnnotation(foo = anno_mark(at = to.mark, labels = genes, labels_gp = gpar(fontsize = 8, col = colors)))

pdf('Desktop/MDS.cell.type.splicing.annotated.pdf', width = 10, height = 20)
Heatmap(hm, split=6,cluster_columns = F, show_row_names = FALSE ,col = colorRampPalette(c("light grey", "white", "red"))(25), name = "z score", right_annotation = ha)
dev.off()

```

##Testing that genes are there
```{r}
mep = read.table("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/7.Permutation_junction_output/7.Comb_patients_ind_celltype_merge_counts_2WT/MDS_P5_P6/MEP/logOR_within_cell_type_ALT_3P_Junctions_with_threshold_info.txt")
mep[which(mep$gene == "SLC25A39"),]
mep = read.table("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/7.Permutation_junction_output/5.Comb_patients_ind_celltype_merge_counts/MDS_P5_P6/MEP/logOR_within_cell_type_ALT_3P_Junctions_with_threshold_info.txt")
mep[which(mep$gene == "SLC25A39"),]

gene <- "MCM4"
ep[which(ep$gene == gene),]
ep.cryptic.heat[which(ep.cryptic.heat$gene == gene),]
all.cell.types.filt[which(all.cell.types.filt$gene == gene),]
```



## -------------------------- BAX MDS/CH comparison - Figure 5E -------------------------- ##


##Load all the data and generate the cell type fraction plots
```{r}

## ------- MDS ------- ##
load("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/16.Pseudotime.Window/MDS.output/MDSP5.MDSP6.merged.info.no.sig.thresh.bucket.400.usingCITE.ct.sub.onlyMut.wCTcounts.Rdata")
load("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/16.Pseudotime.Window/junctions.to.use.MDS.noSigThresh.ct.sub.Rdata")

mds.rownames <- junctions.of.interest.final

extracted_cols <- lapply(win.mutpsi, function(x) x[,"mut.psi"]) #note the added comma!
mds.mut.psi.per.window <- as.data.frame(do.call(cbind, extracted_cols))
colnames(mds.mut.psi.per.window) = as.character(1:ncol(mds.mut.psi.per.window))
rownames(mds.mut.psi.per.window) <- mds.rownames

annotation_col = data.frame(ep.fraction = as.numeric(paste(ct.fractions.2["EP",])), 
                            mep.fraction = as.numeric(paste(ct.fractions.2["MEP",])),
                            mkp.fraction = as.numeric(paste(ct.fractions.2["MkP",])),
                            hsc.fraction = as.numeric(paste(ct.fractions.2["HSPC",])),
                            mean.CD71.exp = ADT.CD71.per.window)

rownames(annotation_col) = paste(1:nrow(annotation_col))

cell.freq <- melt(t(annotation_col[,1:4]),var=c("CellType",'window'))
mycolors <- c("#980043","#4B0082","#DD3497","#313695")

pdf("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/scripts/FinalFigures/MDS.perc.ct.per.bin.mut.only.8bins.pdf", width = 20, height = 2)
ggplot(cell.freq, aes(x = window, y = value, fill = CellType)) + geom_col(position = "fill") + scale_fill_manual(values=mycolors) + theme_void() + theme(legend.key.size = unit(0.5, 'cm'), #change legend key size
        legend.key.height = unit(0.5, 'cm'), #change legend key height
        legend.key.width = unit(0.5, 'cm'), #change legend key width
        legend.title = element_text(size=10), #change legend title font size
        legend.text = element_text(size=8)) #change legend text font size
dev.off()


## ------- CH ------- ##
load("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/16.Pseudotime.Window/CH.output/CH259.CH305.merged.info.sig.in.any.dPSI2.bucket60.onlyMUT.wCTcounts.Rdata")
load("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/16.Pseudotime.Window/junctions.to.use.CH.dPSI2.Rdata")

ch.rownames <- junctions.of.interest.final

extracted_cols <- lapply(win.mutpsi, function(x) x[,"mut.psi"]) #note the added comma!
ch.mut.psi.per.window <- as.data.frame(do.call(cbind, extracted_cols))
colnames(ch.mut.psi.per.window) = as.character(1:ncol(ch.mut.psi.per.window))
rownames(ch.mut.psi.per.window) <- ch.rownames


annotation_col = data.frame(ep.fraction = as.numeric(paste(ct.fractions.2["EP",])), 
                            mep.fraction = as.numeric(paste(ct.fractions.2["MEP",])), 
                            imp.fraction = as.numeric(paste(ct.fractions.2["IMP",])),
                            hsc.fraction = as.numeric(paste(ct.fractions.2["HSPC",])),  
                            mean.pseudotime = pseudotime.mean)

rownames(annotation_col) = paste(1:nrow(annotation_col))

cell.freq <- melt(t(annotation_col[,1:4]),var=c("CellType",'window'))
mycolors <- c("#980043","#4B0082","#045A8D","#313695")

pdf("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/scripts/FinalFigures/CH.perc.ct.per.bin.mut.only.8bins.pdf", width = 20, height = 2)
ggplot(cell.freq, aes(x = window, y = value, fill = CellType)) + geom_col(position = "fill") + scale_fill_manual(values=mycolors) + theme_void() + theme(legend.key.size = unit(0.5, 'cm'), #change legend key size
        legend.key.height = unit(0.5, 'cm'), #change legend key height
        legend.key.width = unit(0.5, 'cm'), #change legend key width
        legend.title = element_text(size=10), #change legend title font size
        legend.text = element_text(size=8)) #change legend text font size
dev.off()

```


##Generate the mut.psi plots for each

```{r}
junction <- "BAX:chr19:48960914:48961482:+"

df <- as.data.frame(matrix(ncol=2, nrow=ncol(ch.mut.psi.per.window)))
colnames(df) <- c("MDS.psi","CH.psi")

df$MDS.psi <- as.numeric(paste(mds.mut.psi.per.window[junction,]))
df$CH.psi <- as.numeric(paste(ch.mut.psi.per.window[junction,]))
df$bin.num <- rownames(df)

Color <- "steelblue"

#MDS
m <- floor(min(df$MDS.psi))
df$MDS.psi <- df$MDS.psi - m

pdf("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/scripts/FinalFigures/MDS.BAX.mutpsi.8bins.pdf", width = 12, height = 8)
ggplot(df, mapping = aes(x, y, group =1)) + 
  geom_bar(data = data.frame(x = df$bin.num, y = df$MDS.psi), stat = 'identity', fill=Color, alpha=0.3) +
  stat_smooth(aes(x = bin.num, y = MDS.psi), color=Color, method = "loess", formula = y ~ x, se = F, span=1) +
  scale_y_continuous(
    name = "MDS.mut.psi",
    breaks=seq(0,50),
    labels=(seq(0,50) + m)) +
  theme_classic() +
  labs(x="bin",
       title = paste0("MDS mut.psi for cryptic junction - ", junction)) +
  theme(plot.title = element_text(vjust = 2.5,hjust = 0.5, size = 13), 
        axis.title.y = element_text(color=Color  , size=12)) 
dev.off()


#CH
m <- floor(min(df$CH.psi))
df$CH.psi <- df$CH.psi - m

pdf("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/scripts/FinalFigures/CH.BAX.mutpsi.8bins.pdf", width = 12, height = 8)
ggplot(df, mapping = aes(x, y, group =1)) + 
  geom_bar(data = data.frame(x = df$bin.num, y = df$CH.psi), stat = 'identity', fill=Color, alpha=0.3) +
  stat_smooth(aes(x = bin.num, y = CH.psi), color=Color, method = "loess", formula = y ~ x, se = F, span=1) +
  scale_y_continuous(
    name = "CH.mut.psi",
    breaks=seq(0,50),
    labels=(seq(0,50) + m)) +
  theme_classic() +
  labs(x="bin",
       title = paste0("CH mut.psi for cryptic junction - ", junction)) +
  theme(plot.title = element_text(vjust = 2.5,hjust = 0.5, size = 13), 
        axis.title.y = element_text(color=Color, size=12)) 

dev.off()

```




##--------------------------------------------------- Supplementary Figure MDS ---------------------------------------------------##

##Generate PDF for MDS paer patient breakdown:


```{r}
load("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/4.Integrated_seurat_objs/mds.sample.integrated.p4.5.6.in.progress_2_withPseudotime.RData")
#loaded as seurat
mds.samples.integrated <- seurat 

#Idents(samples.integrated) # put it back to this after generating plots
Idents(mds.samples.integrated) <- "patient.id"
Idents(mds.samples.integrated) <- "orig.ident"

plots <- list()

patient_ids <- unique(mds.samples.integrated@meta.data$orig.ident)
patient_ids
##Getting the Seurat Pallete (change Idents to suite)
require(scales)
# Create vector with levels of object@ident
identities <- levels(mds.samples.integrated@active.ident)
# Create vector of default ggplot2 colors
my_color_palette <- hue_pal()(length(identities))

for (i in 1:length(patient_ids)){
  #plots[[i]] <- DimPlot(mds.samples.integrated , reduction = "umap", label = TRUE, cells.highlight = WhichCells(mds.samples.integrated, idents = c(patient_ids[i]))) + scale_color_manual(labels = c("Unselected", patient_ids[i]), values = c("grey", my_color_palette[i]))
  plots[[patient_ids[i]]] <- ggplot(mds.samples.integrated@meta.data, aes(x = UMAP_1, y = UMAP_2)) + 
  geom_point(color = "grey", pch = 19, size = 3) + 
  geom_point(data=mds.samples.integrated@meta.data[which(mds.samples.integrated@meta.data$orig.ident == patient_ids[i]), ], aes(x=UMAP_1, y=UMAP_2), color="black", fill="#8D38DA", pch = 21, size=3, stroke = 0.3)+ ggtitle(patient_ids[i]) +
  #geom_point(data=mds.samples.integrated@meta.data[which(mds.samples.integrated@meta.data$patient.id == patient_ids[i]), ], aes(x=UMAP_1, y=UMAP_2), color="black", fill="#8D38DA", pch = 21, size=3, stroke = 0.3)+
  theme_void() +  labs(fill = "Patient ID") 
  
}


plots$MDS_P4
Idents(mds.samples.integrated) <- "seurat_clusters"


pdf(file = "/gpfs/commons/groups/landau_lab/SF3B1_splice_project/scripts/FinalFigures/MDS.UMAP.per.patient.P5.split.pdf", width = 20, height = 20)
grid.arrange(grobs = list(plots$MDS_P6, plots$MDS_P5_1,plots$MDS_P5_2,plots$MDS_P4), ncol = 2)
dev.off()




```




##Generate Heatmap showing marker gene expression per Cell Type
```{r}
library(Seurat)
load("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/4.Integrated_seurat_objs/mds.sample.integrated.p4.5.6.in.progress_2_withPseudotime.RData")
#loaded as seurat
mds.samples.integrated <- seurat 
mds.samples.integrated@meta.data$Cell.Assignment_adj <- factor(mds.samples.integrated@meta.data$Cell.Assignment_adj, levels = c("Mat_Mono","DC","MonoDC","NP","IMP","HSPC","MkP","MEP","EP","E/B/M","PreB","Bcells","Tcells"))

Idents(mds.samples.integrated) <- "Cell.Assignment_adj"
DefaultAssay(mds.samples.integrated) <- "RNA"

mds.samples.integrated.CellTypeMarkers <- FindAllMarkers(mds.samples.integrated, only.pos = TRUE, min.pct = 0.2, assay = "RNA")
save(mds.samples.integrated.CellTypeMarkers, file="/gpfs/commons/groups/landau_lab/SF3B1_splice_project/scripts/FinalFigures/mds.samples.integrated.CellTypeMarkers.Rdata")

top10_markers_ct <- mds.samples.integrated.CellTypeMarkers %>% group_by(cluster) %>% top_n(10, wt = avg_logFC)
top10_markers_ct <- top10_markers_ct %>% arrange(cluster)

##Use the RNA markers but look at the expression on the integrated object ?
#DefaultAssay(mds.samples.integrated) <- "integrated"

## Using the seurat heatmap funciton:
plot_hm <- DoHeatmap(mds.samples.integrated, features = top10_markers_ct$gene, label=FALSE) + scale_fill_gradient2(low = "blue", mid = "black", high = "red")

pdf(file = "/gpfs/commons/groups/landau_lab/SF3B1_splice_project/scripts/FinalFigures/MDS.CellType.Heatmap.pdf", width = 20, height = 14)
plot_hm
dev.off()


##Making a subset of seurat object to only plot some cell types on heatmap
mds.samples.integrated.sub <- subset(x = mds.samples.integrated, subset = Cell.Assignment_adj %in% c("Mat_Mono","NP","IMP","HSPC","MkP","MEP","EP"))

mds.samples.integrated.sub@meta.data

top10_markers_ct_sub <- top10_markers_ct[which(top10_markers_ct$cluster %in% c("Mat_Mono","NP","IMP","HSPC","MkP","MEP","EP")),]

Idents(mds.samples.integrated.sub) <- "Cell.Assignment_adj"
DefaultAssay(mds.samples.integrated.sub) <- "RNA"

plot_hm_sub <- DoHeatmap(mds.samples.integrated.sub, features = top10_markers_ct_sub$gene, label=F, group.colors=c("lightskyblue4","#006D2C","#045A8D","#313695","#DD3497","#4B0082","#980043")) + scale_fill_gradient2(low = "blue", mid = "black", high = "red")

pdf(file = "/gpfs/commons/groups/landau_lab/SF3B1_splice_project/scripts/FinalFigures/MDS.CellType.sub.Heatmap.pdf", width = 20, height = 14)
plot_hm_sub
dev.off()
```


###Lineage specific marker gene expression plots 
```{r fig.height=10, fig.width=10}

library(gridExtra)
#c Lineage Specific Genes/Modules ####
#cannonical_genes <- c("AVP", "HOPX", "AZU1" ,"MPO", "DNTT", "IRF8","PLEK", "ITGA2B" ,"CA1","BLVRB","CLC")
##Use GATA1???
#cannonical_genes <- c("AVP", "HOPX","MPO","IRF8","CD14","DNTT","PLEK","CA1","BLVRB","CLC","IL32","CD19")
#Only project markers fro major cell tyes
cannonical_genes <- c("AVP", "HOPX","MPO","IRF8","CD14","PLEK","CA1","BLVRB")

plots <- list()
for (i in cannonical_genes){
  plots[[i]] <- FeaturePlot(mds.samples.integrated, i, cols = c("#FDDA88", "#8D38DA"), order = TRUE) + theme(axis.line=element_blank(),axis.text.x=element_blank(), axis.text.y=element_blank(),axis.ticks=element_blank(), axis.title.x=element_blank(),axis.title.y=element_blank(), legend.position="none")
}

grid.arrange(grobs = plots, ncol = 3)

pdf(file = "/gpfs/commons/groups/landau_lab/SF3B1_splice_project/scripts/FinalFigures/MDS.Canonical.Markers.sub.pdf", width = 12, height = 9)
grid.arrange(grobs = plots, ncol = 4)
dev.off()
```


##Per Patient distribution for num gene and num UMIs 

```{r fig.height=4, fig.width=8}
##Num UMIs per patient
mds.samples.integrated@meta.data$orig.ident <- factor(mds.samples.integrated@meta.data$orig.ident, levels = c("MDS_P6","MDS_P5_1","MDS_P5_2","MDS_P4"))

pdf(file = "/gpfs/commons/groups/landau_lab/SF3B1_splice_project/scripts/FinalFigures/MDS.UMIs.per.patient.P5.split.pdf", width=10, height=4)
ggplot(mds.samples.integrated@meta.data, aes(x=orig.ident , y=nCount_RNA)) + stat_boxplot(geom='errorbar', linetype=2, width=0.5)+ geom_boxplot(outlier.shape = NA, colour = "black",coef = 0)  + coord_flip() + theme_bw() + theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) + scale_y_continuous(breaks = round(seq(0, max(mds.samples.integrated@meta.data$nCount_RNA), by = 10000),10000)) + ylab("Number of UMIs per cell") + xlab("Patient ID")
dev.off()


##Num Genes per patient
pdf(file = "/gpfs/commons/groups/landau_lab/SF3B1_splice_project/scripts/FinalFigures/MDS.Genes.per.patient.P5.split.pdf", width=10, height=4)
ggplot(mds.samples.integrated@meta.data, aes(x=orig.ident , y=nFeature_RNA)) + stat_boxplot(geom='errorbar', linetype=2, width=0.5)+ geom_boxplot(color="black", coef = 0, outlier.shape = NA)+ coord_flip() + theme_bw() +theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) + ylab("Number of Genes per cell") + xlab("Patient ID")
dev.off()

```



##--------------------------------------------------- Supplementary Figure CH ---------------------------------------------------##


##Generate PDF for CH per patient breakdown:

```{r}
library(Seurat)
load("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/4.Integrated_seurat_objs/ch.integrated.Rdata")
#loaded as samples.integrated

Idents(samples.integrated) <- "orig.ident"

plots <- list()

patient_ids <- unique(samples.integrated@meta.data$orig.ident)
patient_ids

##Getting the Seurat Pallete (change Idents to suite)
require(scales)
# Create vector with levels of object@ident
identities <- levels(samples.integrated@active.ident)
# Create vector of default ggplot2 colors
my_color_palette <- hue_pal()(length(identities))

for (i in 1:length(patient_ids)){
  plots[[patient_ids[i]]] <- ggplot(samples.integrated@meta.data, aes(x = UMAP_1, y = UMAP_2)) + 
  geom_point(color = "grey", pch = 19, size = 3) + 
  geom_point(data=samples.integrated@meta.data[which(samples.integrated@meta.data$orig.ident == patient_ids[i]), ], aes(x=UMAP_1, y=UMAP_2), color="black", fill="#8D38DA", pch = 21, size=3, stroke = 0.3)+ ggtitle(patient_ids[i]) +
  theme_void() +  labs(fill = "Patient ID") 
  
}


pdf(file = "/gpfs/commons/groups/landau_lab/SF3B1_splice_project/scripts/FinalFigures/CH.UMAP.per.patient.pdf", width = 22, height = 15)
grid.arrange(grobs = plots, ncol = 2)
dev.off()


```



##Generate Heatmap showing marker gene expression per Cell Type
```{r}
samples.integrated@meta.data$CellType <- factor(samples.integrated@meta.data$CellType, levels = c("NP","IMP","HSPC","MkP","MEP","EP","EBM","PreB"))


Idents(samples.integrated) <- "CellType"
DefaultAssay(samples.integrated) <- "RNA"

samples.integrated.CellTypeMarkers <- FindAllMarkers(samples.integrated, only.pos = TRUE, min.pct = 0.2, assay = "RNA")

save(samples.integrated.CellTypeMarkers, file="/gpfs/commons/groups/landau_lab/SF3B1_splice_project/scripts/FinalFigures/ch.samples.integrated.CellTypeMarkers.Rdata")

top10_markers_ct <- samples.integrated.CellTypeMarkers %>% group_by(cluster) %>% top_n(10, wt = avg_logFC)
top10_markers_ct <- top10_markers_ct %>% arrange(cluster)

top10_markers_ct

##Making a subset of seurat object to only plot some cell types on heatmap
samples.integrated.sub <- subset(x = samples.integrated, subset = CellType %in% c("NP","IMP","HSPC","MkP","MEP","EP"))

samples.integrated.sub@meta.data

top10_markers_ct_sub <- top10_markers_ct[which(top10_markers_ct$cluster %in% c("NP","IMP","HSPC","MkP","MEP","EP")),]

Idents(samples.integrated.sub) <- "CellType"
DefaultAssay(samples.integrated.sub) <- "RNA"

plot_hm_sub <- DoHeatmap(samples.integrated.sub, features = top10_markers_ct_sub$gene, label=F, group.colors=c("#006D2C","#045A8D","#313695","#DD3497","#4B0082","#980043")) + scale_fill_gradient2(low = "blue", mid = "black", high = "red")

pdf(file = "/gpfs/commons/groups/landau_lab/SF3B1_splice_project/scripts/FinalFigures/CH.CellType.sub.Heatmap.pdf", width = 20, height = 14)
plot_hm_sub
dev.off()


```


###Lineage specific marker gene expression plots 

###Lineage specific marker gene expression plots 
```{r fig.height=10, fig.width=10}

library(gridExtra)

#Only project markers for major cell tyes
cannonical_genes <- c("AVP", "HOPX","MPO","PLEK","CA1","BLVRB")

plots <- list()
for (i in cannonical_genes){
  plots[[i]] <- FeaturePlot(samples.integrated, i, cols = c("#FDDA88", "#8D38DA"), order = TRUE, reduction = "umap") + theme(axis.line=element_blank(),axis.text.x=element_blank(), axis.text.y=element_blank(),axis.ticks=element_blank(), axis.title.x=element_blank(),axis.title.y=element_blank(), legend.position="none")
}

grid.arrange(grobs = plots, ncol = 3)

pdf(file = "/gpfs/commons/groups/landau_lab/SF3B1_splice_project/scripts/FinalFigures/CH.Canonical.Markers.sub.pdf", width = 10, height = 10)
grid.arrange(grobs = plots, ncol = 3)
dev.off()
```



##Per Patient distribution for num gene and num UMIs 

```{r fig.height=4, fig.width=8}
##Num UMIs per patient
samples.integrated@meta.data$orig.ident <- factor(samples.integrated@meta.data$orig.ident, levels = c("CH_259","CH_305"))

pdf(file = "/gpfs/commons/groups/landau_lab/SF3B1_splice_project/scripts/FinalFigures/CH.UMIs.per.patient.pdf", width=10, height=4)
ggplot(samples.integrated@meta.data, aes(x=orig.ident , y=nCount_RNA)) + stat_boxplot(geom='errorbar', linetype=2, width=0.5)+ geom_boxplot(outlier.shape = NA, colour = "black", coef = 0)  +coord_flip() + theme_bw() + theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) + ylab("Number of UMIs per cell") + xlab("Patient ID") + ylim(0,8000)
dev.off()

#+ scale_y_continuous(breaks = round(seq(0, max(samples.integrated@meta.data$nCount_RNA), by = 10000),10000))

##Num Genes per patient
pdf(file = "/gpfs/commons/groups/landau_lab/SF3B1_splice_project/scripts/FinalFigures/CH.Genes.per.patient.pdf", width=10, height=4)
ggplot(samples.integrated@meta.data, aes(x=orig.ident , y=nFeature_RNA)) + stat_boxplot(geom='errorbar', linetype=2, width=0.5)+ geom_boxplot(color="black", coef = 0, outlier.shape = NA)+ coord_flip() + theme_bw() +theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) + ylab("Number of Genes per cell") + xlab("Patient ID") + ylim(0,4000)
dev.off()

```

