

##Looking into exon skipping 

```{r}
cryptic.3p.splicing.table <- read.table("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/9.Splice_pipeline_example_run/CH259/output_files/diff_transcript_output/merge_final_output/logOR_within_cell_type_ALT_3P_Junctions_with_threshold_info.txt")

#junctions.annotations <- read.csv("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/9.Splice_pipeline_example_run/CH259/output_files/leafcutter_outputs/CH259_output/CH259_all.introns.info.w.primary.annotations.txt")

##New Annotation
junctions.annotations <- read.csv("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/9.Splice_pipeline_example_run/CH259/output_files/leafcutter_outputs/CH259_output/CH259_all.introns.info.w.primary.annotations.w.skipping.csv")


junctions.annotations[which(junctions.annotations$gene == "XRN2"),]
```


```{r}
junctions.annotations[which(junctions.annotations$gene =="RPL22L1"),] %>% arrange(five_prime)
junctions.annotations[which(junctions.annotations$gene =="RPL22L1"),c("chr","start","end","strand_1","five_prime","three_prime","gene","verdict","relStart","relEnd","startClass","endClass")] %>% arrange(five_prime)
cryptic.3p.splicing.table[which(cryptic.3p.splicing.table$Final_Verdict =="Canonical" & cryptic.3p.splicing.table$pvalue <= 0.05),]
junctions.annotations[which(junctions.annotations$gene == gene & junctions.annotations$five_prime == "108216411" ),c("chr","start","end","strand_1","five_prime","three_prime","gene","verdict","relStart","relEnd","startClass","endClass")] %>% arrange(desc(three_prime))
```


##Adding exon skipping information
```{r}

cryptic.3p.splicing.table
rownames(cryptic.3p.splicing.table) <- cryptic.3p.splicing.table$intron_junction

junctions.annotations$intron.junction <- paste0(junctions.annotations$chr,":",junctions.annotations$start,":",junctions.annotations$end,":",junctions.annotations$strand)

junctions.annotations.sub <- junctions.annotations[which(junctions.annotations$intron.junction %in% intersect(cryptic.3p.splicing.table$intron_junction,junctions.annotations$intron.junction )),]

length(junctions.annotations.sub$intron.junction)
length(unique(junctions.annotations.sub$intron.junction))
nrow(cryptic.3p.splicing.table)

junctions.annotations.sub$strand_1 <- paste(junctions.annotations.sub$strand_1)
junctions.annotations.sub$strand <- paste(junctions.annotations.sub$strand)

list <- junctions.annotations.sub[duplicated(junctions.annotations.sub$intron.junction), ]$intron.junction

to_remove <- junctions.annotations.sub[which(junctions.annotations.sub$intron.junction %in% list & (junctions.annotations.sub$strand != junctions.annotations.sub$strand_1)),]$X

junctions.annotations.sub <- junctions.annotations.sub[which(!(junctions.annotations.sub$X %in% to_remove)),]
rownames(junctions.annotations.sub) <- junctions.annotations.sub$intron.junction

junctions.annotations.sub$num.exon.skips <- abs(junctions.annotations.sub$relStartExon_skipping - junctions.annotations.sub$relEndExon_skipping)

cryptic.3p.splicing.table$num.exon.skips <- junctions.annotations.sub[rownames(cryptic.3p.splicing.table),]$num.exon.skips
cryptic.3p.splicing.table
cryptic.3p.splicing.table[which(cryptic.3p.splicing.table$pvalue <= 0.05 & cryptic.3p.splicing.table$num.exon.skips >0),]
```


```{r}
cryptic.3p.splicing.table[which(cryptic.3p.splicing.table$pvalue <= 0.05 & cryptic.3p.splicing.table$num.exon.skips >0),] %>% arrange(desc(num.exon.skips))
```

```{r}
cryptic.3p.splicing.table[which(cryptic.3p.splicing.table$pvalue <= 0.05 & cryptic.3p.splicing.table$num.exon.skips >0),] %>% arrange(desc(num.exon.skips))
```

```{r}
cryptic.3p.splicing.table[which(cryptic.3p.splicing.table$pvalue <= 0.05 & cryptic.3p.splicing.table$num.exon.skips >0 & cryptic.3p.splicing.table$gene =="RPL32"),] %>% arrange(five_prime)
junctions.annotations.sub[which(junctions.annotations.sub$gene =="RPL32"),c("chr","start","end","strand_1","five_prime","three_prime","gene","verdict","relStart","relEnd","startClass","endClass")] %>% arrange(five_prime)
```

```{r}
junctions.annotations.sub
```


```{r}
junctions.annotations[which(junctions.annotations$gene == "SEPT6"),] %>% arrange(desc(relStartExon_skipping))

junctions.annotations[which(junctions.annotations$gene == "SEPT6" & junctions.annotations$end == 119629317),] %>% arrange(end)


junctions.annotations[which(junctions.annotations$gene == "SEPT6" & junctions.annotations$end == 119629317),c("chr","start","end","gene","verdict","relStartExon_skipping","relEndExon_skipping")] %>% arrange(start)
```


```{r}
gene <- "RPL32"
cryptic.3p.splicing.table[which(cryptic.3p.splicing.table$pvalue <= 0.05 & cryptic.3p.splicing.table$num.exon.skips >0 & cryptic.3p.splicing.table$gene ==gene),] %>% arrange(five_prime)
junctions.annotations.sub[which(junctions.annotations.sub$gene ==gene),c("chr","start","end","strand_1","five_prime","three_prime","gene","verdict","relStart","relEnd","startClass","endClass","relStartExon_skipping","relEndExon_skipping")] %>% arrange(five_prime)
```


```{r}
cryptic.3p.splicing.table[which(cryptic.3p.splicing.table$pvalue <= 0.05 & cryptic.3p.splicing.table$num.exon.skips >0 & cryptic.3p.splicing.table$gene =="MYL6B"),]

junctions.annotations.sub[which(junctions.annotations.sub$gene =="MYL6B"),c("chr","start","end","strand_1","five_prime","three_prime","gene","verdict","relStart","relEnd","startClass","endClass")] %>% arrange(five_prime)

```


```{r}
junctions.annotations.sub[which(junctions.annotations.sub$gene =="MYL6B"),]
```


```{r}
cryptic.3p.splicing.table[which(cryptic.3p.splicing.table$pvalue <= 0.05 & cryptic.3p.splicing.table$num.exon.skips > 0 & cryptic.3p.splicing.table$gene =="SEPT6"),] %>% arrange(five_prime)
junctions.annotations.sub[which(junctions.annotations.sub$gene =="SEPT6"),c("chr","start","end","strand_1","five_prime","three_prime","gene","verdict","relStart","relEnd","startClass","endClass")] %>% arrange(five_prime)
```

##Load in the annotation reference file to figure out what's happening 

```{r}
##Annotation reference:
annotation.ref <- read.table("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/ONT_Splice_Pipeline/bin/annotation_reference/leafviz_all_introns.bed.gz")

gene <- "RPL32"
gene <- "SEPT6"

gene <- "ITGA2B"

gene <- "DYNLL1"

gene <- "LRR1"


annotation.ref[which(annotation.ref$V4 == gene),]

#junctions.annotations.sub[which(junctions.annotations.sub$gene ==gene & junctions.annotations.sub$startClass == "main" & junctions.annotations.sub$endClass == "main"),] %>% arrange(five_prime)

#cryptic.3p.splicing.table[which(cryptic.3p.splicing.table$pvalue <= 0.05 & cryptic.3p.splicing.table$num.exon.skips >0 & cryptic.3p.splicing.table$gene ==gene),] %>% arrange(five_prime)

junctions.annotations[which(junctions.annotations$gene ==gene),] %>% arrange(five_prime)


junctions.annotations.sub[which(junctions.annotations.sub$gene ==gene),c("chr","start","end","strand_1","five_prime","three_prime","gene","verdict","relStart","relEnd","startClass","endClass")] %>% arrange(three_prime)
```


####MDS combined new annotations - overlappying info onto final results 

```{r}
cryptic.3p.splicing.table <- read.table("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/7.Permutation_junction_output/6.Comb_patients_2WT/5_read_threshold/MDS_P5_P6/logOR_within_cell_type_ALT_3P_Junctions_with_threshold_info.txt") 

junctions.annotations <- read.csv("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/9.Splice_pipeline_example_run/MDS_P5_P6_2WT/5_read_threshold/diff_transcript_combined_merge_counts_2WT/combined_metadata/annotations_MDS_combined.csv")

length(cryptic.3p.splicing.table$intron_junction)
length(unique(cryptic.3p.splicing.table$intron_junction))
list <- cryptic.3p.splicing.table[duplicated(cryptic.3p.splicing.table$intron_junction), ]$intron_junction

##remove duplicated intron junctions from cryptic splice table for now. 
cryptic.3p.splicing.table.sub <- cryptic.3p.splicing.table[which(!cryptic.3p.splicing.table$intron_junction %in% list),]

rownames(cryptic.3p.splicing.table.sub) <- cryptic.3p.splicing.table.sub$intron_junction


junctions.annotations$intron.junction <- paste0(junctions.annotations$chr,":",junctions.annotations$start,":",junctions.annotations$end,":",junctions.annotations$strand)

junctions.annotations.sub <- junctions.annotations[which(junctions.annotations$intron.junction %in% intersect(cryptic.3p.splicing.table.sub$intron_junction,junctions.annotations$intron.junction )),]

length(junctions.annotations.sub$intron.junction)
length(unique(junctions.annotations.sub$intron.junction))
nrow(cryptic.3p.splicing.table.sub)

##Ally already filtered out things with disagreeing stradn information

list <- junctions.annotations.sub[duplicated(junctions.annotations.sub$intron.junction), ]$intron.junction

junctions.annotations.final <- junctions.annotations.sub[!duplicated(junctions.annotations.sub$intron.junction),]

rownames(junctions.annotations.final) <- junctions.annotations.final$intron.junction
nrow(junctions.annotations.final)

junctions.annotations.final$num.exon.skips <- junctions.annotations.final$relStartExon_skipping - junctions.annotations.final$relEndExon_skipping

junctions.annotations.final
##Add in the exon skipping info
cryptic.3p.splicing.table.sub$tags_skipping <- junctions.annotations.final[rownames(cryptic.3p.splicing.table.sub),]$tags_skipping

cryptic.3p.splicing.table.sub$curTag_skipping <- junctions.annotations.final[rownames(cryptic.3p.splicing.table.sub),]$curTag_skipping


cryptic.3p.splicing.table.sub$relStartExon <- junctions.annotations.final[rownames(cryptic.3p.splicing.table.sub),]$relStartExon_skipping

cryptic.3p.splicing.table.sub$relEndExon <- junctions.annotations.final[rownames(cryptic.3p.splicing.table.sub),]$relEndExon_skipping

cryptic.3p.splicing.table.sub$num.exon.skips <- junctions.annotations.final[rownames(cryptic.3p.splicing.table.sub),]$num.exon.skips


##Make sure it's accurate

cryptic.3p.splicing.table.sub[which(cryptic.3p.splicing.table.sub$gene == "RPL32"),] %>% arrange(intron_junction)
junctions.annotations.final[which(junctions.annotations.final$gene == "RPL32"),] %>% arrange(intron.junction)

cryptic.3p.splicing.table.sub[which(cryptic.3p.splicing.table.sub$pvalue <= 0.05 & cryptic.3p.splicing.table.sub$num.exon.skips >0),] %>% arrange(desc(num.exon.skips))

write.table(cryptic.3p.splicing.table.sub, file="/gpfs/commons/groups/landau_lab/SF3B1_splice_project/7.Permutation_junction_output/6.Comb_patients_2WT/5_read_threshold/MDS_P5_P6/logOR_within_cell_type_ALT_3P_Junctions_with_threshold_and_exon.skipping_info.txt")
```

##Make sure all the information is consistent across patients 
```{r}
junction <- "chr7:55978395:55979762:+"
junctions.annotations.sub[which(junctions.annotations.sub$intron.junction ==junction),]
```


```{r}
table(cryptic.3p.splicing.table.sub$Final_Verdict)
```


##Trouble shoot with one example
```{r}

```


#MDS_combined ct outputs add exon skipping info

```{r}
junctions.annotations <- read.csv("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/9.Splice_pipeline_example_run/MDS_P5_P6_2WT/5_read_threshold/diff_transcript_combined_merge_counts_2WT/combined_metadata/annotations_MDS_combined.csv")

junctions.annotations$intron.junction <- paste0(junctions.annotations$chr,":",junctions.annotations$start,":",junctions.annotations$end,":",junctions.annotations$strand)
junctions.annotations
```

```{r}

ct <- "NP"

cryptic.3p.splicing.table <- read.table(paste0("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/7.Permutation_junction_output/7.Comb_patients_ind_celltype_merge_counts_2WT/MDS_P5_P6/",ct,"/logOR_within_cell_type_ALT_3P_Junctions_with_threshold_info.txt"))

length(cryptic.3p.splicing.table$intron_junction)
length(unique(cryptic.3p.splicing.table$intron_junction))

list <- cryptic.3p.splicing.table[duplicated(cryptic.3p.splicing.table$intron_junction), ]$intron_junction
#cryptic.3p.splicing.table[which(cryptic.3p.splicing.table$intron_junction %in% list), ] %>% arrange(intron_junction)
##remove duplicated intron junctions from cryptic splice table for now. 
cryptic.3p.splicing.table.sub <- cryptic.3p.splicing.table[which(!cryptic.3p.splicing.table$intron_junction %in% list),]

rownames(cryptic.3p.splicing.table.sub) <- cryptic.3p.splicing.table.sub$intron_junction

junctions.annotations.sub <- junctions.annotations[which(junctions.annotations$intron.junction %in% intersect(cryptic.3p.splicing.table.sub$intron_junction,junctions.annotations$intron.junction)),]

length(junctions.annotations.sub$intron.junction)
length(unique(junctions.annotations.sub$intron.junction))
nrow(cryptic.3p.splicing.table.sub)

##Ally already filtered out things with disagreeing stradn information
list <- junctions.annotations.sub[duplicated(junctions.annotations.sub$intron.junction), ]$intron.junction
#junctions.annotations.sub[which(junctions.annotations.sub$intron.junction %in% list),] %>% arrange(intron.junction)

junctions.annotations.final <- junctions.annotations.sub[!duplicated(junctions.annotations.sub$intron.junction),]
rownames(junctions.annotations.final) <- junctions.annotations.final$intron.junction
nrow(junctions.annotations.final)

junctions.annotations.final$num.exon.skips <- junctions.annotations.final$relStartExon_skipping - junctions.annotations.final$relEndExon_skipping

##Add in the exon skipping info
cryptic.3p.splicing.table.sub$num.exon.skips <- junctions.annotations.final[rownames(cryptic.3p.splicing.table.sub),]$num.exon.skips


##Make sure it's accurate
#cryptic.3p.splicing.table.sub[which(cryptic.3p.splicing.table.sub$gene == "RPL32"),] %>% arrange(intron_junction)
#junctions.annotations.final[which(junctions.annotations.final$gene == "RPL32"),] %>% arrange(intron.junction)

#cryptic.3p.splicing.table.sub[which(cryptic.3p.splicing.table.sub$pvalue <= 0.05 & cryptic.3p.splicing.table.sub$num.exon.skips >0),] %>% arrange(desc(num.exon.skips))



write.table(cryptic.3p.splicing.table.sub, file=paste0("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/7.Permutation_junction_output/7.Comb_patients_ind_celltype_merge_counts_2WT/MDS_P5_P6/",ct,"/logOR_within_cell_type_ALT_3P_Junctions_with_threshold_and_exon.skipping_info.txt"))



```

