

##Looking specifically at BAX - how do the mut PSI and wt PSI look first by discrete cell type
```{r}

```


##Compare use of canonical and cryptic for WT and MUT cells
```{r}
genotyping.table.file.path <- "/gpfs/commons/groups/landau_lab/SF3B1_splice_project/12.Bulk.PSI/merging.patient.data/MDSP5.MDSP6.merged.genotype.info.with.cell.type.txt"

genotyping.table <- read.table(genotyping.table.file.path)

genotyping.table[10000:13000,]
```


##Try using multiple genotyping threhsold for both MUT and WT cells
##Get the dPSIs per threhsold and the fishers exact test value for usage

```{r}
seurat@meta.data

ct <- "MEP"
table(seurat@meta.data[which(seurat@meta.data$Total.calls >= 2 & seurat@meta.data$Cell.Assignment_adj == ct & seurat@meta.data$orig.ident != "MDS_P4"),]$Genotype_1UMI)

table(seurat@meta.data[which(seurat@meta.data$Total.calls >= 3 & seurat@meta.data$Cell.Assignment_adj == ct & seurat@meta.data$orig.ident != "MDS_P4"),]$Genotype_1UMI)

table(seurat@meta.data[which(seurat@meta.data$Total.calls >= 4 & seurat@meta.data$Cell.Assignment_adj == ct & seurat@meta.data$orig.ident != "MDS_P4"),]$Genotype_1UMI)


ct <- "EP"

table(seurat@meta.data[which(seurat@meta.data$Total.calls >= 2 & seurat@meta.data$Cell.Assignment_adj == ct & seurat@meta.data$orig.ident != "MDS_P4"),]$Genotype_1UMI)

table(seurat@meta.data[which(seurat@meta.data$Total.calls >= 3 & seurat@meta.data$Cell.Assignment_adj == ct & seurat@meta.data$orig.ident != "MDS_P4"),]$Genotype_1UMI)

table(seurat@meta.data[which(seurat@meta.data$Total.calls >= 4 & seurat@meta.data$Cell.Assignment_adj == ct & seurat@meta.data$orig.ident != "MDS_P4"),]$Genotype_1UMI)

```


```{r}
psi.info.minUMI.1 <- read.csv("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/16.Pseudotime.Window/BAX/MDS.alt3p.junction.info.w.quantile.and.ct.counts.minUMI.1.MDSP5.MDSP6.txt", header=T)

psi.info.minUMI.1 <- read.csv("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/16.Pseudotime.Window/BAX/MDS.alt3p.junction.info.w.quantile.and.ct.counts.minUMI.1.MDSP5.txt", header=T)
bax.psi.info.minUMI.1 <- psi.info.minUMI.1[which(psi.info.minUMI.1$gene =="BAX"),]

psi.info.minUMI.2 <- read.csv("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/16.Pseudotime.Window/BAX/MDS.alt3p.junction.info.w.quantile.and.ct.counts.minUMI.2.MDSP5.txt", header=T)
bax.psi.info.minUMI.2 <- psi.info.minUMI.2[which(psi.info.minUMI.2$gene =="BAX"),]

psi.info.minUMI.3 <- read.csv("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/16.Pseudotime.Window/BAX/MDS.alt3p.junction.info.w.quantile.and.ct.counts.minUMI.3.MDSP5.txt", header=T)
bax.psi.info.minUMI.3 <- psi.info.minUMI.3[which(psi.info.minUMI.3$gene =="BAX"),]



psi.info.minUMI.4 <- read.csv("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/16.Pseudotime.Window/BAX/MDS.alt3p.junction.info.w.quantile.and.ct.counts.minUMI.4.txt", header=T)
bax.psi.info.minUMI.4 <- psi.info.minUMI.4[which(psi.info.minUMI.4$gene =="BAX"),]



```



##Looking specifically at BAX - how do the mut PSI and wt PSI look first by CD71 pseudotime quantile

```{r}
#bax.psi.info.minUMI.1.sub <- bax.psi.info.minUMI.1[which(bax.psi.info.minUMI.1$clusterID == "clu_77300"),]
#bax.psi.info.minUMI.2.sub <- bax.psi.info.minUMI.2[which(bax.psi.info.minUMI.2$clusterID == "clu_77300"),]
bax.psi.info.minUMI.3.sub <- bax.psi.info.minUMI.3[which(bax.psi.info.minUMI.3$clusterID == "clu_77300"),]
#bax.psi.info.minUMI.4.sub <- bax.psi.info.minUMI.4[which(bax.psi.info.minUMI.4$clusterID == "clu_77300"),]

bax.info <- bax.psi.info.minUMI.3.sub

bax.info
```


```{r}
#ch.psi.info.minUMI.1 <- read.csv("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/16.Pseudotime.Window/BAX/CH.alt3p.junction.info.w.quantile.and.ct.counts.minUMI.1.CH259.CH305.txt", header=T)

ch.psi.info.minUMI.1 <- read.csv("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/16.Pseudotime.Window/BAX/CH.alt3p.junction.info.w.quantile.and.ct.counts.minUMI.1.CH259.txt", header=T)
ch.bax.psi.info.minUMI.1 <- ch.psi.info.minUMI.1[which(ch.psi.info.minUMI.1$gene =="BAX"),]

ch.psi.info.minUMI.2 <- read.csv("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/16.Pseudotime.Window/BAX/CH.alt3p.junction.info.w.quantile.and.ct.counts.minUMI.2.CH259.txt", header=T)
ch.bax.psi.info.minUMI.2 <- ch.psi.info.minUMI.2[which(ch.psi.info.minUMI.2$gene =="BAX"),]

bax.info <- ch.bax.psi.info.minUMI.2
```



##dot/line plot of the mut and wt PSI's per quantile

```{r}
##Grab the psi's per quantile

mut.wt.psi.info.per.quantile <- as.data.frame(matrix(ncol=1, nrow=4))
rownames(mut.wt.psi.info.per.quantile) <- c(1:4)
colnames(mut.wt.psi.info.per.quantile) <- c("Quantile")

mut.wt.psi.info.per.quantile$Quantile <- c("Q1","Q2","Q3","Q4")

mut.wt.psi.info.per.quantile$mut.junc.cov <- paste(bax.info[which(bax.info$intron.junction == "chr19:48960914:48961482:+"),c("Q1.mut.junc.cov","Q2.mut.junc.cov","Q3.mut.junc.cov","Q4.mut.junc.cov")])
mut.wt.psi.info.per.quantile$mut.cluster.cov <- paste(bax.info[which(bax.info$intron.junction == "chr19:48960914:48961482:+"),c("Q1.mut.cluster.cov","Q2.mut.cluster.cov","Q3.mut.cluster.cov","Q4.mut.cluster.cov")])
mut.wt.psi.info.per.quantile$mut.psi <- paste(bax.info[which(bax.info$intron.junction == "chr19:48960914:48961482:+"),c("Q1.mut.bulk.psi","Q2.mut.bulk.psi","Q3.mut.bulk.psi","Q4.mut.bulk.psi")])


mut.wt.psi.info.per.quantile$wt.junc.cov <- paste(bax.info[which(bax.info$intron.junction == "chr19:48960914:48961482:+"),c("Q1.wt.junc.cov","Q2.wt.junc.cov","Q3.wt.junc.cov","Q4.wt.junc.cov")])
mut.wt.psi.info.per.quantile$wt.cluster.cov <- paste(bax.info[which(bax.info$intron.junction == "chr19:48960914:48961482:+"),c("Q1.wt.cluster.cov","Q2.wt.cluster.cov","Q3.wt.cluster.cov","Q4.wt.cluster.cov")])

mut.wt.psi.info.per.quantile$wt.psi <- paste(bax.info[which(bax.info$intron.junction == "chr19:48960914:48961482:+"),c("Q1.wt.bulk.psi","Q2.wt.bulk.psi","Q3.wt.bulk.psi","Q4.wt.bulk.psi")])

mut.wt.psi.info.per.quantile$dPSI <- as.numeric(mut.wt.psi.info.per.quantile$mut.psi) - as.numeric(mut.wt.psi.info.per.quantile$wt.psi)

mut.wt.psi.info.per.quantile


```



##dot/line plot of the mut and wt PSI's per cell type 
```{r}
mut.wt.psi.info.per.ct <- as.data.frame(matrix(ncol=1, nrow=4))
rownames(mut.wt.psi.info.per.ct) <- c(1:4)
colnames(mut.wt.psi.info.per.ct) <- c("CellType")

mut.wt.psi.info.per.ct$CellType <- c("HSPC","IMP","MEP","EP")
mut.wt.psi.info.per.ct$mut.junc.cov <- paste(bax.info[which(bax.info$intron.junction == "chr19:48960914:48961482:+"),c("HSPC.mut.junc.cov","IMP.mut.junc.cov","MEP.mut.junc.cov","EP.mut.junc.cov")])
mut.wt.psi.info.per.ct$mut.cluster.cov <- paste(bax.info[which(bax.info$intron.junction == "chr19:48960914:48961482:+"),c("HSPC.mut.cluster.cov","IMP.mut.cluster.cov","MEP.mut.cluster.cov","EP.mut.cluster.cov")])
mut.wt.psi.info.per.ct$mut.psi <- paste(bax.info[which(bax.info$intron.junction == "chr19:48960914:48961482:+"),c("HSPC.mut.bulk.psi","IMP.mut.bulk.psi","MEP.mut.bulk.psi","EP.mut.bulk.psi")])

mut.wt.psi.info.per.ct$wt.junc.cov <- paste(bax.info[which(bax.info$intron.junction == "chr19:48960914:48961482:+"),c("HSPC.wt.junc.cov","IMP.wt.junc.cov","MEP.wt.junc.cov","EP.wt.junc.cov")])
mut.wt.psi.info.per.ct$wt.cluster.cov <- paste(bax.info[which(bax.info$intron.junction == "chr19:48960914:48961482:+"),c("HSPC.wt.cluster.cov","IMP.wt.cluster.cov","MEP.wt.cluster.cov","EP.wt.cluster.cov")])

mut.wt.psi.info.per.ct$wt.psi <- paste(bax.info[which(bax.info$intron.junction == "chr19:48960914:48961482:+"),c("HSPC.wt.bulk.psi","IMP.wt.bulk.psi","MEP.wt.bulk.psi","EP.wt.bulk.psi")])

mut.wt.psi.info.per.ct$dPSI <- as.numeric(mut.wt.psi.info.per.ct$mut.psi) - as.numeric(mut.wt.psi.info.per.ct$wt.psi)

mut.wt.psi.info.per.ct

```




```{r}
library(reshape2)
df <- melt(mut.wt.psi.info.per.quantile, measure.vars = c("mut.psi", "wt.psi"))
df$value <- as.numeric(df$value)
df$Quantile <- factor(df$Quantile, levels = c("Q1","Q2","Q3","Q4"))
ggplot(df, aes(x=Quantile, y=value, group=variable)) + geom_point(aes(y=value, color=variable)) +geom_line(aes(y=value, color=variable)) + theme_classic() + ggtitle("BAX-w isoform MUT and WT PSI's in CH305 [Genotype 1UMI threhsold]")


df <- melt(mut.wt.psi.info.per.ct, measure.vars = c("mut.psi", "wt.psi"))
df$value <- as.numeric(df$value)
df$CellType <- factor(df$CellType, levels = c("HSPC","IMP","MEP","EP"))
ggplot(df, aes(x=CellType, y=value, group=variable)) + geom_point(aes(y=value, color=variable)) +geom_line(aes(y=value, color=variable)) + theme_classic() + ggtitle("BAX-w isoform MUT and WT PSI's in CH305  [Genotype 1UMI threhsold]")
```


```{r}
library(reshape2)
df <- melt(mut.wt.psi.info.per.quantile, measure.vars = c("mut.psi", "wt.psi"))
df$value <- as.numeric(df$value)
df$Quantile <- factor(df$Quantile, levels = c("Q1","Q2","Q3","Q4"))
ggplot(df, aes(x=Quantile, y=value, group=variable)) + geom_point(aes(y=value, color=variable)) +geom_line(aes(y=value, color=variable)) + theme_classic() + ggtitle("BAX-w isoform MUT and WT PSI's in MDSP6 [Genotype 1UMI threhsold]")


df <- melt(mut.wt.psi.info.per.ct, measure.vars = c("mut.psi", "wt.psi"))
df$value <- as.numeric(df$value)
df$CellType <- factor(df$CellType, levels = c("HSPC","IMP","MEP","EP"))
ggplot(df, aes(x=CellType, y=value, group=variable)) + geom_point(aes(y=value, color=variable)) +geom_line(aes(y=value, color=variable)) + theme_classic() + ggtitle("BAX-w isoform MUT and WT PSI's in MDSP6  [Genotype 1UMI threhsold]")
```


##Getting junction count info
```{r}

```

##Check other apoptotic genes

```{r}
pro.apoptotic.gene.set <- read.table("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/16.Pseudotime.Window/BAX/reactome.apoptosis.gene.list.txt", header = T)

pro.apoptotic.gene.list <- paste(pro.apoptotic.gene.set$REACTOME_APOPTOSIS)
pro.apoptotic.gene.list 


junctions.bulk <- read.table("/gpfs/commons/groups/landau_lab/SF3B1_splice_project/7.Permutation_junction_output/6.Comb_patients_2WT/5_read_threshold/MDS_P5_P6/logOR_within_cell_type_ONLY_CRYPTIC_3P_Junctions_with_threshold_info.txt")

junctions.bulk[which(junctions.bulk$gene %in% pro.apoptotic.gene.list & junctions.bulk$dPSI >0 & junctions.bulk$pvalue <= 0.05),]
```


```{r}
junctions.bulk
```


#MDS
```{r}
p.val <- 0.1
junctions.bulk[which(junctions.bulk$gene %in%  pro.apoptotic.gene.list & junctions.bulk$dPSI >0 & junctions.bulk$pvalue <= p.val),]
junctions.hspc[which(junctions.hspc$gene %in%  pro.apoptotic.gene.list & junctions.hspc$dPSI >0 & junctions.hspc$pvalue <= p.val),]
junctions.imp[which(junctions.imp$gene %in%  pro.apoptotic.gene.list & junctions.imp$dPSI >0 & junctions.imp$pvalue <= p.val),]
junctions.mep[which(junctions.mep$gene %in%  pro.apoptotic.gene.list & junctions.mep$dPSI >0 & junctions.mep$pvalue <= p.val),]
junctions.ep[which(junctions.ep$gene %in%  pro.apoptotic.gene.list & junctions.ep$dPSI >0 & junctions.ep$pvalue <= p.val),]
junctions.mkp[which(junctions.mkp$gene %in%  pro.apoptotic.gene.list & junctions.mkp$dPSI >0 & junctions.mkp$pvalue <= p.val),]
junctions.np[which(junctions.np$gene %in%  pro.apoptotic.gene.list & junctions.np$dPSI >0 & junctions.np$pvalue <= p.val),]
```


#CH
```{r}
p.val <- 0.5
junctions.bulk[which(junctions.bulk$gene %in%  pro.apoptotic.gene.list & junctions.bulk$dPSI >0 & junctions.bulk$pvalue <= p.val),]
junctions.hspc[which(junctions.hspc$gene %in%  pro.apoptotic.gene.list & junctions.hspc$dPSI >0 & junctions.hspc$pvalue <= p.val),]
#junctions.mkp[which(junctions.mkp$gene %in%  pro.apoptotic.gene.list & junctions.mkp$dPSI >0 & junctions.mkp$pvalue <= p.val),]
junctions.mep[which(junctions.mep$gene %in%  pro.apoptotic.gene.list & junctions.mep$dPSI >0 & junctions.mep$pvalue <= p.val),]
junctions.ep[which(junctions.ep$gene %in%  pro.apoptotic.gene.list & junctions.ep$dPSI >0 & junctions.ep$pvalue <= p.val),]
junctions.imp[which(junctions.imp$gene %in%  pro.apoptotic.gene.list & junctions.imp$dPSI >0 & junctions.imp$pvalue <= p.val),]
#junctions.np[which(junctions.np$gene %in%  pro.apoptotic.gene.list & junctions.np$dPSI >0 & junctions.np$pvalue <= p.val),]
```

##What do we want to keep?

```{r}
junction.table <- junctions.ep
junction.table[which(junction.table$gene %in%  pro.apoptotic.gene.list & junction.table$dPSI >0 & junction.table$pvalue <= p.val),c("pvalue","intron_junction","gene" ,"obs.wt","obs.mut", "three.wt.cluster.cov","three.mut.cluster.cov","threep_distance","wt.psi","mut.psi","dPSI")]

```

##Look at the signal per patient

```{r}

```


